---
title: "TCOPIES Report (Draft)"
date: January 15, 2021
output: 
  bookdown::word_document2:
  fig_caption: yes
params:
  conventional_only: FALSE
---


```{r message=FALSE, warning=FALSE, include=F}
CONVENTIONAL_ONLY <- params$conventional_only
```

```{r message=FALSE, warning=FALSE, include=FALSE}
rm(list=grep("^CONVENTIONAL_ONLY$", ls(), invert = T, value = T))

pacman::p_load(openxlsx,tidyverse,knitr,flextable,officer,reshape2,summarytools,tableone,stringr,
               ggplot2,ggrepel,stringdist,igraph,gridExtra,VennDiagram,vegan,SPECIES,grid,dplyr,ineq,future.apply,
               circlize,RColorBrewer,RCircos,ggpubr,MASS,questionr,pander, officedown)
#pander( data, style='rmarkdown' ) 

source("helper_functions.R")

newBKM <- function(){
  if(!exists("BKM_CNT")){
    BKM_CNT <<- 0
  }
  
  BKM_CNT <<- BKM_CNT + 1
  return(paste0("tab", BKM_CNT))
}


# get data
tissue.path <- "../Daten/Tissue/"
out.path <- "../Output/_final/"
div.path <- "../Output/diversity_analysis/"
data.path <- "../Daten/All_rerun_corr/clean_tcopies_fold5_freq1e-10_ar2/"
wein.path <- "../Daten/weinberger_tcrb/data/"

load(file = paste0("TCopies_merged_corr.RData"))

save_plots=T
set.seed(1234)
rep.dwns=1000
tcopies.TCMGRset <- c("R42","R43","R106","R132","R172","R34")
wein.TCMGRset <- c("P03","P06","P07","P08","P15","P16", "P18", "P20", "P21", "P22", "P23", "P39", "P43", "P44")

# Paths to folders with results
folder_CD4_allo <- "../Output/network_tcopies_CD4_allo_LV1_final/"

# Read in tissue data
filename <- "vdj.R34_bulk_tissue_1_clones_allo.txt"
tissue_34 <- read.table(paste0(tissue.path,filename), header=T, stringsAsFactors = F)
filename <- "vdj.R43_bulk_tissue_1_clones_allo.txt"
tissue_43 <- read.table(paste0(tissue.path,filename), header=T, stringsAsFactors = F)

# Get files from directory
files.tcr <- grep(list.files(data.path), pattern=".txt",value=T)
files.tcr <- files.tcr[!(grepl("low_2", files.tcr))]
files.tcr <- files.tcr[!(grepl("high", files.tcr))]
files.tcr <- files.tcr[!(grepl("R40", files.tcr))]

# tbl_clonesR43 <- merged[["R43"]]
# new_43 <- full_join(tbl_clonesR43, tissue_43[,c("cdr3nt", "count", "freq")], by="cdr3nt")
# colnames(new_43)[(ncol(new_43)-1):ncol(new_43)] <- c("Tissue.count", "Tissue.freq")
# write.xlsx(new_43, paste0(out.path,"tbl_clonesR43.xlsx"), overwrite=T)
# saveRDS(new_43, file = paste0(out.path,"RObjects/","clones_R43.RData"))
# 
# tbl_clonesR34 <- merged[["R34"]]
# new_34 <- full_join(tbl_clonesR34, tissue_34[,c("cdr3nt", "count", "freq")], by="cdr3nt")
# colnames(new_34)[(ncol(new_34)-1):ncol(new_34)] <- c("Tissue.count", "Tissue.freq")
# saveRDS(new_34, file = paste0(out.path,"RObjects/","clones_R34.RData"))
# write.xlsx(new_34, paste0(out.path,"tbl_clonesR34.xlsx"), overwrite=T)
# 
# 
# for(i in 1:length(merged)){
#   fname <- extract_nameinfo(names(merged)[i])
#   saveRDS(merged[[i]], paste0(out.path,"RObjects/","clones_",fname[1],".RData"))
# }
```

# Data cleaning and downsampling


```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE}

df <- data.frame(Parameter=c("Ambiguity.Ratio", "Fold.change","Frequency","Downsampling.Reps"), Values=c(2, 5, format(0.0000000001, scientific = FALSE), rep.dwns))

 df%>% 
  flextable()  %>%
  autofit() %>%
  set_caption("Parameters for data cleaning and downsampling", autonum = run_autonum(seq_id = "tab", bkm = newBKM())) %>% 
  # bold header
  bold(part = "header") 

```


If after data cleaning, phenotype allocation remained ambiguous the mixed up clonotypes were assigned to the class in which its frequency exceeded twice the frequency occurrence of the other phenotype group. Further, non-productive clonotypes (identified by cdr3nt sequence containing stop or start codon * or _) were omitted.



<!-- ## Venn-Diagramms of sample R34 and R43 -->

<!-- ```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, , fig.width=4, fig.height=4} -->
<!-- # Venn R34 -->
<!-- blood_34$BL <- apply(blood_34[,grepl("BL_1.count",colnames(blood_34))],1, function(x) any(!is.na(x))) -->
<!-- blood_34$MLR <- apply(blood_34[,grepl(c("low_1.count"), colnames(blood_34)) | grepl(c("low_2.count"), colnames(blood_34))],1, function(x) any(!is.na(x))) -->
<!-- blood_34$Late <- apply(blood_34[,grepl("late_1.count", colnames(blood_34))],1, function(x) any(!is.na(x))) -->
<!-- new_blood <- blood_34[,c("cdr3nt","BL","MLR","Late")] -->
<!-- venn.list <- list() -->
<!-- venn.list[[1]] <- new_blood[blood_34$BL,]$cdr3nt -->
<!-- venn.list[[2]] <- new_blood[blood_34$MLR,]$cdr3nt -->
<!-- venn.list[[3]] <- new_blood[blood_34$Late,]$cdr3nt -->
<!-- venn.list[[4]] <- tissue_34$cdr3nt -->
<!-- names(venn.list) <- c("BL", "MLR", "Late", "Tissue") -->

<!-- plt1 <- venn.diagram( -->
<!--   x = venn.list, -->
<!--   category.names = c("BL", "allo", "Bx", "Tissue"), -->
<!--   main="R34", -->
<!--   filename = NULL) -->


<!-- # Venn R43 -->
<!-- blood_43$BL <- apply(blood_43[,grepl("BL_1.count",colnames(blood_43))],1, function(x) any(!is.na(x))) -->
<!-- blood_43$MLR <- apply(blood_43[,grepl(c("low_1.count"), colnames(blood_43)) | grepl(c("low_2.count"), colnames(blood_43))],1, function(x) any(!is.na(x))) -->
<!-- blood_43$Late <- apply(blood_43[,grepl("late_1.count", colnames(blood_43))],1, function(x) any(!is.na(x))) -->
<!-- new_blood <- blood_43[,c("cdr3nt","BL","MLR","Late")] -->

<!-- venn.list <- list() -->
<!-- venn.list[[1]] <- new_blood[blood_43$BL,]$cdr3nt -->
<!-- venn.list[[2]] <- new_blood[blood_43$MLR,]$cdr3nt -->
<!-- venn.list[[3]] <- new_blood[blood_43$Late,]$cdr3nt -->
<!-- venn.list[[4]] <- tissue_43$cdr3nt -->
<!-- names(venn.list) <- c("BL", "MLR", "Late", "Tissue") -->

<!-- plt2 <- venn.diagram( -->
<!--   x = venn.list, -->
<!--   category.names = c("BL", "allo", "Bx", "Tissue"), -->
<!--   main="R43", -->
<!--   filename = NULL) -->

<!-- grid.draw(plt1) -->
<!-- grid.newpage() -->
<!-- grid.draw(plt2) -->
<!-- ``` -->

# Requested Figures

<!-- ## Descriptive Analysis -->

<!-- ```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE} -->

<!-- analyse_samples <- function(df){ -->

<!--   ind <- apply(df[,2:ncol(df)], 1, function(x) all(is.na(x))) -->
<!--   df <- df[ !ind, ] -->
<!--   cdratio.tissue <- data.frame(cbind("Patient"=c(""),table(df$CDtype), prop.table(table(df$CDtype)))) -->

<!--   df$pheno <- substr(colnames(df)[apply(df,1,which.max)],1,3) -->
<!--   df$BL <- apply(df[,str_detect(names(df),"BL")], 1, function(x) any(!is.na(x)))*1 -->
<!--   df$MLR <- apply(df[,str_detect(names(df),"low")], 1, function(x) any(!is.na(x)))*1 -->
<!--   df$Late <- apply(df[,str_detect(names(df),"late")], 1, function(x) any(!is.na(x)))*1 -->

<!--   BL.count <- table(df[df$BL==1,]$BL, df[df$BL==1,]$pheno) -->
<!--   BL.prop <- prop.table(table(df[df$BL==1,]$BL, df[df$BL==1,]$pheno)) -->
<!--   BL.reads <- apply(df[,str_detect(colnames(df), "BL") & str_detect(colnames(df), "count")],2, function (x) sum(x, na.rm=T)) -->

<!--   MLR.count <- table(df[df$MLR==1,]$MLR, df[df$MLR==1,]$pheno) -->
<!--   MLR.prop <- prop.table(table(df[df$MLR==1,]$MLR, df[df$MLR==1,]$pheno)) -->
<!--   MLR.reads <- apply(df[,str_detect(colnames(df), "low_1") & str_detect(colnames(df), "count")],2, function (x) sum(x, na.rm=T)) -->

<!--   Late.count <- table(df[df$Late==1,]$Late, df[df$Late==1,]$pheno) -->
<!--   Late.prop <- prop.table(table(df[df$Late==1,]$Late, df[df$Late==1,]$pheno)) -->
<!--   Late.reads <- apply(df[,str_detect(colnames(df), "late") & str_detect(colnames(df), "count")],2, function (x) sum(x, na.rm=T)) -->

<!--   Overall.count <- table(df$pheno) -->
<!--   Overall.prop <- prop.table(table(df$pheno)) -->
<!--   Overall.reads <- BL.reads + MLR.reads + Late.reads -->

<!--   return(data.frame("CD4.count"=c(Overall.count[1], BL.count[1], MLR.count[1], Late.count[1]), -->
<!--              "CD4.prop"=c(Overall.prop[1], BL.prop[1], MLR.prop[1], Late.prop[1]), -->
<!--             "CD4.reads"=c(Overall.reads[1], BL.reads[1], MLR.reads[1], Late.reads[1]), -->
<!--              "CD8.count"=c(Overall.count[2], BL.count[2], MLR.count[2], Late.count[2]), -->
<!--              "CD8.prop"=c(Overall.prop[2], BL.prop[2], MLR.prop[2], Late.prop[2]), -->
<!--             "CD8.reads"=c(Overall.reads[2], BL.reads[2], MLR.reads[2], Late.reads[2]), -->
<!--              "Total.count"=c(Overall.count[1]+Overall.count[2], BL.count[1]+BL.count[2], MLR.count[1]+MLR.count[2], -->
<!--                              Late.count[1]+ Late.count[2]), -->
<!--             "Total.reads"=c(Overall.reads[1]+Overall.reads[2], BL.reads[1]+BL.reads[2], MLR.reads[1]+MLR.reads[2], -->
<!--                             Late.reads[1]+ Late.reads[2]))) -->
<!-- } -->
<!-- plan(multisession, workers=5) -->
<!-- tmp.phenotypes <- future_lapply(merged, function(x) analyse_samples(x)) -->
<!-- plan(sequential) -->
<!-- df.phenotypes <- do.call(rbind, tmp.phenotypes) -->
<!-- df.phenotypes$Patient <- rep(names(merged), each=4) -->
<!-- df.phenotypes$Stage <- rep(c("Overall", "PreTX", "donor-reactive", "PostTX")) -->

<!-- df.phenotypes  <- df.phenotypes %>% -->
<!--   relocate(Patient, .before=CD4.count) %>% -->
<!--   relocate(Stage, .before=CD4.count) -->

<!-- write.xlsx(df.phenotypes, paste0(out.path,"table_tcopies_phenodistr_blood.xlsx"), overwrite=T) -->
<!-- ``` -->


## Unique clone counts BL, MLR and Late for CD4 and CD8
Number of unique clonotypes was assessed based on the number of distinct and unique CDR3 sequences in the pre-transplant (unstimulated), the mixed lymphocyte reaction (donor-reactive) and the post-transplant (BX) sample repertoires. Each dot in Figure 3.1 represents an observed value in the patient set. The distribution of the 12 observed values per group are displayed by boxplots. The box includes observed values within the inter-quartile range (IQR) with the whiskers stretching to the last observed value contained within 1.5*IQR from the lower and upper quartile.

```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=4, fig.cap="Absolute number of unique clonotypes grouped by phenotype groups CD4 and CD8 for the unstimulated and the donor-reactive repertoire samples"}
get_uc <- function(filename){
   
  #print(filename)
  tmp <- read.table(paste0(data.path,filename), header=T, stringsAsFactors = F)
  unique.clones <- length(unique(tmp$cdr3nt))
  unique.allo <- length(unique(tmp[tmp$allo==1,]$cdr3nt))
  file.info <- extract_nameinfo(filename)
  
  return(c(file.info[2:5], unique.clones, unique.allo))
}

if(!file.exists("table_tcopies_unique_clones.xlsx")){
 df_uc <- data.frame(t(sapply(files.tcr, function(filename) get_uc(filename))))
 df_uc$V5 <- as.numeric(df_uc$V5)
 df_uc[df_uc$stage == "BL1",]$stage <- "PreTX"
 df_uc[df_uc$stage == "low1",]$stage <- "donor-reactive"
 df_uc[df_uc$stage == "late1",]$stage <- "PostTX"
 
 df_uc$tmp <- paste0(df_uc$CDtype," " ,df_uc$stage)
 df_uc$tmp <- factor(df_uc$tmp,levels = c("CD4 unstim", "CD8 unstim", "CD4 donor-reactive", "CD8 donor-reactive", "CD4 BX", "CD8 BX"))
 df_uc$stage <- factor(df_uc$stage, levels=c("PreTX","PostTX","donor-reactive"))
 write.xlsx(df_uc, paste0(out.path, "table_tcopies_unique_clones.xlsx"), overwrite=T)
}

df_uc <- read.xlsx(paste0(out.path, "table_tcopies_unique_clones.xlsx"))
df_uc$stage <- factor(df_uc$stage, levels=c("PreTX","PostTX","donor-reactive"))


my_cols <-brewer.pal(8, "Paired")[c(1,2,7)]
p1 <- ggplot(df_uc, aes(x = stage, y = V5)) +
  stat_boxplot(geom ='errorbar', width = 0.2) +
  geom_boxplot(aes(x=stage, y=V5, fill=stage),width=0.5) +
  geom_dotplot(binaxis='y', stackdir='center', position=position_dodge(1), dotsize=0.3) +
  scale_fill_manual(values=my_cols) +
  scale_y_continuous(expand=c(0,0),"Unique clones",limits=c(0,300000), breaks = seq(0,300000, 50000)) +
  scale_x_discrete(expand=c(0,0),"",guide = guide_axis(angle = 45)) +
  theme_bw() + 
  theme(legend.position = "none", text = element_text(size=11)) +
  facet_wrap(~CDtype)

filename <- paste0("plot_uniqueclones_tcopies")
p1
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p1, width = 20, height = 20, dpi = 640, units="cm", limitsize = F)
```


```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE}
# ggplot(df_uc, aes(x = tmp, y = V5)) + 
#   geom_violin(trim = FALSE)+
#   geom_dotplot(aes(x = tmp, y = V5, fill=stage),binaxis='y', stackdir='center', dotsize = 0.75) +
#   scale_fill_brewer(palette="Dark2") +
#   scale_y_continuous(expand=c(0,0),"unique clones",limits=c(0,300000), breaks = seq(0,300000, 50000))  +
#   scale_x_discrete(expand=c(0,0),"",guide = guide_axis(angle = 45)) +
#   theme_bw()

anova.cd4 <- summary(aov(df_uc[df_uc$CDtype %in% "CD4",]$V5 ~ df_uc[df_uc$CDtype %in% "CD4",]$stage))[[1]]

anova.cd8 <- summary(aov(df_uc[df_uc$CDtype %in% "CD8",]$V5 ~ df_uc[df_uc$CDtype %in% "CD8",]$stage))[[1]]

anova.tcopies <- data.frame("Pheno"=c("CD4", "CD8"), rbind(anova.cd4[1,c(4,1,3,2,5)], anova.cd8[1,c(4,1,3,2,5)]))

anova.tcopies %>%
  mutate(F.value = sprintf("%0.3f", F.value)) %>%
  mutate(Mean.Sq = sprintf("%0.3f", Mean.Sq)) %>%
  mutate(Sum.Sq = sprintf("%0.3f", Sum.Sq)) %>%
  mutate(Pr..F. = sprintf("%0.3f", Pr..F.)) %>%
  `colnames<-`(c("Pheno","Statistic", "df", "Mean squared", "Sum squared","p")) %>%
  data.frame() %>%
  flextable() %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Analysis of variance (ANOVA) for multiple group comparsion of unqiue clone counts grouped by phenotype", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))
  
ttest.uc1 <- t.test(df_uc[df_uc$CDtype %in% "CD4" & df_uc$stage %in% "PreTX",]$V5, df_uc[df_uc$CDtype %in% "CD4" &df_uc$stage %in% "PostTX",]$V5, paired=T)
ttest.uc2 <- t.test(df_uc[df_uc$CDtype %in% "CD4" & df_uc$stage %in% "PreTX",]$V5, df_uc[df_uc$CDtype %in% "CD4" &df_uc$stage %in% "donor-reactive",]$V5, paired=T)
ttest.uc3 <- t.test(df_uc[df_uc$CDtype %in% "CD4" & df_uc$stage %in% "PostTX",]$V5, df_uc[df_uc$CDtype %in% "CD4" &df_uc$stage %in% "donor-reactive",]$V5, paired=T)

ttest.uc4 <- t.test(df_uc[df_uc$CDtype %in% "CD8" & df_uc$stage %in% "PreTX",]$V5, df_uc[df_uc$CDtype %in% "CD8" &df_uc$stage %in% "PostTX",]$V5, paired=T)
ttest.uc5 <- t.test(df_uc[df_uc$CDtype %in% "CD8" & df_uc$stage %in% "PreTX",]$V5, df_uc[df_uc$CDtype %in% "CD8" &df_uc$stage %in% "donor-reactive",]$V5, paired=T)
ttest.uc6 <- t.test(df_uc[df_uc$CDtype %in% "CD8" & df_uc$stage %in% "PostTX",]$V5, df_uc[df_uc$CDtype %in% "CD8" &df_uc$stage %in% "donor-reactive",]$V5, paired=T)

ttest.uc <- data.frame("Pheno"=c(rep("CD4", 3),rep("CD8",3)), "Groups"=rep(c("Pre-Post", "Pre-allo", "Post-allo"),2),
                          "t"=c(ttest.uc1$statistic, ttest.uc2$statistic, ttest.uc3$statistic, ttest.uc4$statistic, ttest.uc5$statistic, ttest.uc6$statistic),
                          "df"=c(ttest.uc1$parameter, ttest.uc2$parameter, ttest.uc3$parameter, ttest.uc4$parameter, ttest.uc5$parameter, ttest.uc6$parameter), 
                          "p"=c(ttest.uc1$p.value, ttest.uc2$p.value, ttest.uc3$p.value, ttest.uc4$p.value, ttest.uc5$p.value, ttest.uc6$p.value), 
                          "Mean diff."=c(ttest.uc1$estimate[1], ttest.uc2$estimate[1], ttest.uc3$estimate[1], 
                                         ttest.uc4$estimate[1], ttest.uc5$estimate[1], ttest.uc6$estimate[1]),
                          "ci.lo"=c(ttest.uc1$conf.int[1],ttest.uc2$conf.int[1],ttest.uc3$conf.int[1],
                                    ttest.uc4$conf.int[1],ttest.uc5$conf.int[1],ttest.uc6$conf.int[1]),
                          "ci.up"=c(ttest.uc1$conf.int[2],ttest.uc2$conf.int[2],ttest.uc3$conf.int[2],
                                    ttest.uc4$conf.int[2],ttest.uc4$conf.int[2],ttest.uc6$conf.int[2]))
ttest.uc[,c(3:8)] <- apply(ttest.uc[,c(3:8)],2, function(x) round(x,3))

ttest.uc %>% 
  unite("CI", ci.lo:ci.up, sep=",") %>%
  mutate(CI = paste0("(",CI,")")) %>%
  `colnames<-`(c("Phenotype","Groups", "t", "df", "p", "Mean Difference", "95% CI")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Results of the paired t test assessing differences in unique clonotypes in the pre-and donor-reactive repertoire", autonum = run_autonum(seq_id = "tab", bkm = newBKM())) 

```



<!-- ## VDJ overlap -->

<!-- Figure 3.2 shows a circos plot of V and J pairings in the donor-reactive repertoire of a selected sample (R106 CD4 donor-reactive).  Arcs correspond to unique V and J segments which were scaled to their frequency in sample. The links represent pairings of V- and J-segments and their intensity is scaled to the pairing frequency. -->

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=6, fig.cap="V-J junction circos plot for a single donor-reactive sample."}

# load data and preproc to fit formats
plot_VJcircos <- function(data.tcr, file.tcr=NULL, name=NULL){
  # data.tcr; file.tcr=files.tcr[3]
  if(!is.null(file.tcr)){file.info <- extract_nameinfo(file.tcr)
  }else{file.info <- name}

  mat <- prop.table(wtd.table(data.tcr$v,data.tcr$j, weights=data.tcr$count))

  col_sum = apply(mat, 2, sum)
  row_sum = apply(mat, 1, sum)

  mat <- mat[order(row_sum), order(col_sum)]

  # equal number of characters for visualizaiton

  rn <- rownames(mat)
  cn <- colnames(mat)

  maxrn <- max(nchar(rn))
  maxcn <- max(nchar(cn))

  for(i in seq_len(length(rn))) {
        rn[i] <- paste(rn[i], paste(rep(" ", maxrn - nchar(rn[i])), collapse = ''))
  }

  for(i in seq_len(length(cn))) {
        cn[i] <- paste(cn[i], paste(rep(" ", maxcn - nchar(cn[i])), collapse = ''))
  }

  rownames(mat) <- rn
  colnames(mat) <- cn

  # viz using circlize
  #ncolors <- c(rep(floor(nrow(mat)/3),2), floor(nrow(mat)/3) + nrow(mat)%%3)
 # rcols <- c(colorRampPalette(brewer.pal(9,"Oranges"))(ncolors[1]),colorRampPalette(brewer.pal(9,"Purples"))(ncolors[2]), colorRampPalette(brewer.pal(9,"Greens"))(ncolors[3]))
  colfunc<-colorRampPalette(c("red","yellow","springgreen4"))
  rcols <- colfunc(nrow(mat))
  ccols <- colorRampPalette(brewer.pal(9,"Blues"))(ncol(mat))

  names(rcols) <- gtools::mixedsort(rownames(mat))
  names(ccols) <- gtools::mixedsort(colnames(mat))

  col_fun = colorRamp2(range(mat), c("#FFEEEE", "#FF0000"), transparency = 0.25)

  if(!is.null(file.tcr)){
      filename <- paste0("plot_VJusage_", file.info["patID"],"_",file.info["stage"], "_", file.info["CDtype"])
  }else{filename <- paste0("plot_VJusage_", file.info)}
  
  png(file = paste(out.path, filename,".png", sep = ""), width = 8*600, height = 6*600, units = "px", res = 600)
  circos.par(gap.degree = c(rep(2, nrow(mat)-1), 5, rep(1, ncol(mat)-1), 10), start.degree = 5)

  chordDiagram(mat, annotationTrack = "grid",
               grid.col = c(rcols, ccols), col=col_fun, order = c(names(rcols), names(ccols)),
               preAllocateTracks = list(track.height = 0.3), transparency = 0.5, reduce=0)

  circos.trackPlotRegion(track.index = 1, bg.border = NA,
         panel.fun = function(x, y) {
                     sector.name = get.cell.meta.data("sector.index")
                     xlim = get.cell.meta.data("xlim")
                     ylim = get.cell.meta.data("ylim")
                     circos.text(mean(xlim), ylim[1], cex = 0.4, sector.name, facing = "clockwise", adj = c(0, 0.3))
                     }
         )
  dev.off()

  circos.clear()

}


#Selected plot in report
data.tcr<- read.table(paste0(data.path, files.tcr[3]), header=T, stringsAsFactors =F)

plot_VJcircos(data.tcr, files.tcr[3])

# # Print and save all once
# for(i in 1:length(files.tcr)){
#   print(files.tcr[i])
#   data.tcr<- read.table(paste0(data.path, files.tcr[i]), header=T, stringsAsFactors =F)
#   plot_VJcircos(data.tcr, files.tcr[i])
# }


## Create legends for the circos plots
# library("grDevices")
# plot(1:20,pch=20,col=col)
# col=c(colorRampPalette(brewer.pal(9,"Oranges"))(ncolors[1]),colorRampPalette(brewer.pal(9,"Purples"))(ncolors[2]), colorRampPalette(brewer.pal(9,"Greens"))(ncolors[3]))
# colfunc<-colorRampPalette(c("red","yellow","green"))
# col=colfunc(63)
# #col=colorRampPalette(c("white", "red"))(20) 
# M <- as.matrix(seq(1:63),1,63)
# image(M,col = col,frame=F,xaxt="n",yaxt="n",main="V-genes")
# 
# 
# library("grDevices")
# plot(1:20,pch=20,col=col)
# col <- colorRampPalette(brewer.pal(9,"Blues"))(ncol(mat))
# M <- as.matrix(seq(1:13),1,13)
# image(M,col = col,frame=F,xaxt="n",yaxt="n",main="J-genes")


```


## Power law plots and slope estimation for the bulk component
The power law slope of each repertoire was computed by separating the frequency distribution consisting of clones with unique cell counts into two components: the bulk component and the high frequency component consisting of the small fraction f clones with a large unique cell count (following DeWolf et al. 2018). Class allocation was performed using the lowest frequency unique expanded clone as a cut-off value. The slope of the power law fit is then computed by fitting a linear regression between the log frequency and the log counts of the unique clones contained in the high frequency component and taking the absolute value of the effect size of the log counts.

```{r echo=FALSE, cache=T, cache=TRUE, message=FALSE, warning=FALSE, fig.align="center", fig.width=12, fig.height=16, fig.cap="Power law slopes of the bulk component of selected unstimulated repertoire samples grouped by phenotype"}
# The clonotype abundance frequency distribution  (x=frequency, y=number of clones with frequency x)
plot_powerLaw <- function(filename){
  sample <- read.table(paste0(data.path,filename), header=T, stringsAsFactors =F)

  tbl.counts <- table(sample$count)
  threshold.counts <- as.numeric(names(tbl.counts[tbl.counts == 1][2]))

  # Separate in bulk and high frequency component
  sample.bulk <- sample[sample$count < threshold.counts,]
  sample.hf <- sample[sample$count >= threshold.counts,]

  # Calculate power law slope of the bulk component
  clone_nr.bulk <- c(table(sample.bulk$freq))
  freq.bulk <- as.numeric(as.character(names(clone_nr.bulk)))
  names(clone_nr.bulk) <- NULL
  lm.fit <- lm(log(clone_nr.bulk)~log(freq.bulk))
  slope <- lm.fit$coefficients[2]

  # Include log clone nr and freq for plotting
  clone_nr.hf <- c(table(sample.hf$freq))
  freq.hf <- as.numeric(as.character(names(clone_nr.hf)))
  names(clone_nr.hf) <- NULL

  df.bulk = data.frame(log.clone=log(clone_nr.bulk), log.freq=log(freq.bulk),pred = predict(lm.fit), bulk=T)
  df.hf = data.frame(log.clone=log(clone_nr.hf), log.freq=log(freq.hf),pred = NA, bulk=F)

  df <- rbind(df.bulk, df.hf)
  file.info <- extract_nameinfo(filename)

  return(list("res"=c(file.info[2:5], "slope"=slope), "res.df"=df))
}

plot_PL <- function(list.df){
  res  <- list.df[[1]]
  df <- list.df[[2]]
  p <- ggplot(data=df, mapping=aes(x=log.freq, y=log.clone, col=bulk)) +
    geom_point() +
    geom_line(mapping=aes(y=pred))+
    labs(y=expression(log[10](Clone~number)), x=expression(log[10](Frequency)))  +
    scale_color_manual(values=c("#999999", "black"), name = "", labels = c("High-frequency component","Bulk component")) +
    scale_x_continuous(limits = c(-11,-2), breaks = seq(-11,-2,1)) +
    scale_y_continuous(limits = c(0,10)) +
    theme_bw() +
    ggtitle(paste0(res[[1]], " - ", res[2])) +
    annotate("text", x=-2.5, y=10, label= paste("slope = ",round(as.numeric(res[5]),1))) +
    theme(legend.position = "bottom", plot.title = element_text(hjust = 0.5), text = element_text(size=12))

  return(p)
}

rel.files = files.tcr[!str_detect(files.tcr,"late|low")]
df.powerslope <- lapply(rel.files, function(filename) plot_powerLaw(filename))
plots <- lapply(df.powerslope, function(x) plot_PL(x))
names(plots) <- rel.files
gridExtra::grid.arrange(grobs = plots[1:6], ncol=2, name="Baseline")

# for(i in 1:length(plots)){
#   file.info <- extract_nameinfo(names(plots)[i])
#   filename <- paste0("plot_PLslopes_", file.info["patID"],"_",file.info["stage"], "_", file.info["CDtype"])
#   png(file = paste(out.path, filename,".png", sep = ""), width = 8*600, height = 6*600, units = "px", res = 600)
#   print(plots[i])
#   dev.off()
# }


```


```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=12, fig.height=16, fig.cap="Power law slopes of the bulk component of selected donor-reactive repertoire samples grouped by phenotype"}
rel.files = files.tcr[!str_detect(files.tcr,"BL|late")]
df.powerslope <- lapply(rel.files, function(filename) plot_powerLaw(filename))
plots <- lapply(df.powerslope, function(x) plot_PL(x))
names(plots) <- rel.files
gridExtra::grid.arrange(grobs = plots[1:6], ncol=2, name="MLR")

# for(i in 1:length(plots)){
#   print(i)
#   file.info <- extract_nameinfo(names(plots)[i])
#   filename <- paste0("plot_PLslopes_", file.info["patID"],"_",file.info["stage"], "_", file.info["CDtype"])
#   ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p1, width = 20, height = 20, dpi = 640, units="cm", limitsize = F)
# }
```


```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=4, fig.height=3, fig.cap="Boxplot of the computed absolute values of the power law slopes grouped by phenotype groups CD4 and CD8 for the unstimulated and the donor-reactive repertoire samples"}
df.powerslope <- lapply(files.tcr, function(filename) plot_powerLaw(filename))
df.ps <- data.frame(do.call(Map, c(f = rbind, df.powerslope))$res)

df.ps[df.ps$stage == "BL1",]$stage <- "PreTX"
df.ps[df.ps$stage == "low1",]$stage <- "donor-reactive"
df.ps[df.ps$stage == "late1",]$stage <- "PostTX"
# df.ps$tmp <- paste0(df.ps$CDtype," " ,df.ps$stage)
# df.ps$tmp <- factor(df.ps$tmp,levels = c("CD4 unstim", "CD8 unstim", "CD4 donor-reactive", "CD8 donor-reactive", "CD4 BX", "CD8 BX"))
df.ps$slope.log.freq. <- abs(as.numeric(df.ps$slope.log.freq.))
df.ps <- df.ps[df.ps$stage %in% c("PreTX", "donor-reactive"),]
df.ps$stage <- factor(df.ps$stage, levels=c("PreTX", "donor-reactive"))

dat_text <- data.frame(
  label = c("p<0.001", "p<0.001"),
  CDtype   = c("CD4", "CD8"),
  stage     = c(1.5, 1.5),
  slope.log.freq.     = c(4.25, 4.25)
)

my_cols <-brewer.pal(8, "Paired")[c(1,7)]
p1<- ggplot(df.ps, aes(x = stage, y = slope.log.freq.)) +
  stat_boxplot(geom ='errorbar', width = 0.2) +
  geom_boxplot(aes(x=stage, y=slope.log.freq., fill=stage),width=0.5) +
  geom_dotplot(binaxis='y', stackdir='center', position=position_dodge(1), dotsize=0.5) +
  scale_y_continuous("Slope (S)", limits=c(0,4.5))  +
  scale_x_discrete("")  +
  geom_text(data    = dat_text,  mapping = aes(x = stage, y = slope.log.freq., label = label)) +
  scale_fill_manual(values=my_cols) +
  theme_bw() +
  theme(legend.position = "none", text = element_text(size=14)) +
  facet_wrap(~CDtype)


p1
ggsave(filename = paste(out.path,"plot_PLslopes_tcopies.png", sep = ""), width = 20, height = 20, dpi = 640, units="cm", limitsize = F)

tttest.slope.cd4 <- t.test(df.ps[df.ps$stage %in% "PreTX" & df.ps$CDtype %in% "CD4",]$slope.log.freq., df.ps[df.ps$stage %in% "donor-reactive" & df.ps$CDtype %in% "CD4",]$slope.log.freq., paired=T)
tttest.slope.cd8 <- t.test(df.ps[df.ps$stage %in% "PreTX" & df.ps$CDtype %in% "CD8",]$slope.log.freq., df.ps[df.ps$stage %in% "donor-reactive" & df.ps$CDtype %in% "CD8",]$slope.log.freq., paired=T)

ttest.slope <- data.frame("Pheno"=c("CD4", "CD8"), "Variable"=rep("PL slope",2),"t"=c(tttest.slope.cd4$statistic, tttest.slope.cd8$statistic),
                          "df"=c(tttest.slope.cd4$parameter, tttest.slope.cd8$parameter), "p"=c(tttest.slope.cd4$p.value, tttest.slope.cd8$p.value), "Mean
                          diff."=c(tttest.slope.cd4$estimate[1], tttest.slope.cd8$estimate[1]), "ci.lo"=c(tttest.slope.cd4$conf.int[1],tttest.slope.cd8$conf.int[1]),
                          "ci.up"=c(tttest.slope.cd4$conf.int[2], tttest.slope.cd8$conf.int[2]))
ttest.slope[,c(3:8)] <- apply(ttest.slope[,c(3:8)],2, function(x) round(x,3))

ttest.slope %>% 
  unite("CI", ci.lo:ci.up, sep=",") %>%
  mutate(CI = paste0("(",CI,")")) %>%
  `colnames<-`(c("Phenotype","Variable", "t", "df", "p", "Mean Difference", "95% CI")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Results of the paired t test assessing differences in the power law slopes in the pre-and donor-reactive repertoire", autonum = run_autonum(seq_id = "tab", bkm = newBKM())) 
  
```




## Jensen-Shannon Divergence (JSD) between the BL and BX sample for CD4 and CD8 grouped by rejection
In order to assess repertoire overlap between the pre-transplant and post-transplant repertoire,  the Jensen-Shannon divergence (JSD) was computed. The overlap measure quantifies the extent of overlap between pairs of repertoires in a standardized way and takes clone number as well as its frequency into account. The JSD is normalized in between 0 and 1 where 1 indicates a no overlap between the two repertoires.


```{r echo=FALSE, cache=T, message=FALSE, cache=TRUE, warning=FALSE, fig.align="center", fig.width=12, fig.height=6, fig.cap="Boxplot of the Jensen-Shannon divergence (JSD) between the pre-transplant and post-transplant samples grouped by the outcome T-cell mediated graft rejection"}
JensenShannonDivergence <- function(df1, df2, var_freq=F){

  if(!var_freq){
      df.merged <- dplyr::full_join(df1[,c("cdr3nt", "freq")],df2[,c("cdr3nt", "freq")], by="cdr3nt")
    p <- df.merged$freq.x
    q <- df.merged$freq.y
  }else{
    p <- df1
    q <- df2
  }
  p[is.na(p)] <- 0
  q[is.na(q)] <- 0
  
  JSD <- -sum(((p+q)/2)* log(((p+q)/2),2)) - (1/2)* (-sum(p * log(p,2), na.rm = T) - sum(q*log(q,2),na.rm = T))
  
  return(JSD)
}

wrapper_JSD <- function(x, x_filename=T){
  
  if(x_filename){
      dfP <- read.table(paste0(data.path,x[1]), header=T, stringsAsFactors = F)
      dfQ <- read.table(paste0(data.path,x[2]), header=T, stringsAsFactors = F)
  }else{
    dfP <- x[[1]]
    dfQ <- x[[2]]
  }
  
  JSD.all <- JensenShannonDivergence(dfP, dfQ)
  JSD.1000 <- JensenShannonDivergence(dfP[1:1000,], dfQ[1:1000,])
  JSD.500 <- JensenShannonDivergence(dfP[1:500,], dfQ[1:500,])
  JSD.100 <- JensenShannonDivergence(dfP[1:100,], dfQ[1:100,])
  
  return(c(JSD.all, JSD.1000, JSD.500, JSD.100))
}


get_VJpairs <- function(x, classes=2){
  dfP <- read.table(paste0(data.path,x[1]), header=T, stringsAsFactors = F)
  dfQ <- read.table(paste0(data.path,x[2]), header=T, stringsAsFactors = F)
  
  dfP$group <- paste0(dfP$v,"x",dfP$j)
  dfQ$group <- paste0(dfQ$v,"x",dfQ$j)
  
  P.vj <-  data.frame(prop.table(wtd.table(dfP$group, weights=dfP$count)))
  Q.vj <- data.frame(prop.table(wtd.table(dfQ$group, weights=dfQ$count)))

  df.merged <- dplyr::full_join(P.vj[,c("Var1","Freq")],Q.vj[,c("Var1","Freq")], by=c("Var1"))
  colnames(df.merged) <- c("VJ", "PreTX.freq", "PostTX.freq")
  df.merged$Ratio.PostPre <- df.merged$PostTX.freq/df.merged$PreTX.freq

  df.merged <- df.merged[order(df.merged$PreTX.freq, decreasing = T),]

  info <- extract_nameinfo(x[1])
  write.xlsx(df.merged, paste0(out.path, "tbl_vjpairs_", info[2],"_", info[3],".xlsx"), overwrite=T)
  
  JSD.vj.all <- JensenShannonDivergence(df1=df.merged$PreTX.freq, df2=df.merged$PostTX.freq, var_freq = T)
  
  P.vj <-  data.frame(prop.table(wtd.table(dfP[1:1000,]$group, weights=dfP[1:1000,]$count)))
  Q.vj <- data.frame(prop.table(wtd.table(dfQ[1:1000,]$group, weights=dfQ[1:1000,]$count)))
  df.merged <- dplyr::full_join(P.vj[,c("Var1", "Freq")],Q.vj[,c("Var1","Freq")], by=c("Var1"))
  colnames(df.merged) <- c("VJ", "PreTX.freq", "PostTX.freq")
  
  JSD.vj.1000 <- JensenShannonDivergence(df1=df.merged$PreTX.freq, df2=df.merged$PostTX.freq, var_freq = T)
  
  P.vj <-  data.frame(prop.table(wtd.table(dfP[1:500,]$group, weights=dfP[1:500,]$count)))
  Q.vj <- data.frame(prop.table(wtd.table(dfQ[1:500,]$group, weights=dfQ[1:500,]$count)))
  df.merged <- dplyr::full_join(P.vj[,c("Var1","Freq")],Q.vj[,c("Var1","Freq")], by=c("Var1"))
  colnames(df.merged) <- c("VJ", "PreTX.freq", "PostTX.freq")

  JSD.vj.500 <- JensenShannonDivergence(df1=df.merged$PreTX.freq, df2=df.merged$PostTX.freq, var_freq = T)
  
  P.vj <-  data.frame(prop.table(wtd.table(dfP[1:100,]$group, weights=dfP[1:100,]$count)))
  Q.vj <- data.frame(prop.table(wtd.table(dfQ[1:100,]$group, weights=dfQ[1:100,]$count)))
  df.merged <- dplyr::full_join(P.vj[,c("Var1", "Freq")],Q.vj[,c("Var1","Freq")], by=c("Var1"))
  colnames(df.merged) <- c("VJ", "PreTX.freq", "PostTX.freq")

  JSD.vj.100 <- JensenShannonDivergence(df1=df.merged$PreTX.freq, df2=df.merged$PostTX.freq, var_freq = T)

  
  return(c(JSD.vj.all, JSD.vj.1000, JSD.vj.500, JSD.vj.100))
  
}

# Jensen-Shannon Divergence for VJ pairs
if(!file.exists("table_tcopies_JSD_VJ.xlsx")){
 rel.files <- rbind(cbind(files.tcr[str_detect(files.tcr, "BL")], files.tcr[str_detect(files.tcr, "late")]),
                    cbind(files.tcr[str_detect(files.tcr, "BL")], files.tcr[str_detect(files.tcr, "low")]),
                    cbind(files.tcr[str_detect(files.tcr, "low")], files.tcr[str_detect(files.tcr, "late")]))
 JSD.val.vj <- apply(rel.files, 1, function(x) get_VJpairs(x))
 JSD.VJ <- data.frame(cbind(t(sapply(rel.files[,1], extract_nameinfo)), t(sapply(rel.files[,2], extract_nameinfo)), t(JSD.val.vj)))
 colnames(JSD.VJ)[11:14] <- c("JSD.all", "JSD.1000", "JSD.500", "JSD.100")
 JSD.VJ$rejection <- ifelse(JSD.VJ$patID %in% tcopies.TCMGRset, "Rejectors",  "Controls")
 JSD.VJ[, 11:14]<- apply(JSD.VJ[, 11:14], 2, as.numeric)
 write.xlsx(JSD.VJ, paste0(out.path, "table_tcopies_JSD_VJ.xlsx"), overwrite=T)
}

JSD.VJ <- read.xlsx(paste0(out.path, "table_tcopies_JSD_VJ.xlsx"))


# # Jensen-Shannon Divergence for clonal distribution
if(!file.exists("table_tcopies_JSD.xlsx")){
 JSD.val <- apply(rel.files, 1, wrapper_JSD)
 JSD.df <- data.frame(cbind(t(sapply(rel.files[,1], extract_nameinfo)), t(sapply(rel.files[,2], extract_nameinfo)), t(JSD.val)))
 colnames(JSD.df)[11:14] <- c("JSD.all", "JSD.1000", "JSD.500", "JSD.100")
 JSD.df$rejection <- ifelse(JSD.df$patID %in% tcopies.TCMGRset, "Rejectors",  "Controls")
 JSD.df[, 11:14]<- apply(JSD.df[, 11:14], 2, as.numeric)
 write.xlsx(JSD.df, paste0(out.path, "table_tcopies_JSD.xlsx"), overwrite=T)
}

JSD.df <- read.xlsx(paste0(out.path, "table_tcopies_JSD.xlsx"))

# JSD.df %>%
#   flextable %>%
#   autofit %>%
#   colformat_num(digits=2) %>%
#   bold(part = "header") %>%
#   set_caption("JSD)", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))


JSD.CD4 <- JSD.df[JSD.df$CDtype %in% "CD4" & JSD.df$stage %in% "BL1" & JSD.df$stage.1 %in% "late1",]
JSD.CD8 <- JSD.df[JSD.df$CDtype %in% "CD8" & JSD.df$stage %in% "BL1" & JSD.df$stage.1 %in% "late1",]

p1 <- ggplot(JSD.CD4, aes(x = rejection, y = JSD.all)) +
  stat_boxplot(geom ='errorbar', width = 0.2) +
  geom_boxplot(aes(x=rejection, y=JSD.all, fill=rejection),width=0.5) +
  geom_dotplot(binaxis='y', stackdir='center', position=position_dodge(1), dotsize=0.5) +
  scale_y_continuous(expand=c(0,0),"Jensen-Shannon Divergence", limits = c(0,1), breaks = seq(0,1,0.1))  +
  scale_x_discrete(expand=c(0,0)," ", labels=c("0"="Controls", "1"="Rejectors")) +
  scale_fill_brewer(palette="Dark2","") +
  ggtitle("CD4") +
  theme_bw() +
  theme(legend.position = "none", plot.title = element_text(hjust = 0.5), text = element_text(size=22))


p2 <- ggplot(JSD.CD8, aes(x = rejection.1, y = JSD.all)) +
  stat_boxplot(geom ='errorbar', width = 0.2) +
  geom_boxplot(aes(x=rejection.1, y=JSD.all, fill=rejection.1),width=0.5) +
  geom_dotplot(binaxis='y', stackdir='center', position=position_dodge(1), dotsize=0.5) +
  scale_y_continuous(expand=c(0,0),"", limits = c(0,1), breaks = seq(0,1,0.1))  +
  scale_x_discrete(expand=c(0,0), "", labels=c("0"="Controls", "1"="Rejectors"))  +
  scale_fill_brewer(palette="Dark2","") +
  ggtitle("CD8") +
  theme_bw() +
  theme(legend.position = "none", plot.title = element_text(hjust = 0.5), text = element_text(size=22),
        axis.text.y=element_blank(), axis.ticks.y=element_blank())


p.JSD <- ggarrange(p1,p2, nrow=1, common.legend = T)
p.JSD
ggsave(filename = paste(out.path,"plot_tcopies_JSD_BLvsLate.png", sep = ""), plot=p.JSD, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)

# Only the top 1000
p3 <- ggplot(JSD.CD4, aes(x = rejection, y = JSD.1000)) +
  stat_boxplot(geom ='errorbar', width = 0.2) +
  geom_boxplot(aes(x=rejection, y=JSD.1000, fill=rejection),width=0.5) +
  geom_dotplot(binaxis='y', stackdir='center', position=position_dodge(1), dotsize=0.5) +
  scale_y_continuous(expand=c(0,0),"Jensen-Shannon Divergence", limits = c(0,0.15), breaks = seq(0,0.15,0.05))  +
  scale_x_discrete(expand=c(0,0)," ", labels=c("0"="Controls", "1"="Rejectors")) +
  scale_fill_brewer(palette="Dark2","") +
  ggtitle("CD4") +
  theme_bw() +
  theme(legend.position = "none", plot.title = element_text(hjust = 0.5), text = element_text(size=22))


p4 <- ggplot(JSD.CD8, aes(x = rejection.1, y = JSD.1000)) +
  stat_boxplot(geom ='errorbar', width = 0.2) +
  geom_boxplot(aes(x=rejection.1, y=JSD.1000, fill=rejection.1),width=0.5) +
  geom_dotplot(binaxis='y', stackdir='center', position=position_dodge(1), dotsize=0.5) +
  scale_y_continuous(expand=c(0,0),"",  limits = c(0,0.15), breaks = seq(0,0.15,0.05))  +
  scale_x_discrete(expand=c(0,0), "", labels=c("0"="Controls", "1"="Rejectors"))  +
  scale_fill_brewer(palette="Dark2","") +
  ggtitle("CD8") +
  theme_bw() +
  theme(legend.position = "none", plot.title = element_text(hjust = 0.5), text = element_text(size=22),
        axis.text.y=element_blank(), axis.ticks.y=element_blank())


p.JSD <- ggarrange(p3,p4, nrow=1, common.legend = T)
p.JSD
ggsave(filename = paste(out.path,"plot_tcopies_JSD_BLvsLate_Top1000.png", sep = ""), plot=p.JSD, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)


# Hypothesis testing
ttest.1 <- t.test(JSD.CD4[JSD.CD4$rejection %in% "Rejectors",]$JSD.all, JSD.CD4[JSD.CD4$rejection %in% "Controls",]$JSD.all,paired=F)
ttest.2 <- t.test(JSD.CD8[JSD.CD8$rejection %in% "Rejectors",]$JSD.all, JSD.CD8[JSD.CD8$rejection %in% "Controls",]$JSD.all,paired=F)

ttest.full <- data.frame("Pheno"=c("CD4", "CD8"), "Groups"=rep(c("Pre-Post"),2),
                          "t"=c(ttest.1$statistic, ttest.2$statistic),
                          "df"=c(ttest.1$parameter, ttest.2$parameter), 
                          "p"=c(ttest.1$p.value, ttest.2$p.value), 
                          "Mean diff."=c(ttest.1$estimate[1], ttest.2$estimate[1]),
                          "ci.lo"=c(ttest.1$conf.int[1],ttest.2$conf.int[1]),
                          "ci.up"=c(ttest.1$conf.int[2],ttest.2$conf.int[2]))
ttest.full[,c(3:8)] <- apply(ttest.full[,c(3:8)],2, function(x) round(x,3))

ttest.full %>% 
  unite("CI", ci.lo:ci.up, sep=",") %>%
  mutate(CI = paste0("(",CI,")")) %>%
  `colnames<-`(c("Phenotype","Groups", "t", "df", "p", "Mean Difference", "95% CI")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Results of unpaired Student's t testing to assess differences in overall JSD between PreTX and PostTX in rejectors and controls", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))
```


```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=12, fig.height=6, fig.cap="Boxplot of the Jensen-Shannon divergence (JSD) between the pre-transplant and donor-reactive repertoire samples grouped by the outcome T-cell mediated graft rejection"}
JSD.CD4 <- JSD.df[JSD.df$CDtype %in% "CD4" & JSD.df$stage %in% "BL1" & JSD.df$stage.1 %in% "low1",]
JSD.CD8 <- JSD.df[JSD.df$CDtype %in% "CD8" & JSD.df$stage %in% "BL1" & JSD.df$stage.1 %in% "low1",]

p1 <- ggplot(JSD.CD4, aes(x = rejection, y = JSD.all)) +
  stat_boxplot(geom ='errorbar', width = 0.2) +
  geom_boxplot(aes(x=rejection, y=JSD.all, fill=rejection),width=0.5) +
  geom_dotplot(binaxis='y', stackdir='center', position=position_dodge(1), dotsize=0.5) +
  scale_y_continuous(expand=c(0,0),"Jensen-Shannon Divergence", limits = c(0,1), breaks = seq(0,1,0.1))  +
  scale_x_discrete(expand=c(0,0)," ", labels=c("0"="Controls", "1"="Rejectors")) +
  scale_fill_brewer(palette="Dark2","") +
  ggtitle("CD4") +
  theme_bw() +
  theme(legend.position = "none", plot.title = element_text(hjust = 0.5), text = element_text(size=18))


p2 <- ggplot(JSD.CD8, aes(x = rejection.1, y = JSD.all)) +
  stat_boxplot(geom ='errorbar', width = 0.2) +
  geom_boxplot(aes(x=rejection.1, y=JSD.all, fill=rejection.1),width=0.5) +
  geom_dotplot(binaxis='y', stackdir='center', position=position_dodge(1), dotsize=0.5) +
  scale_y_continuous(expand=c(0,0),"", limits = c(0,1), breaks = seq(0,1,0.1))  +
  scale_x_discrete(expand=c(0,0), "", labels=c("0"="Controls", "1"="Rejectors"))  +
  scale_fill_brewer(palette="Dark2","") +
  ggtitle("CD8") +
  theme_bw() +
  theme(legend.position = "none", plot.title = element_text(hjust = 0.5), text = element_text(size=18),
        axis.text.y=element_blank(), axis.ticks.y=element_blank())


p.JSD <- ggarrange(p1,p2, nrow=1, common.legend = T)
p.JSD
ggsave(filename = paste(out.path,"plot_JSD_BLvsMLR_tcopies.png", sep = ""), plot=p.JSD, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)


ttest.1 <- t.test(JSD.CD4[JSD.CD4$rejection %in% "Rejectors",]$JSD.all, JSD.CD4[JSD.CD4$rejection %in% "Controls",]$JSD.all,paired=F)
ttest.2 <- t.test(JSD.CD8[JSD.CD8$rejection %in% "Rejectors",]$JSD.all, JSD.CD8[JSD.CD8$rejection %in% "Controls",]$JSD.all,paired=F)

ttest.full <- data.frame("Pheno"=c("CD4", "CD8"), "Groups"=rep(c("Pre-allo"),2),
                          "t"=c(ttest.1$statistic, ttest.2$statistic),
                          "df"=c(ttest.1$parameter, ttest.2$parameter), 
                          "p"=c(ttest.1$p.value, ttest.2$p.value), 
                          "Mean diff."=c(ttest.1$estimate[1], ttest.2$estimate[1]),
                          "ci.lo"=c(ttest.1$conf.int[1],ttest.2$conf.int[1]),
                          "ci.up"=c(ttest.1$conf.int[2],ttest.2$conf.int[2]))
ttest.full[,c(3:8)] <- apply(ttest.full[,c(3:8)],2, function(x) round(x,3))

ttest.full %>% 
  unite("CI", ci.lo:ci.up, sep=",") %>%
  mutate(CI = paste0("(",CI,")")) %>%
  `colnames<-`(c("Phenotype","Groups", "t", "df", "p", "Mean Difference", "95% CI")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Results of unpaired Student's t testing to assess differences in overall JSD between PreTX and allo in rejectors and controls", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))
```


```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=12, fig.height=6, fig.cap="Boxplot of the Jensen-Shannon divergence (JSD) between the donor-reactive and post-transplant repertoire samples grouped by the outcome T-cell mediated graft rejection"}
JSD.CD4 <- JSD.df[JSD.df$CDtype %in% "CD4" & JSD.df$stage %in% "low1" & JSD.df$stage.1 %in% "late1",]
JSD.CD8 <- JSD.df[JSD.df$CDtype %in% "CD8" & JSD.df$stage %in% "low1" & JSD.df$stage.1 %in% "late1",]

p1 <- ggplot(JSD.CD4, aes(x = rejection, y = JSD.all)) +
  stat_boxplot(geom ='errorbar', width = 0.2) +
  geom_boxplot(aes(x=rejection, y=JSD.all, fill=rejection),width=0.5) +
  geom_dotplot(binaxis='y', stackdir='center', position=position_dodge(1), dotsize=0.5) +
  scale_y_continuous(expand=c(0,0),"Jensen-Shannon Divergence", limits = c(0,1), breaks = seq(0,1,0.1))  +
  scale_x_discrete(expand=c(0,0)," ", labels=c("0"="Controls", "1"="Rejectors")) +
  scale_fill_brewer(palette="Dark2","") +
  ggtitle("CD4") +
  theme_bw() +
  theme(legend.position = "none", plot.title = element_text(hjust = 0.5), text = element_text(size=18))


p2 <- ggplot(JSD.CD8, aes(x = rejection.1, y = JSD.all)) +
  stat_boxplot(geom ='errorbar', width = 0.2) +
  geom_boxplot(aes(x=rejection.1, y=JSD.all, fill=rejection.1),width=0.5) +
  geom_dotplot(binaxis='y', stackdir='center', position=position_dodge(1), dotsize=0.5) +
  scale_y_continuous(expand=c(0,0),"", limits = c(0,1), breaks = seq(0,1,0.1))  +
  scale_x_discrete(expand=c(0,0), "", labels=c("0"="Controls", "1"="Rejectors"))  +
  scale_fill_brewer(palette="Dark2","") +
  ggtitle("CD8") +
  theme_bw() +
  theme(legend.position = "none", plot.title = element_text(hjust = 0.5), text = element_text(size=18),
        axis.text.y=element_blank(), axis.ticks.y=element_blank())


p.JSD <- ggarrange(p1,p2, nrow=1, common.legend = T)
p.JSD
ggsave(filename = paste(out.path,"plot_JSD_MLRvsLate_tcopies.png", sep = ""), plot=p.JSD, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)


ttest.1 <- t.test(JSD.CD4[JSD.CD4$rejection %in% "Rejectors",]$JSD.all, JSD.CD4[JSD.CD4$rejection %in% "Controls",]$JSD.all,paired=F)
ttest.2 <- t.test(JSD.CD8[JSD.CD8$rejection %in% "Rejectors",]$JSD.all, JSD.CD8[JSD.CD8$rejection %in% "Controls",]$JSD.all,paired=F)

ttest.full <- data.frame("Pheno"=c("CD4", "CD8"), "Groups"=rep(c("Post-allo"),2),
                          "t"=c(ttest.1$statistic, ttest.2$statistic),
                          "df"=c(ttest.1$parameter, ttest.2$parameter), 
                          "p"=c(ttest.1$p.value, ttest.2$p.value), 
                          "Mean diff."=c(ttest.1$estimate[1], ttest.2$estimate[1]),
                          "ci.lo"=c(ttest.1$conf.int[1],ttest.2$conf.int[1]),
                          "ci.up"=c(ttest.1$conf.int[2],ttest.2$conf.int[2]))
ttest.full[,c(3:8)] <- apply(ttest.full[,c(3:8)],2, function(x) round(x,3))

ttest.full %>% 
  unite("CI", ci.lo:ci.up, sep=",") %>%
  mutate(CI = paste0("(",CI,")")) %>%
  `colnames<-`(c("Phenotype","Groups", "t", "df", "p", "Mean Difference", "95% CI")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Results of unpaired Student's t testing to assess differences in overall JSD between PostTX and allo in rejectors and controls", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))

```

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=12, fig.height=6, fig.cap="Jensen-Shannon divergence (JSD) of VJ combinations between pre-transplant and post-transplant repertoire samples grouped by phenotype"}
JSD.BLlate <- JSD.VJ[JSD.VJ$stage %in% "BL1" & JSD.VJ$stage.1 %in% "late1",]

tmp <- melt(JSD.BLlate, id.vars = colnames(JSD.BLlate)[1:10])

p.vj <- ggplot(data=tmp, aes(y=value, x=variable,label=patID))+
  geom_point(aes(x=variable, label=patID),stat="identity", position = position_dodge()) +
  geom_line(aes(x=variable, group=patID), alpha=1) +
  scale_x_discrete( "",label=c("All", "Top 1000", "Top 500", "Top 100")) +
  scale_y_continuous("Jensen-Shannon Divergence") +
  theme_bw() +
  scale_fill_manual(values=c("firebrick2", "deepskyblue4")) +
  scale_color_manual(values=my_cols) +
  facet_wrap(~CDtype,scales="free_x") +
  theme(plot.title = element_text(hjust = 0.5),legend.position="none", text = element_text(size=16))


p.vj
ggsave(filename = paste(out.path,"plot_JSD_VJ_BLvsLate_tcopies.png", sep = ""), plot=p.vj, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)

```

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=12, fig.height=6, fig.cap="Jensen-Shannon divergence (JSD) of VJ combinations between pre-transplant and donor-reactive repertoire samples grouped by phenotype"}

JSD.BLlow <- JSD.VJ[JSD.VJ$stage %in% "BL1" & JSD.VJ$stage.1 %in% "low1",]

tmp <- melt(JSD.BLlow, id.vars = colnames(JSD.BLlow)[1:10])

p.vj <- ggplot(data=tmp, aes(y=value, x=variable,label=patID))+
  geom_point(aes(x=variable, label=patID),stat="identity", position = position_dodge()) +
  geom_line(aes(x=variable, group=patID), alpha=1) +
  scale_x_discrete( "",label=c("All", "Top 1000", "Top 500", "Top 100")) +
  scale_y_continuous("Jensen-Shannon Divergence") +
  theme_bw() +
  scale_fill_manual(values=c("firebrick2", "deepskyblue4")) +
  scale_color_manual(values=my_cols) +
  facet_wrap(~CDtype,scales="free_x") +
  theme(plot.title = element_text(hjust = 0.5),legend.position="none", text = element_text(size=16))


p.vj
ggsave(filename = paste(out.path,"plot_JSD_VJ_BLvsMLR_tcopies.png", sep = ""), plot=p.vj, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)

```

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=12, fig.height=6, fig.cap="Jensen-Shannon divergence (JSD) of VJ combinations between donor-reactive and post-transplant repertoire samples grouped by phenotype"}

JSD.MLRLate <- JSD.VJ[JSD.VJ$stage %in% "low1" & JSD.VJ$stage.1 %in% "late1",]

tmp <- melt(JSD.MLRLate, id.vars = colnames(JSD.MLRLate)[1:10])

p.vj <- ggplot(data=tmp, aes(y=value, x=variable,label=patID))+
  geom_point(aes(x=variable, label=patID),stat="identity", position = position_dodge()) +
  geom_line(aes(x=variable, group=patID), alpha=1) +
  scale_x_discrete( "",label=c("All", "Top 1000", "Top 500", "Top 100")) +
  scale_y_continuous("Jensen-Shannon Divergence") +
  theme_bw() +
  scale_fill_manual(values=c("firebrick2", "deepskyblue4")) +
  scale_color_manual(values=my_cols) +
  facet_wrap(~CDtype,scales="free_x") +
  theme(plot.title = element_text(hjust = 0.5),legend.position="none", text = element_text(size=16))


p.vj
ggsave(filename = paste(out.path,"plot_JSD_VJ_MLRvsLate_tcopies.png", sep = ""), plot=p.vj, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)

```

## Donor-reactive clonotypes grouped in BL and LATE (downsampling)
Normalization is performed by down-sampling the datasets to the size of the smallest reads for the phenotype groups CD4 and CD8, respectively. The
illustrated estimate represents the mean estimate over 1000 repeats.

```{r echo=FALSE, cache=T, cache=F,message=FALSE, warning=FALSE, fig.align="center", fig.width=12, fig.height=8, fig.cap="Absolute number of detected donor-reactive clonotypes after normalization by downsamling to the smallest number of reads"}
downsampling <- function(df, data.dwns, incl.stages=NULL, dwns.size){

  # Downsampling
  df.down <- sample(df$cdr3nt, dwns.size, prob = df$freq, replace=T)
  df1 <- reshape2::melt(table(df.down))
  df1$freq <- df1$value/sum(df1$value)
  colnames(df1) <- c("cdr3nt", "count", "freq")

  df.down <- merge(df1,df[,!(colnames(df)%in% c("count","freq"))],by="cdr3nt")

  abs.allo <- sum(df.down$allo==1)
  abs.allo.clones <- sum(df.down[df.down$allo==1,]$count)
  rel.allo <- sum(df.down$allo==1)/nrow(df.down)*100
  rel.allo.clones <-  sum(df.down[df.down$allo==1,]$count)/sum(df.down$count)*100
  cum.allo <- sum(df.down[df.down$allo==1,]$freq)

  return(c(abs.allo, abs.allo.clones, rel.allo, rel.allo.clones, cum.allo))
}

wrapper_downsampling <- function(filename, path, df.dwns, rep.dwns, incl.stages=NULL, dwns.size=NULL){

  file.info <- extract_nameinfo(filename)
  df <- read.table(paste0(path, filename), header=T, stringsAsFactors = F)
  if(is.null(dwns.size)){dwns.size <- min(df.dwns[df.dwns$CDtype%in%file.info["CDtype"] & df.dwns$stage %in% incl.stages,]$smallest.reads)}

  # Original values
  org.abs.allo <- sum(df$allo==1)
  org.abs.allo.clones <- sum(df[df$allo==1,]$count)
  org.rel.allo <- sum(df$allo==1)/nrow(df)*100
  org.rel.allo.clones <- sum(df[df$allo==1,]$count)/sum(df$count)*100
  org.cum.allo <- sum(df[df$allo==1,]$freq)

  # Downsampling
  res <- t(sapply(1:rep.dwns, function(x) downsampling(df, data.dwns=df.dwns, incl.stages, dwns.size)))
  out <- c(apply(res, 2, function(x) c(mean(x, na.rm=T), quantile(x,c(0.25,0.75),na.rm=T))))

  return(c(file.info[2:5], org.abs.allo, org.abs.allo.clones, org.rel.allo, org.rel.allo.clones ,org.cum.allo, out))

}

extract_ttest <- function(df1, df2, pairedt=TRUE){
  res <- t.test(df1,df2, paired = pairedt)
  dt <- c(summary(df1), sd(df1), summary(df2), sd(df1))
  
  return(list("test"=c(res$statistic,res$parameter, res$p.value, res$estimate[1], res$conf.int[1], res$conf.int[2]),
       "data"=dt))
}

#Get reads for all samples
tcopies.reads <- sapply(read_AllSamples(data.path), function (x) sum(x[c('count')]))
tcopies.reads <- data.frame(cbind(t(sapply(names(tcopies.reads), extract_nameinfo)), reads=tcopies.reads))
tcopies.reads$reads <- as.numeric(as.character(tcopies.reads$reads))
rownames(tcopies.reads) <- NULL

# Get smallest reads for the categories CD4/CD8 and BL/MLR/late and store it in df.dws
tcopies.reads$group <- paste0(tcopies.reads$CDtype,"+",tcopies.reads$stage)
tcopies.split <- split(tcopies.reads, as.factor(tcopies.reads$group))
df.dwns <- do.call("rbind", lapply(tcopies.split,function(x) min(x$reads)))
df.dwns <- data.frame(cbind(unique(tcopies.reads[,c("CDtype","stage")]), smallest.reads=df.dwns))

 df.dwns %>%
   flextable()  %>%
     bold(part = "header") %>%
   set_caption("Smallest read size for each phenotype group and time point", autonum = run_autonum(seq_id = "tab", bkm = newBKM())) %>%
   autofit() %>%
   fit_to_width(6.5)

if(!file.exists("table_tcopies_allo_dwns.xlsx")){
 rel.files <- files.tcr[str_detect(files.tcr,"low", negate = T)]
 plan(multisession, workers=5)
 res.dwns <- data.frame(t(future_sapply(rel.files, function(x) wrapper_downsampling(x, path=data.path, df.dwns, rep.dwns = rep.dwns, incl.stages = c("BL1","late1")))))
 plan(sequential)
 colnames(res.dwns) <- c("patID","CDtype","stage","rejection",
                         "org.abs.allo",  "org.abs.allo.clones", "org.rel.allo", "org.rel.allo.clones", "org.cum.allo",
                        "abs.allo", "abs.allo.lo", "abs.allo.up",
                        "abs.allo.clones", "abs.allo.clones.lo", "abs.allo.clones.up",
                        "rel.allo", "rel.allo.lo", "rel.allo.up",
                        "rel.allo.clones", "rel.allo.clones.lo", "rel.allo.clones.up",
                        "cum.allo","cum.allo.lo", "cum.allo.up")
 
 res.dwns[,5:24] <- apply(res.dwns[,5:24],2, as.numeric)
 res.dwns[res.dwns$rejection==0,]$rejection <- "Controls"
 res.dwns[res.dwns$rejection==1,]$rejection <- "Rejectors"
 write.xlsx(res.dwns, paste0(out.path, "table_tcopies_allo_dwns.xlsx"), overwrite=T)
}

res.dwns <- read.xlsx(paste0(out.path, "table_tcopies_allo_dwns.xlsx"))

# ------------------------------------------ CD4 --------------------------------------------------------------------------------------------
tmp <- res.dwns[res.dwns$stage %in% c("BL1", "late1") & res.dwns$CDtype %in% "CD4",]
p1b.CD4 <- ggplot(data=tmp, aes(y=patID, x=abs.allo, xmin=abs.allo.lo, xmax=abs.allo.up, fill=stage, col=stage))+
  geom_point(aes(fill=stage)) +
  geom_errorbarh(height=.1)+
  geom_vline(xintercept=0, color="black", linetype="dashed", alpha=.5)+
  scale_color_brewer(palette="Dark2", name = "", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_y_discrete("") +
  scale_x_continuous("Absolute number of detected donor-reactive clonotypes", limits = c(0,1500), breaks = seq(0,1500,250),guide = guide_axis(angle = 45)) +
  theme_bw() +
  theme(text = element_text(size=18)) +
  facet_grid(rejection~., scales= "free", space="free")

my_cols <-brewer.pal(8, "Paired")[c(1,2)]
p1a.CD4 <- ggplot(data=tmp, aes(y=abs.allo, x=patID, fill=stage))+
  geom_bar(stat="identity", position = position_dodge()) +
  scale_fill_manual(values=my_cols, name = "", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_x_discrete("",guide = guide_axis(angle = 45)) +
  scale_y_continuous("Absolute number of detected donor-reactive clonotypes") +
  theme_bw() +
  ggtitle("CD4") +
  facet_wrap(~rejection,scales="free_x") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=16))

p2a.CD4 <- ggplot(data=tmp, aes(y=rel.allo, x=patID, fill=stage))+
  geom_bar(stat="identity", position = position_dodge()) +
  scale_fill_manual(values=my_cols, name = "", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_x_discrete("",guide = guide_axis(angle = 45)) +
  scale_y_continuous("Percentage of detected donor-reactive clonotypes", limits = c(0, 5.6), breaks = c(0,1,3,5)) +
  theme_bw() +
  ggtitle("CD4") +
  facet_wrap(~rejection,scales="free_x") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=16))

p3a.CD4 <- ggplot(data=tmp, aes(y=cum.allo, x=patID, fill=stage))+
  geom_bar(stat="identity", position = position_dodge()) +
  scale_fill_manual(values=my_cols, name = "", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_x_discrete("",guide = guide_axis(angle = 45)) +
  scale_y_continuous("", breaks = c(0,0.01,0.03,0.05)) +
  theme_bw() +
  ggtitle("CD4") +
  facet_wrap(~rejection,scales="free_x") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=16))

# ------Hypothesis testing
# Clonotypes
tt.rel.cd4.clts <- extract_ttest(tmp[tmp$stage%in% "BL1",]$rel.allo, tmp[tmp$stage%in% "late1",]$rel.allo)
tt.abs.cd4.clts <- extract_ttest(tmp[tmp$stage%in% "BL1",]$abs.allo, tmp[tmp$stage%in% "late1",]$abs.allo)

res.tt.cd4.clts <- data.frame(rbind(tt.rel.cd4.clts[[1]],tt.abs.cd4.clts[[1]]))
colnames(res.tt.cd4.clts) <- c("t", "df", "p-value", "dd.value", "ci.lo", "ci.up")

res.data.cd4.clts <- data.frame(rbind(tt.rel.cd4.clts[[2]],tt.abs.cd4.clts[[2]]))
colnames(res.data.cd4.clts) <- c("Min", "1stQu", "Median", "Mean", "3rdQu", "Max", "SD")

# Clones
tt.rel.cd4.cs <- extract_ttest(tmp[tmp$stage%in% "BL1",]$rel.allo.clones, tmp[tmp$stage%in% "late1",]$rel.allo.clones)
tt.abs.cd4.cs <- extract_ttest(tmp[tmp$stage%in% "BL1",]$abs.allo.clones, tmp[tmp$stage%in% "late1",]$abs.allo.clones)

res.tt.cd4.cs <- data.frame(rbind(tt.rel.cd4.cs[[1]],tt.abs.cd4.cs[[1]]))
colnames(res.tt.cd4.cs) <- c("t", "df", "p-value", "dd.value", "ci.lo", "ci.up")

res.data.cd4.cs <- data.frame(rbind(tt.rel.cd4.cs[[2]],tt.abs.cd4.cs[[2]]))
colnames(res.data.cd4.cs) <- c("Min", "1stQu", "Median", "Mean", "3rdQu", "Max", "SD")

# -------------------------------------------  CD8  ---------------------------------------------------------------
tmp <- res.dwns[res.dwns$stage %in% c("BL1", "late1") & res.dwns$CDtype %in% "CD8",]
p1b.CD8 <- ggplot(data=tmp, aes(y=patID, x=abs.allo, xmin=abs.allo.lo, xmax=abs.allo.up, fill=stage))+
  geom_point(aes(fill=stage)) +
  geom_errorbarh(height=.1)+
  geom_vline(xintercept=0, color="black", linetype="dashed", alpha=.5)+
  scale_color_brewer(palette="Dark2", name = "", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_y_discrete("") +
  scale_x_continuous("Absolute number of detected donor-reactive clonotypes", limits = c(0,1500), breaks = seq(0,1500,250),guide = guide_axis(angle = 45)) +
  theme_bw() +
  facet_grid(rejection~., scales= "free", space="free")

p1a.CD8 <- ggplot(data=tmp, aes(y=abs.allo, x=patID, fill=stage))+
  geom_bar(stat="identity", position = position_dodge()) +
  scale_fill_manual(values=my_cols, name = "", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_x_discrete("",guide = guide_axis(angle = 45)) +
  scale_y_continuous("") +
  theme_bw() +
  ggtitle("CD8") +
  facet_wrap(~rejection,scales="free_x") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=16))

p2a.CD8 <- ggplot(data=tmp, aes(y=rel.allo, x=patID, fill=stage))+
  geom_bar(stat="identity", position = position_dodge()) +
  scale_fill_manual(values=my_cols, name = "", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_x_discrete("",guide = guide_axis(angle = 45)) +
  scale_y_continuous("", limits = c(0, 5.6), breaks = c(0,1,3,5)) +
  theme_bw() +
  ggtitle("CD8") +
  facet_wrap(~rejection,scales="free_x") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=16))

p3a.CD8 <- ggplot(data=tmp, aes(y=cum.allo, x=patID, fill=stage))+
  geom_bar(stat="identity", position = position_dodge()) +
  scale_fill_manual(values=my_cols, name = "", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_x_discrete("",guide = guide_axis(angle = 45)) +
  scale_y_continuous("", breaks = c(0,0.01,0.03,0.05)) +
  theme_bw() +
  ggtitle("CD8") +
  facet_wrap(~rejection,scales="free_x") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=16))


filename <- paste0("plot_absolute_allo_tcopies")
p1a <- ggarrange(p1a.CD4, p1a.CD8, ncol=2, nrow=1, common.legend = TRUE, legend="bottom")
p1a
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p1a, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)


# ------Hypothesis testing
# Clonotypes
tt.rel.cd8.clts <- extract_ttest(tmp[tmp$stage%in% "BL1",]$rel.allo, tmp[tmp$stage%in% "late1",]$rel.allo)
tt.abs.cd8.clts <- extract_ttest(tmp[tmp$stage%in% "BL1",]$abs.allo, tmp[tmp$stage%in% "late1",]$abs.allo)

res.tt.cd8.clts <- data.frame(rbind(tt.rel.cd8.clts[[1]],tt.abs.cd8.clts[[1]]))
colnames(res.tt.cd8.clts) <- c("t", "df", "p-value", "dd.value", "ci.lo", "ci.up")

res.data.cd8.clts <- data.frame(rbind(tt.rel.cd8.clts[[2]],tt.abs.cd8.clts[[2]]))
colnames(res.data.cd8.clts) <- c("Min", "1stQu", "Median", "Mean", "3rdQu", "Max", "SD")

# Clones
tt.rel.cd8.cs <- extract_ttest(tmp[tmp$stage%in% "BL1",]$rel.allo.clones, tmp[tmp$stage%in% "late1",]$rel.allo.clones)
tt.abs.cd8.cs <- extract_ttest(tmp[tmp$stage%in% "BL1",]$abs.allo.clones, tmp[tmp$stage%in% "late1",]$abs.allo.clones)

res.tt.cd8.cs <- data.frame(rbind(tt.rel.cd8.cs[[1]],tt.abs.cd8.cs[[1]]))
colnames(res.tt.cd8.cs) <- c("t", "df", "p-value", "dd.value", "ci.lo", "ci.up")

res.data.cd8.cs <- data.frame(rbind(tt.rel.cd8.cs[[2]],tt.abs.cd8.cs[[2]]))
colnames(res.data.cd8.cs) <- c("Min", "1stQu", "Median", "Mean", "3rdQu", "Max", "SD")


```

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=12, fig.height=8, fig.cap="Percentage of detetcted donor-reactive clonotypes after normalization by downsamling to the smallest number of reads"}

filename <- paste0("plot_percentage_allo_tcopies")
p2a <- ggarrange(p2a.CD4, p2a.CD8, ncol=2, nrow=1, common.legend = TRUE, legend="bottom")
p2a
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p2a, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)


```

\textbf{Paired Student's t test for alloclonal differences in PreTX vs PostTX on (1) the clonotype and (2) the clone level}

```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE}
# Test difference on clonotype level
res.data.clts <- data.frame("Phenotype"=c(rep("CD4",2),rep("CD8",2)), 
                           "Variable"=rep(c("Alloclonotype fraction", "Absolute alloclonotype counts"),2),
                           round(rbind(res.data.cd4.clts, res.data.cd8.clts),2))

res.data.clts %>%
  `colnames<-`(c("Phenotype","Variable", 
                 paste(rep(c("BL","Late"), each = length(c("Min", "Q1", "Median", "Mean", "Q3","Max", "SD"))), 
                       c("Min", "Q1", "Median", "Mean", "Q3","Max","SD"), sep = "."))) %>%
  flextable()  %>%
  add_header_row(top = TRUE, 
                     values = c("","PreTX", "PostTX"), 
                     colwidths = c(2,7,7)) %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Overview of alloclonotype fraction and absolute alloclonotype counts in the pre-transplant and post-transplant repertoires after downsampling", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))


res.ttpaired.clts <- data.frame("Phenotype"=c(rep("CD4",2),rep("CD8",2)), 
                           "Variable"=rep(c("Alloclonotype fraction","Absolute alloclonotype counts"),2),
                           rbind(res.tt.cd4.clts, res.tt.cd8.clts))

res.ttpaired.clts %>%
  mutate(t = sprintf("%0.1f", t)) %>%
  mutate(p.value = sprintf("%0.3f", p.value)) %>%
  mutate(dd.value = sprintf("%0.3f", dd.value)) %>%
  mutate(ci.lo = sprintf("%0.3f", ci.lo)) %>%
  mutate(ci.up = sprintf("%0.3f", ci.up)) %>%
  unite("CI", ci.lo:ci.up, sep=",") %>%
  mutate(CI = paste0("(",CI,")")) %>%
  `colnames<-`(c("Phenotype","Variable", "t", "df", "p", "Mean Difference", "95% CI")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Results of the paired t test assessing differences in absolute alloconotype counts and alloclonotype fraction in the pre-and post-transplant repertoire", autonum = run_autonum(seq_id = "tab", bkm = newBKM())) 



# Test difference on clone level
res.data.cs <- data.frame("Phenotype"=c(rep("CD4",2),rep("CD8",2)), 
                           "Variable"=rep(c("Alloclone fraction", "Absolute allo counts"),2),
                           round(rbind(res.data.cd4.cs, res.data.cd8.cs),2))

res.data.cs %>%
  `colnames<-`(c("Phenotype","Variable", 
                 paste(rep(c("BL","Late"), each = length(c("Min", "Q1", "Median", "Mean", "Q3","Max", "SD"))), 
                       c("Min", "Q1", "Median", "Mean", "Q3","Max","SD"), sep = "."))) %>%
  flextable()  %>%
  add_header_row(top = TRUE, 
                     values = c("","PreTX", "PostTX"), 
                     colwidths = c(2,7,7)) %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Overview of alloclonal fraction and absolute alloclone counts in the pre-transplant and post-transplant repertoires after downsampling", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))


res.ttpaired.cs <- data.frame("Phenotype"=c(rep("CD4",2),rep("CD8",2)), 
                           "Variable"=rep(c("Alloclone fraction","Absolute allo counts"),2),
                           rbind(res.tt.cd4.cs, res.tt.cd8.cs))

res.ttpaired.cs %>%
  mutate(t = sprintf("%0.1f", t)) %>%
  mutate(p.value = sprintf("%0.3f", p.value)) %>%
  mutate(dd.value = sprintf("%0.3f", dd.value)) %>%
  mutate(ci.lo = sprintf("%0.3f", ci.lo)) %>%
  mutate(ci.up = sprintf("%0.3f", ci.up)) %>%
  unite("CI", ci.lo:ci.up, sep=",") %>%
  mutate(CI = paste0("(",CI,")")) %>%
  `colnames<-`(c("Phenotype","Variable", "t", "df", "p", "Mean Difference", "95% CI")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Results of the paired t test assessing differences in absolute allocone counts and alloclonal fraction in the pre-and post-transplant repertoire", autonum = run_autonum(seq_id = "tab", bkm = newBKM())) 

```


\textbf{Unpaired Student's t test of (1) alloclonotype fraction and (2) alloclonal fraction between the two outcome groups in PostTX}
```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE}
# Hypothesis testing of alloclonotype fraction between the two outcome groups in PostTX
res.tt.groups.cd4 <- t.test(res.dwns[res.dwns$CDtype %in% "CD4" & res.dwns$stage %in% "late1" & res.dwns$rejection %in% "Controls",]$rel.allo,
                                   res.dwns[res.dwns$CDtype %in% "CD4" & res.dwns$stage %in% "late1" & res.dwns$rejection %in% "Rejectors",]$rel.allo, paired = F)
res.tt.groups.cd8 <- t.test(res.dwns[res.dwns$CDtype %in% "CD8" & res.dwns$stage %in% "late1" & res.dwns$rejection %in% "Controls",]$rel.allo,
                                   res.dwns[res.dwns$CDtype %in% "CD8" & res.dwns$stage %in% "late1" & res.dwns$rejection %in% "Rejectors",]$rel.allo, paired = F)
res.tmp <- data.frame("Pheno"=c("CD4", "CD8"), 
                      "Variable"=rep("Alloclonotype fraction",2),
                      "t"=c(res.tt.groups.cd4$statistic,res.tt.groups.cd8$statistic), 
                          "df"=c(res.tt.groups.cd4$parameter, res.tt.groups.cd8$parameter), 
                          "p"=c(res.tt.groups.cd4$p.value, res.tt.groups.cd8$p.value),
                          "mean.diff"=c(res.tt.groups.cd4$estimate[1]-res.tt.groups.cd4$estimate[2], res.tt.groups.cd8$estimate[1]-res.tt.groups.cd8$estimate[2]),
                          "ci.lo"=c(res.tt.groups.cd4$conf.int[1], res.tt.groups.cd8$conf.int[1]),
                          "ci.up"=c(res.tt.groups.cd4$conf.int[2], res.tt.groups.cd8$conf.int[2]))
res.tmp[,3:8] <- apply(res.tmp[,3:8],2, function(x) round(x,3))

res.tmp %>%
    unite("CI", ci.lo:ci.up, sep=",") %>%
  mutate(CI = paste0("(",CI,")")) %>%
  `colnames<-`(c("Phenotype","Variable", "t", "df", "p", "Mean Difference", "95% CI")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Results of the unpaired Student's t test assessing differences in alloclonotype fraction between rejectors (N=6) and non-rejectors (N=6)", autonum = run_autonum(seq_id = "tab", bkm = newBKM())) 


# Hypothesis testing of alloclonal fraction between the two outcome groups in PostTX
res.tt.groups.cd4 <- t.test(res.dwns[res.dwns$CDtype %in% "CD4" & res.dwns$stage %in% "late1" & res.dwns$rejection %in% "Controls",]$rel.allo.clones,
                                   res.dwns[res.dwns$CDtype %in% "CD4" & res.dwns$stage %in% "late1" & res.dwns$rejection %in% "Rejectors",]$rel.allo.clones, paired = F)
res.tt.groups.cd8 <- t.test(res.dwns[res.dwns$CDtype %in% "CD8" & res.dwns$stage %in% "late1" & res.dwns$rejection %in% "Controls",]$rel.allo.clones,
                                   res.dwns[res.dwns$CDtype %in% "CD8" & res.dwns$stage %in% "late1" & res.dwns$rejection %in% "Rejectors",]$rel.allo.clones, paired = F)
res.tmp <- data.frame("Pheno"=c("CD4", "CD8"), 
                      "Variable"=rep("Alloclone fraction",2),
                      "t"=c(res.tt.groups.cd4$statistic,res.tt.groups.cd8$statistic), 
                          "df"=c(res.tt.groups.cd4$parameter, res.tt.groups.cd8$parameter), 
                          "p"=c(res.tt.groups.cd4$p.value, res.tt.groups.cd8$p.value),
                          "mean.diff"=c(res.tt.groups.cd4$estimate[1]-res.tt.groups.cd4$estimate[2], res.tt.groups.cd8$estimate[1]-res.tt.groups.cd8$estimate[2]),
                          "ci.lo"=c(res.tt.groups.cd4$conf.int[1], res.tt.groups.cd8$conf.int[1]),
                          "ci.up"=c(res.tt.groups.cd4$conf.int[2], res.tt.groups.cd8$conf.int[2]))
res.tmp[,3:8] <- apply(res.tmp[,3:8],2, function(x) round(x,3))

res.tmp %>%
    unite("CI", ci.lo:ci.up, sep=",") %>%
  mutate(CI = paste0("(",CI,")")) %>%
  `colnames<-`(c("Phenotype","Variable", "t", "df", "p", "Mean Difference", "95% CI")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Results of the unpaired Student's t test assessing differences in alloclone fraction between rejectors (N=6) and non-rejectors (N=6)", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))
```


## Diversity estimates
T cell diversity of the repertoires was assessed by the R20 in the post-transplant samples and in the donor-reactive samples. The R20 measure quantifies the extent to which the top 20% of the most frequent clonotypes are expanded in contrast to the bottom 80% by computing the ratio between the number of unique clonotypes within the top 20% compared to the number in the bottom 80% of all detected clonotypes. 


```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE}
calc_RepertoireStatistics <- function(filename=NULL, data=NULL, path=data.path){
  # Generate vector of the following repertoire characteristics:
  # *list characteristics*
  if(is.null(data)){
  sample <- read.table(paste0(path,filename), header=T, stringsAsFactors = F)
  fileinfo <- extract_nameinfo(filename)
  }else{
    fileinfo <- NA
    sample<-data
  }
  
  reads <- sum(sample$count)
  largest.clone.freq <- max(sample$freq)
  largest.clone.count <- max(sample$count)
  sr.nuc <- length(unique(sample$cdr3nt))
  sr.aa <- length(unique(sample$cdr3aa))
  
  # RX diversity
  R10 <- min(which(cumsum(sample$freq)>=0.10))/nrow(sample)
  R20 <- min(which(cumsum(sample$freq)>=0.20))/nrow(sample)
  R50 <- min(which(cumsum(sample$freq)>=0.50))/nrow(sample)

  sample.allo <- sample[sample$allo==1,]
  sample.allo$freq.allo <- sample.allo$count/sum(sample.allo$count)
  R10allo <- min(which(cumsum(sample.allo$freq.allo)>=0.10))/nrow(sample.allo)
  R20allo <- min(which(cumsum(sample.allo$freq.allo)>=0.20))/nrow(sample.allo)
  R50allo <- min(which(cumsum(sample.allo$freq.allo)>=0.50))/nrow(sample.allo)
  
  Hobs <- -sum(sample$freq*log(sample$freq,2)) #Shannon
  Hmax <- log(nrow(sample),2)
  pielou <- Hobs/Hmax
  clonality <- 1- pielou
  si <- sum(sample$freq^2)
  invsi <- 1/si
 # ginic <- Gini(sample$count)
  bpi <- max(sample$freq)
  invbpi <- 1/bpi

  # The clonotype abundance frequency distribution  (x=frequency, y=number of clones with frequency x)
  tbl.counts <- table(sample$count)
  threshold.counts <- as.numeric(names(tbl.counts[tbl.counts == 1][2]))
  sample.bulk <- sample[sample$count < threshold.counts,]
  clone_nr <- c(table(sample.bulk$freq))
  freq <- as.numeric(as.character(names(clone_nr)))
  names(clone_nr) <- NULL
  if(all(is.na(log(freq))) | all(is.na(log(clone_nr)))){
    slope<-NA
  }else{
    slope <- as.numeric(abs(lm(log(clone_nr)~log(freq))$coefficients[2]))
  }

  # Non-parametric estimators of species richness
  df <- cbind(ceiling(as.numeric(names(table(sample$count)))),table(sample$count))
  chao1 <- chao1984(df, conf = 0.95)
  ACE_t10 <- ChaoLee1992(df, t = 10, method = "ACE",conf = 0.95)
  ACEbc_t10 <- ChaoLee1992(df, t = 10, method = "ACE-1",conf = 0.95)

  repstats <- c(c(fileinfo), round(c(Reads=reads, largest.clone.freq=largest.clone.freq, largest.clone.count=largest.clone.count,
                                     SpeciesRichness.Nuc=sr.nuc, SpeciesRichness.AA=sr.aa,
                                     R10=R10, R10allo=R10allo, R20=R20,R20allo=R20allo, R50=R50, R50allo=R50allo,
                                     ShannonE=Hobs, maxH=Hmax,PielouDiversity = pielou, Clonality=clonality,
                                     SimpsonIndex=si, invSimpsonIndex=invsi,
                                     BergerParkerIndex=bpi, invBergerParkerIndex=invbpi, pl.slope=slope,
                                     Chao1.est=chao1$Nhat, Chao1.se=chao1$SE,
                                     ACEbc.t10.est=ACEbc_t10$Nhat, ACEbc.t10.se=ACEbc_t10$SE),5))
}


#Generate table of repertoire characteristics for all rep samples
if(!file.exists("table_tcopies_diversity.xlsx")){
  tcopies.stats <- data.frame(t(sapply(files.tcr, calc_RepertoireStatistics, USE.NAMES = F)))
  varsToNumeric <- colnames(tcopies.stats)[5:ncol(tcopies.stats)]
  tcopies.stats[,varsToNumeric] <- lapply(tcopies.stats[,varsToNumeric], function(x) as.numeric(as.character(x)))
  tcopies.stats[tcopies.stats$rejection==0,]$rejection <- "Controls"
  tcopies.stats[tcopies.stats$rejection==1,]$rejection <- "Rejectors"
  write.xlsx(tcopies.stats, paste0(out.path, "table_tcopies_diversity.xlsx"), overwrite=T)
}

tcopies.stats <- read.xlsx(paste0(out.path, "table_tcopies_diversity.xlsx"))

#------- CD4
tmp4 <- tcopies.stats[tcopies.stats$CDtype%in%"CD4" & tcopies.stats$stage%in%c("BL1","late1"),]
p1 <-  ggplot(tmp4, aes(x=stage, y=R20, fill=rejection)) +
  geom_boxplot(position=position_dodge(0.8), notch=F, width = .4, alpha=0.8) +
  geom_dotplot(binaxis = "y", stackdir = "center",  position=position_dodge(0.8), dotsize=0.45, alpha=1) +
  stat_boxplot(geom ='errorbar', width = 0.2, position=position_dodge(0.8), alpha=0.8) +
  scale_y_continuous(expand=c(0,0),"R20", limits=c(0,0.075)) +
  scale_x_discrete(expand=c(0,0),"", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_fill_brewer(palette="Dark2",name = "", labels = c("Controls", "Rejectors")) +
  ggtitle("CD4") +
  theme_bw() +
  theme(legend.position="none", plot.title = element_text(hjust = 0.5), text = element_text(size=14))

p3 <- ggplot(tmp4, aes(x=stage, y=R20allo, fill=rejection)) +
  geom_boxplot(position=position_dodge(0.8), notch=F, width = .4, alpha=0.8) +
  geom_dotplot(binaxis = "y", stackdir = "center",  position=position_dodge(0.8), dotsize=0.45, alpha=1) +
  stat_boxplot(geom ='errorbar', width = 0.2, position=position_dodge(0.8), alpha=0.8) +
  scale_y_continuous(expand=c(0,0),"R20 donor-reactive",  limits=c(0,0.065)) +
  scale_x_discrete(expand=c(0,0),"", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_fill_brewer(palette="Dark2",name = "", labels = c("Controls", "Rejectors")) +
  ggtitle("CD4") +
  theme_bw() +
  theme(legend.position="none", plot.title = element_text(hjust = 0.5), text = element_text(size=14))

p5 <- ggplot(tmp4, aes(x=stage, y=Clonality, fill=rejection)) +
  geom_boxplot(position=position_dodge(0.8), notch=F, width = .4, alpha=0.8) +
  geom_dotplot(binaxis = "y", stackdir = "center",  position=position_dodge(0.8), dotsize=0.45, alpha=1) +
  stat_boxplot(geom ='errorbar', width = 0.2, position=position_dodge(0.8), alpha=0.8) +
  scale_y_continuous(expand=c(0,0),"Clonality", limits = c(0,0.55)) +
  scale_x_discrete(expand=c(0,0),"", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_fill_brewer(palette="Dark2",name = "", labels = c("Controls", "Rejectors")) +
  ggtitle("CD4") +
  theme_bw() +
  theme(legend.position="none", plot.title = element_text(hjust = 0.5), text = element_text(size=14))

p7 <- ggplot(tmp4, aes(x=stage, y=Chao1.est, fill=rejection)) +
  geom_boxplot(position=position_dodge(0.8), notch=F, width = .4, alpha=0.8) +
  geom_dotplot(binaxis = "y", stackdir = "center",  position=position_dodge(0.8), dotsize=0.45, alpha=1) +
  stat_boxplot(geom ='errorbar', width = 0.2, position=position_dodge(0.8), alpha=0.8) +
  scale_y_continuous(expand=c(0,0),"Chao1 Estimator") +
  scale_x_discrete(expand=c(0,0),"", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_fill_brewer(palette="Dark2",name = "", labels = c("Controls", "Rejectors")) +
  ggtitle("CD4") +
  theme_bw() +
  theme(legend.position="none", plot.title = element_text(hjust = 0.5), text = element_text(size=14))

p9 <- ggplot(tmp4, aes(x=stage, y=ACEbc.t10.est, fill=rejection)) +
  geom_boxplot(position=position_dodge(0.8), notch=F, width = .4, alpha=0.8) +
  geom_dotplot(binaxis = "y", stackdir = "center",  position=position_dodge(0.8), dotsize=0.45, alpha=1) +
  stat_boxplot(geom ='errorbar', width = 0.2, position=position_dodge(0.8), alpha=0.8) +
  scale_y_continuous(expand=c(0,0),"Abundance-based Coverage Estimator (ACE)") +
  scale_x_discrete(expand=c(0,0),"", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_fill_brewer(palette="Dark2",name = "", labels = c("Controls", "Rejectors")) +
  ggtitle("CD4") +
  theme_bw() +
  theme(legend.position="none", plot.title = element_text(hjust = 0.5), text = element_text(size=14))

# ---- CD8
tmp8 <- tcopies.stats[tcopies.stats$CDtype%in%"CD8" & tcopies.stats$stage%in%c("BL1","late1"),]
p2 <-  ggplot(tmp8, aes(x=stage, y=R20, fill=rejection)) +
  geom_boxplot(position=position_dodge(0.8), notch=F, width = .4, alpha=0.8) +
  geom_dotplot(binaxis = "y", stackdir = "center",  position=position_dodge(0.8), dotsize=0.45, alpha=1) +
  stat_boxplot(geom ='errorbar', width = 0.2, position=position_dodge(0.8), alpha=0.8) +
  scale_y_continuous(expand=c(0,0),"", limits=c(0,0.0075)) +
  scale_x_discrete(expand=c(0,0),"", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_fill_brewer(palette="Dark2",name = "", labels = c("Controls", "Rejectors")) +
  ggtitle("CD8") +
  theme_bw() +
  theme(legend.position="none", plot.title = element_text(hjust = 0.5), text = element_text(size=14))

p4 <- ggplot(tmp8, aes(x=stage, y=R20allo, fill=rejection)) +
  geom_boxplot(position=position_dodge(0.8), notch=F, width = .4, alpha=0.8) +
  geom_dotplot(binaxis = "y", stackdir = "center",  position=position_dodge(0.8), dotsize=0.45, alpha=1) +
  stat_boxplot(geom ='errorbar', width = 0.2, position=position_dodge(0.8), alpha=0.8) +
  scale_y_continuous(expand=c(0,0),"", limits=c(0,0.065)) +
  scale_x_discrete(expand=c(0,0),"", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_fill_brewer(palette="Dark2",name = "", labels = c("Controls", "Rejectors")) +
  ggtitle("CD8") +
  theme_bw() +
  theme(legend.position="none", plot.title = element_text(hjust = 0.5), text = element_text(size=14))

p6 <- ggplot(tmp8, aes(x=stage, y=Clonality, fill=rejection)) +
  geom_boxplot(position=position_dodge(0.8), notch=F, width = .4, alpha=0.8) +
  geom_dotplot(binaxis = "y", stackdir = "center",  position=position_dodge(0.8), dotsize=0.45, alpha=1) +
  stat_boxplot(geom ='errorbar', width = 0.2, position=position_dodge(0.8), alpha=0.8) +
  scale_y_continuous(expand=c(0,0),"", limits = c(0,0.55)) +
  scale_x_discrete(expand=c(0,0),"", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_fill_brewer(palette="Dark2",name = "", labels = c("Controls", "Rejectors")) +
  ggtitle("CD8") +
  theme_bw() +
  theme(legend.position="none", plot.title = element_text(hjust = 0.5), text = element_text(size=14))

p8 <- ggplot(tmp8, aes(x=stage, y=Chao1.est, fill=rejection)) +
  geom_boxplot(position=position_dodge(0.8), notch=F, width = .4, alpha=0.8) +
  geom_dotplot(binaxis = "y", stackdir = "center",  position=position_dodge(0.8), dotsize=0.45, alpha=1) +
  stat_boxplot(geom ='errorbar', width = 0.2, position=position_dodge(0.8), alpha=0.8) +
  scale_y_continuous(expand=c(0,0),"Chao1 Estimator") +
  scale_x_discrete(expand=c(0,0),"", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_fill_brewer(palette="Dark2",name = "", labels = c("Controls", "Rejectors")) +
  ggtitle("CD8") +
  theme_bw() +
  theme(legend.position="none", plot.title = element_text(hjust = 0.5), text = element_text(size=14))

p10 <- ggplot(tmp8, aes(x=stage, y=ACEbc.t10.est, fill=rejection)) +
  geom_boxplot(position=position_dodge(0.8), notch=F, width = .4, alpha=0.8) +
  geom_dotplot(binaxis = "y", stackdir = "center",  position=position_dodge(0.8), dotsize=0.45, alpha=1) +
  stat_boxplot(geom ='errorbar', width = 0.2, position=position_dodge(0.8), alpha=0.8) +
  scale_y_continuous(expand=c(0,0),"Abundance-based Coverage Estimator (ACE)") +
  scale_x_discrete(expand=c(0,0),"", labels = c("Pre-Transplant", "Post-Transplant")) +
  scale_fill_brewer(palette="Dark2",name = "", labels = c("Controls", "Rejectors")) +
  ggtitle("CD8") +
  theme_bw() +
  theme(legend.position="none", plot.title = element_text(hjust = 0.5), text = element_text(size=14))
```

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap="Boxplot of R20 estimates in the pre- and post-transplant repertoire samples grouped by rejection"}

filename <- paste0("plot_R20_tcopies")
p12 <- ggarrange(p1, p2, ncol=2, nrow=1, common.legend = TRUE, legend="bottom")
p12
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p12, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)


# Paired between BL and Late for CD4 and CD8
# test.cd4 <- t.test(tmp4[tmp4$stage %in% "BL1",]$R20, tmp4[tmp4$stage %in% "late1",]$R20, paired=TRUE) 
# test.cd8 <- t.test(tmp8[tmp8$stage %in% "BL1",]$R20, tmp8[tmp8$stage %in% "late1",]$R20, paired=TRUE) 
# 
# # Unpaired between rejectors and controls for CD4 and CD8
# test.rej.cd4.BL <-t.test(tmp4[tmp4$rejection %in% "Controls" & tmp4$stage %in% "BL1",]$R20, tmp4[tmp4$rejection %in% "Rejectors" & tmp4$stage %in% "BL1",]$R20) 
# wilc.rej.cd4.late <- t.test(tmp4[tmp4$rejection %in% "Controls" & tmp4$stage %in% "late1",]$R20, tmp4[tmp4$rejection %in% "Rejectors" & tmp4$stage %in% "late1",]$R20) 
# 
# test.rej.cd8.BL <- t.test(tmp8[tmp8$rejection %in% "Controls" & tmp8$stage %in% "BL1",]$R20, tmp8[tmp8$rejection %in% "Rejectors" & tmp8$stage %in% "BL1",]$R20) 
# test.rej.cd8.late <- t.test(tmp8[tmp8$rejection %in% "Controls" & tmp8$stage %in% "late1",]$R20, tmp8[tmp8$rejection %in% "Rejectors" & tmp8$stage %in% "late1",]$R20) 


```


```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap="Boxplot of R20 estimates of the donor-reactive clonotypes in the pre- and post-transplant repertoire samples grouped by rejection"}
filename <- paste0("plot_R20allo_tcopies")
p34 <- ggarrange(p3, p4, ncol=2, nrow=1, common.legend = TRUE, legend="bottom")
p34
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p34, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)

```

Further, clonality was computed for the post-transplant repertoire samples. Clonality provides a measure of normalized diversity for a given repertoire. It is derived from Shannon's entropy which can be vary due to the variation in detected cells per experiment. Observed changes in Shannon's entropy can either be a result of sample size variation or differences in the frequency distribution. Thus, normalization of the of the entropy can decouple these effects. Clonality yields values within the range of [0,1], where 0 indicates the most diverse repertoire and 1 the least diverse with only one clonotype present.


```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap="Boxplot of clonality estimates of the pre- and post-transplant repertoire samples grouped by rejection"}
filename <- paste0("plot_clonality_tcopies")
p56 <- ggarrange(p5, p6, ncol=2, nrow=1, common.legend = TRUE, legend="bottom")
p56
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p56, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)

```

The unpaired t test was performed to investigate whether there exist a statistically significant difference in $\Delta R20=R20_{BX}-R20_{BL}$, $\Delta R20^{allo}=R20^{allo}_{BX}-R20^{allo}_{BL}$ and $\Delta C=C_{BX}-C_{BL}$ (C...Clonality) 
between the controls and the rejectors.

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE}

tmp4.new <- merge(tmp4[tmp4$stage %in% "BL1",c("patID", "stage", "rejection", "R20", "R20allo", "Clonality")], tmp4[tmp4$stage %in% "late1",c("patID", "stage", "rejection", "R20", "R20allo", "Clonality")], by="patID")
tmp4.new$deltaR20 <- tmp4.new$R20.y-tmp4.new$R20.x
tmp4.new$deltaR20allo <- tmp4.new$R20allo.y-tmp4.new$R20allo.x
tmp4.new$deltaClonality <- tmp4.new$Clonality.y-tmp4.new$Clonality.x

tmp8.new <- merge(tmp8[tmp8$stage %in% "BL1",c("patID", "stage", "rejection", "R20", "R20allo", "Clonality")], tmp8[tmp8$stage %in% "late1",c("patID", "stage", "rejection", "R20", "R20allo", "Clonality")], by="patID")
tmp8.new$deltaR20 <- tmp8.new$R20.y-tmp8.new$R20.x
tmp8.new$deltaR20allo <- tmp8.new$R20allo.y-tmp8.new$R20allo.x
tmp8.new$deltaClonality <- tmp8.new$Clonality.y-tmp8.new$Clonality.x

perform_ttests<- function(x, y){
  test <- t.test(x,y)
  
  res <-c("t"=test$statistic, test$parameter, "p"=test$p.value, "dd.value"=as.numeric(test$estimate[1]- test$estimate[2]),  "ci.lo"=test$conf.int[1], "ci.up"=test$conf.int[2])
  return(res)
}

res.ttests <- matrix(NA, nrow=6, ncol=6)
res.ttests[1,] <- perform_ttests(x=tmp4.new[tmp4.new$rejection.x %in% "Controls",]$deltaR20,
                                 y=tmp4.new[tmp4.new$rejection.x %in% "Rejectors",]$deltaR20)
res.ttests[2,] <- perform_ttests(x=tmp4.new[tmp4.new$rejection.x %in% "Controls",]$deltaR20allo,
                                 y=tmp4.new[tmp4.new$rejection.x %in% "Rejectors",]$deltaR20allo)
res.ttests[3,] <- perform_ttests(x=tmp4.new[tmp4.new$rejection.x %in% "Controls",]$deltaClonality, 
                                 y=tmp4.new[tmp4.new$rejection.x %in% "Rejectors",]$deltaClonality)

res.ttests[4,] <- perform_ttests(x=tmp8.new[tmp8.new$rejection.x %in% "Controls",]$deltaR20, 
                                 y=tmp8.new[tmp8.new$rejection.x %in% "Rejectors",]$deltaR20)
res.ttests[5,] <- perform_ttests(x=tmp8.new[tmp8.new$rejection.x %in% "Controls",]$deltaR20allo, 
                                 y=tmp8.new[tmp8.new$rejection.x %in% "Rejectors",]$deltaR20allo)
res.ttests[6,] <- perform_ttests(x=tmp8.new[tmp8.new$rejection.x %in% "Controls",]$deltaClonality, 
                                 y=tmp8.new[tmp8.new$rejection.x %in% "Rejectors",]$deltaClonality)
colnames(res.ttests) <- c("t", "df", "p", "dd.value", "ci.lo", "ci.up")
res.ttests <- round(res.ttests, 3)

res <- data.frame("Phenotype"=c(rep("CD4",3), rep("CD8",3)), "Variable"=rep(c("deltaR20", "deltaR20allo", "deltaClonality"),2),res.ttests)

res %>%
  unite("95% CI", ci.lo:ci.up, sep="-") %>%
  `colnames<-`(c("Phenotype","Variable", "t", "df", "p", "Mean Difference", "95% CI")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Results of the two-sample t tests for the diversity estimates between rejectors and controls for CD4 and CD8", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))
  
#- Range of diversity estimates

div.df <- data.frame("Pheno"=c(rep("CD4",4),rep("CD8",4)), "Stage"=rep(c("PreTX","PostTX"),4), 
                     "Variable"=rep(c("R20","R20","Clonality", "Clonality"),2),
                     rbind(summary(tmp4[tmp4$stage %in% "BL1",]$R20),summary(tmp4[tmp4$stage %in% "late1",]$R20), 
      summary(tmp4[tmp4$stage %in% "BL1",]$Clonality), summary(tmp4[tmp4$stage %in% "late1",]$Clonality),
      summary(tmp8[tmp8$stage %in% "BL1",]$R20),summary(tmp8[tmp8$stage %in% "late1",]$R20), 
      summary(tmp8[tmp8$stage %in% "BL1",]$Clonality), summary(tmp8[tmp8$stage %in% "late1",]$Clonality)))
div.df[,4:9] <- apply(div.df[,4:9],2,function(x) round(x,3))

div.df %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Distribution of the diversity estimates across the immune repertoires", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))


```

An existing problem with these naive estimators (R20, species richness) of diversity is their sensitivity to the sequencing depth. A greater depth of sequencing leads to the capture of more and more newly detected clonotypes. More robust estimators are given by the Chao1 estimator and the Abundance-based coverage estimator, which add a correction factor to the number of observed species to account for undetected/unseen clonotypes.


```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap="Boxplot of Chao1 estimates of the pre- and post-transplant repertoire samples grouped by rejection"}

filename <- paste0("plot_chao1_tcopies")
p78 <- ggarrange(p7, p8, ncol=2, nrow=1, common.legend = TRUE, legend="bottom")
p78
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p78, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)

```


```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap="Boxplot of the abundance-based coverage estimator (ACE) of the pre- and post-transplant repertoire samples grouped by rejection"}
filename <- paste0("plot_ace_tcopies")
p910 <- ggarrange(p9, p10, ncol=2, nrow=1, common.legend = TRUE, legend="bottom")
p910
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p910, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)

```



```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE}

tmp4.new <- merge(tmp4[tmp4$stage %in% "BL1",c("patID", "stage", "rejection", "Chao1.est", "ACEbc.t10.est")], tmp4[tmp4$stage %in% "late1",c("patID", "stage", "rejection", "Chao1.est", "ACEbc.t10.est")], by="patID")
tmp4.new$deltaChao1.est <- tmp4.new$Chao1.est.y-tmp4.new$Chao1.est.x
tmp4.new$deltaACEbc.t10.est <- tmp4.new$ACEbc.t10.est.y-tmp4.new$ACEbc.t10.est.x

tmp8.new <- merge(tmp8[tmp8$stage %in% "BL1",c("patID", "stage", "rejection", "Chao1.est", "ACEbc.t10.est")], tmp8[tmp8$stage %in% "late1",c("patID", "stage", "rejection", "Chao1.est", "ACEbc.t10.est")], by="patID")
tmp8.new$deltaChao1.est <- tmp8.new$Chao1.est.y-tmp8.new$Chao1.est.x
tmp8.new$deltaACEbc.t10.est <- tmp8.new$ACEbc.t10.est.y-tmp8.new$ACEbc.t10.est.x

perform_ttests<- function(x, y){
  test <- t.test(x,y)
  
  res <-c("t"=test$statistic, test$parameter, "p"=test$p.value, "dd.value"=as.numeric(test$estimate[1]- test$estimate[2]),  "ci.lo"=test$conf.int[1], "ci.up"=test$conf.int[2])
  return(res)
}

res.ttests <- matrix(NA, nrow=4, ncol=6)
res.ttests[1,] <- perform_ttests(x=tmp4.new[tmp4.new$rejection.x %in% "Controls",]$deltaChao1.est, y=tmp4.new[tmp8.new$rejection.x %in% "Rejectors",]$deltaChao1.est)
res.ttests[2,] <- perform_ttests(x=tmp4.new[tmp4.new$rejection.x %in% "Controls",]$deltaACEbc.t10.est, y=tmp4.new[tmp8.new$rejection.x %in% "Rejectors",]$deltaACEbc.t10.est)

res.ttests[3,] <- perform_ttests(x=tmp8.new[tmp8.new$rejection.x %in% "Controls",]$deltaChao1.est, y=tmp8.new[tmp8.new$rejection.x %in% "Rejectors",]$deltaChao1.est)
res.ttests[4,] <- perform_ttests(x=tmp8.new[tmp8.new$rejection.x %in% "Controls",]$deltaACEbc.t10.est, y=tmp8.new[tmp8.new$rejection.x %in% "Rejectors",]$deltaACEbc.t10.est)
colnames(res.ttests) <- c("t", "df", "p", "dd.value", "ci.lo", "ci.up")
res.ttests <- round(res.ttests, 3)

res <- data.frame("Phenotype"=c(rep("CD4",2), rep("CD8",2)), "Variable"=rep(c("deltaChao1", "deltaACE"),2),res.ttests)

res %>%
  unite("95% CI", ci.lo:ci.up, sep="-") %>%
  `colnames<-`(c("Phenotype","Variable", "t", "df", "p", "Mean Difference", "95% CI")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Results of the two-sample t tests for the diversity estimates between rejectors and controls for CD4 and CD8", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))
  

```

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=8, fig.height=6}


# compute_row <- function(i, v, threshold){
#   res = stringdist(v[i], v[-(1:i)], method = "lv")
#   res_threshold = which(res %in% c(1:threshold))
#   res_threshold
#
#   if(length(res_threshold)>0){
#     df <- data.frame(i = v[i], j = v[i+res_threshold], leven = res[res_threshold])
#   }else{
#     df <- data.frame(i = NULL, j = NULL, leven = NULL)
#   }
#   return(df)
# }
#
# compute_levenshtein <- function(v, threshold, workers = 5) {
#   n = length(v)
#
#   if (workers > 1) {
#     plan(multisession, workers = workers)
#     on.exit(plan(sequential))
#   }
#   res <- future_sapply(1:n, function(i) compute_row(i, v, threshold))
#
#   return(res)
# }
#
# normalized_attr <- function(network){
#   degree_dist <- igraph::degree(network)
#   simple_random_net <- sample_degseq(degree_dist)
#
#
# }
#
# generate_network <- function(df, tissue=F){
#   res <- compute_levenshtein(df$cdr3nt, threshold = 2)
#   adjlist <- as.data.table(do.call(rbind, res))
#   colnames(adjlist) <- c("Node1", "Node2", "leven")
#   net.df <- graph.edgelist(as.matrix(adjlist[leven == 1, c(1,2)]))
#
#   net.dens.full <- edge_density(net.df)
#   net.trans.full <- transitivity(net.df, type="globalundirected")
#   net.cpl.full <- mean_distance(net.df, directed = F, unconnected = FALSE)
#
#   if(!tissue){
#     norm_attr <- normalized_attr(net.df)
#   }
#
#   net.df <- delete.vertices(net.df, names(which(igraph::degree(net.df) == 0)))
#   net.dens.simp <- edge_density(net.df)
#   net.trans.simp <- transitivity(net.df, type="globalundirected")
#   net.cpl.simp <- mean_distance(net.df, directed = F, unconnected = FALSE)
#
#   E(net.df)$color <- 'black'
#   l <- layout_with_graphopt(net.df)
#   p <- plot(net.df, vertex.label=NA, remove.multiple = T, edge.arrow.size=0.1, vertex.size=2, layout=l, main="R34")
#
#   net.attr <- c(net.dens.full, net.trans.full, net.cpl.full, net.dens.simp, net.trans.simp, net.cpl.simp)
#
#   return(list("plot"=p, "attr"=net.attr))
# }

# res.network <- list()
# res.network[[1]] <- generate_network(tissue_34[tissue_34$allo==1,])
# res.network[[2]] <- generate_network(df.34.BL[df.34.BL$allo==1,])
# res.network[[3]] <- generate_network(df.34.late[df.34.late$allo==1,])
# res.network[[4]] <- generate_network(tissue_43[tissue_43$allo==1,])
# res.network[[5]] <- generate_network(df.43.BL[df.43.late$allo==1,])
# res.network[[6]] <- generate_network(df.43.late[df.43.late$allo==1,])

```

# Tissue analysis of the samples R34 and R43


## Readjust CD4:CD8 ratio in R34 and R43 

Frequencies of the pre-and post-transplant sample were adjusted according to the CD4:CD8 ratios:

- R34: CD4= 71.9% and CD8 = 28.1%
- R43: CD4 = 61.9% and CD8 = 38.1%
- R30: CD4 = 81% and CD8 = 19%
- R42: CD4 = 61% and CD8 = 39%
- P148: CD4 = 73% and CD8 = 27%
- P190: CD4 = 50% and CD8 = 50%

Further, all repertoire in Figure XX were downsampled to the size of the smallest reads. In this case, the smallest reads were in the tissue sample of patient R34 with 6186.

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=8, fig.height=6, fig.cap="Barchart of the relative fraction of detected donor-reactive clonotypes after normalization by downsampling to the smallest number of reads"}
downsample_data <- function(df.sample , dwns.size=NULL){

  # Downsampling
  df.down <- sample(df.sample$cdr3nt, dwns.size, prob = df.sample$freq, replace=T)
  df1 <- reshape2::melt(table(df.down))
  df1$freq <- df1$value/sum(df1$value)
  colnames(df1) <- c("cdr3nt", "count", "freq")

  df.down <- merge(df1,df.sample[,!(colnames(df.sample)%in% c("count","freq"))],by="cdr3nt")

  abs.allo <- sum(df.down$allo==1)
  rel.allo <- sum(df.down$allo==1)/nrow(df.down)
  cum.freq.allo <- sum(df.down[df.down$allo==1,]$freq)

  return(c(abs.allo, rel.allo,cum.freq.allo))
}


# R34
df.34.CD4.BL <- read.table(paste0(data.path,"vdj.R34_CD4_BL_1_clones.txt"), header=T, stringsAsFactors = F)
df.34.CD4.BL$freq= df.34.CD4.BL$freq * 0.719
df.34.CD4.BL$count= df.34.CD4.BL$count
df.34.CD8.BL <- read.table(paste0(data.path,"vdj.R34_CD8_BL_1_clones.txt"), header=T, stringsAsFactors = F)
df.34.CD8.BL$freq= df.34.CD8.BL$freq * 0.281
adj.factor.34.BL <-  ((sum(df.34.CD4.BL$count)/0.719)*0.281)/sum(df.34.CD8.BL$count)
df.34.CD8.BL$count = df.34.CD8.BL$count * adj.factor.34.BL

df.34.CD4.low <- read.table(paste0(data.path,"vdj.R34_CD4_low_1_clones.txt"), header=T, stringsAsFactors = F)
df.34.CD4.low$freq= df.34.CD4.low$freq * 0.719
df.34.CD4.low$count= df.34.CD4.low$count 
df.34.CD8.low <- read.table(paste0(data.path,"vdj.R34_CD8_low_1_clones.txt"), header=T, stringsAsFactors = F)
df.34.CD8.low$freq= df.34.CD8.low$freq * 0.281
adj.factor.34.low <- ((sum(df.34.CD4.low$count)/0.719)*0.281)/sum(df.34.CD8.low$count)
df.34.CD8.low$count= df.34.CD8.low$count * adj.factor.34.low


df.34.CD4.late <- read.table(paste0(data.path,"vdj.R34_CD4_late_1_clones.txt"), header=T, stringsAsFactors = F)
df.34.CD4.late$freq= df.34.CD4.late$freq * 0.719
df.34.CD4.late$count= df.34.CD4.late$count * 0.719
df.34.CD8.late <- read.table(paste0(data.path,"vdj.R34_CD8_late_1_clones.txt"), header=T, stringsAsFactors = F)
df.34.CD8.late$freq= df.34.CD8.late$freq * 0.281
adj.factor.34.late <-  ((sum(df.34.CD4.late$count)/0.719)*0.281)/sum(df.34.CD8.late$count)
df.34.CD8.late$count= df.34.CD8.late$count * adj.factor.34.late


df.34.BL <- rbind(df.34.CD4.BL, df.34.CD8.BL)
df.34.low <- rbind(df.34.CD4.low, df.34.CD8.low)
df.34.late <- rbind(df.34.CD4.late, df.34.CD8.late)

df.34.CD4 <- rbind(df.34.CD4.BL,df.34.CD4.low,df.34.CD4.late[,!(colnames(df.34.CD4.late) %in% "newBX")])
df.34.CD8 <- rbind(df.34.CD8.BL,df.34.CD8.low,df.34.CD8.late[,!(colnames(df.34.CD8.late) %in% "newBX")])


# R43
df.43.CD4.BL <- read.table(paste0(data.path,"vdj.R43_CD4_BL_1_clones.txt"), header=T, stringsAsFactors = F)
df.43.CD4.BL$freq= df.43.CD4.BL$freq * 0.619
df.43.CD4.BL$count= df.43.CD4.BL$count * 0.619
df.43.CD8.BL <- read.table(paste0(data.path,"vdj.R43_CD8_BL_1_clones.txt"), header=T, stringsAsFactors = F)
df.43.CD8.BL$freq= df.43.CD8.BL$freq * 0.381
adj.factor.43.BL <-  ((sum(df.43.CD4.BL$count)/0.619)*0.381)/sum(df.43.CD8.BL$count)
df.43.CD8.BL$count= df.43.CD8.BL$count * adj.factor.43.BL

df.43.CD4.low <- read.table(paste0(data.path,"vdj.R43_CD4_low_1_clones.txt"), header=T, stringsAsFactors = F)
df.43.CD4.low$freq= df.43.CD4.low$freq * 0.619
df.43.CD4.low$count= df.43.CD4.low$count * 0.619
df.43.CD8.low <- read.table(paste0(data.path,"vdj.R43_CD8_low_1_clones.txt"), header=T, stringsAsFactors = F)
df.43.CD8.low$freq= df.43.CD8.low$freq * 0.381
adj.factor.43.low <-  ((sum(df.43.CD4.low$count)/0.619)*0.381)/sum(df.43.CD8.low$count)
df.43.CD8.low$count= df.43.CD8.low$count * adj.factor.43.low

df.43.CD4.late <- read.table(paste0(data.path,"vdj.R43_CD4_late_1_clones.txt"), header=T, stringsAsFactors = F)
df.43.CD4.late$freq= df.43.CD4.late$freq * 0.619
df.43.CD4.late$count= df.43.CD4.late$count * 0.619
df.43.CD8.late <- read.table(paste0(data.path,"vdj.R43_CD8_late_1_clones.txt"), header=T, stringsAsFactors = F)
df.43.CD8.late$freq= df.43.CD8.late$freq * 0.381
adj.factor.43.late <-  ((sum(df.43.CD4.late$count)/0.619)*0.381)/sum(df.43.CD8.late$count)
df.43.CD8.late$count= df.43.CD8.late$count * adj.factor.43.late

df.43.BL <- rbind(df.43.CD4.BL, df.43.CD8.BL)
df.43.low <- rbind(df.43.CD4.low, df.43.CD8.low)
df.43.late <- rbind(df.43.CD4.late, df.43.CD8.late)

df.43.CD4 <- rbind(df.43.CD4.BL,df.43.CD4.low,df.43.CD4.late[,!(colnames(df.43.CD4.late) %in% "newBX")])
df.43.CD8 <- rbind(df.43.CD8.BL,df.43.CD8.low,df.43.CD8.late[,!(colnames(df.43.CD8.late) %in% "newBX")])

# # R30
# df.30.CD4.BL <- read.table(paste0(data.path,"vdj.R30_CD4_BL_1_clones.txt"), header=T, stringsAsFactors = F)
# df.30.CD4.BL$freq= df.30.CD4.BL$freq * 0.81
# df.30.CD8.BL <- read.table(paste0(data.path,"vdj.R30_CD8_BL_1_clones.txt"), header=T, stringsAsFactors = F)
# df.30.CD8.BL$freq= df.30.CD8.BL$freq * 0.19
# 
# df.30.CD4.late <- read.table(paste0(data.path,"vdj.R30_CD4_late_1_clones.txt"), header=T, stringsAsFactors = F)
# df.30.CD4.late$freq= df.30.CD4.late$freq * 0.81
# df.30.CD8.late <- read.table(paste0(data.path,"vdj.R30_CD8_late_1_clones.txt"), header=T, stringsAsFactors = F)
# df.30.CD8.late$freq= df.30.CD8.late$freq * 0.19
# 
# df.30.BL <- rbind(df.30.CD4.BL, df.30.CD8.BL)
# df.30.late <- rbind(df.30.CD4.late, df.30.CD8.late)
# 
# # R43
# df.42.CD4.BL <- read.table(paste0(data.path,"vdj.R42_CD4_BL_1_clones.txt"), header=T, stringsAsFactors = F)
# df.42.CD4.BL$freq= df.42.CD4.BL$freq * 0.61
# df.42.CD8.BL <- read.table(paste0(data.path,"vdj.R42_CD8_BL_1_clones.txt"), header=T, stringsAsFactors = F)
# df.42.CD8.BL$freq= df.42.CD8.BL$freq * 0.39
# 
# df.42.CD4.late <- read.table(paste0(data.path,"vdj.R42_CD4_late_1_clones.txt"), header=T, stringsAsFactors = F)
# df.42.CD4.late$freq= df.42.CD4.late$freq * 0.61
# df.42.CD8.late <- read.table(paste0(data.path,"vdj.R42_CD8_late_1_clones.txt"), header=T, stringsAsFactors = F)
# df.42.CD8.late$freq= df.42.CD8.late$freq * 0.39
# 
# df.42.BL <- rbind(df.42.CD4.BL, df.42.CD8.BL)
# df.42.late <- rbind(df.42.CD4.late, df.42.CD8.late)
# 
# # R148
# df.148.CD4.BL <- read.table(paste0(data.path,"vdj.R148_CD4_BL_1_clones.txt"), header=T, stringsAsFactors = F)
# df.148.CD4.BL$freq= df.148.CD4.BL$freq * 0.73
# df.148.CD8.BL <- read.table(paste0(data.path,"vdj.R148_CD8_BL_1_clones.txt"), header=T, stringsAsFactors = F)
# df.148.CD8.BL$freq= df.148.CD8.BL$freq * 0.27
# 
# df.148.CD4.late <- read.table(paste0(data.path,"vdj.R148_CD4_late_1_clones.txt"), header=T, stringsAsFactors = F)
# df.148.CD4.late$freq= df.148.CD4.late$freq * 0.73
# df.148.CD8.late <- read.table(paste0(data.path,"vdj.R148_CD8_late_1_clones.txt"), header=T, stringsAsFactors = F)
# df.148.CD8.late$freq= df.148.CD8.late$freq * 0.27
# 
# df.148.BL <- rbind(df.148.CD4.BL, df.148.CD8.BL)
# df.148.late <- rbind(df.148.CD4.late, df.148.CD8.late)
# 
# # R190
# df.190.CD4.BL <- read.table(paste0(data.path,"vdj.R190_CD4_BL_1_clones.txt"), header=T, stringsAsFactors = F)
# df.190.CD4.BL$freq= df.190.CD4.BL$freq * 0.5
# df.190.CD8.BL <- read.table(paste0(data.path,"vdj.R190_CD8_BL_1_clones.txt"), header=T, stringsAsFactors = F)
# df.190.CD8.BL$freq= df.190.CD8.BL$freq * 0.5
# 
# df.190.CD4.late <- read.table(paste0(data.path,"vdj.R190_CD4_late_1_clones.txt"), header=T, stringsAsFactors = F)
# df.190.CD4.late$freq= df.190.CD4.late$freq * 0.5
# df.190.CD8.late <- read.table(paste0(data.path,"vdj.R190_CD8_late_1_clones.txt"), header=T, stringsAsFactors = F)
# df.190.CD8.late$freq= df.190.CD8.late$freq * 0.5
# 
# df.190.BL <- rbind(df.190.CD4.BL, df.190.CD8.BL)
# df.190.late <- rbind(df.190.CD4.late, df.190.CD8.late)
# 
# # # # Downsampling
# res.34.BL <- c(apply(sapply(1:rep.dwns, function(x) downsample_data(df.34.BL, dwns.size = sum(tissue_34$count))),1, function(y) c(mean(y), median(y), quantile(y, c(0.25,0.75)))))
# res.34.late <- c(apply(sapply(1:rep.dwns, function(x) downsample_data(df.34.late, dwns.size = sum(tissue_34$count))),1, function(y) c(mean(y), median(y),quantile(y, c(0.25,0.75)))))
# 
# res.43.BL <- c(apply(sapply(1:rep.dwns, function(x) downsample_data(df.43.BL, dwns.size = sum(tissue_34$count))),1, function(y) c(mean(y), median(y), quantile(y, c(0.25,0.75)))))
# res.43.late <- c(apply(sapply(1:rep.dwns, function(x) downsample_data(df.43.late, dwns.size = sum(tissue_34$count))),1, function(y) c(mean(y), median(y),quantile(y, c(0.25,0.75)))))
# 
# res.30.BL <- c(apply(sapply(1:rep.dwns, function(x) downsample_data(df.30.BL, dwns.size = sum(tissue_34$count))),1, function(y) c(mean(y), median(y), quantile(y, c(0.25,0.75)))))
# res.30.late <- c(apply(sapply(1:rep.dwns, function(x) downsample_data(df.30.late, dwns.size = sum(tissue_34$count))),1, function(y) c(mean(y), median(y),quantile(y, c(0.25,0.75)))))
# 
# res.42.BL <- c(apply(sapply(1:rep.dwns, function(x) downsample_data(df.42.BL, dwns.size = sum(tissue_34$count))),1, function(y) c(mean(y), median(y), quantile(y, c(0.25,0.75)))))
# res.42.late <- c(apply(sapply(1:rep.dwns, function(x) downsample_data(df.42.late, dwns.size = sum(tissue_34$count))),1, function(y) c(mean(y), median(y),quantile(y, c(0.25,0.75)))))
# 
# res.148.BL <- c(apply(sapply(1:rep.dwns, function(x) downsample_data(df.148.BL, dwns.size = sum(tissue_34$count))),1, function(y) c(mean(y), median(y), quantile(y, c(0.25,0.75)))))
# res.148.late <- c(apply(sapply(1:rep.dwns, function(x) downsample_data(df.148.late, dwns.size = sum(tissue_34$count))),1, function(y) c(mean(y), median(y), quantile(y, c(0.25,0.75)))))
# 
# res.190.BL <- c(apply(sapply(1:rep.dwns, function(x) downsample_data(df.190.BL, dwns.size = sum(tissue_34$count))),1, function(y) c(mean(y), median(y), quantile(y, c(0.25,0.75)))))
# res.190.late <- c(apply(sapply(1:rep.dwns, function(x) downsample_data(df.190.late, dwns.size = sum(tissue_34$count))),1, function(y) c(mean(y), median(y), quantile(y, c(0.25,0.75)))))
# 
# 
# 
# res.all <- data.frame(rbind(res.34.BL, res.34.late, res.43.BL, res.43.late, res.42.BL, res.42.late, res.30.BL, res.30.late, res.148.BL, res.148.late, res.190.BL, res.190.late))
# res.all$stage <- rep(c("BL1", "late1"),6)
# res.all$rejection <- c(rep(1,6),rep(0,6))
# colnames(res.all) <- c("abs.allo", "abs.allo.med","abs.allo.lo","abs.allo.up",
#                               "rel.allo","rel.allo.med","rel.allo.lo","rel.allo.up",
#                               "cum.allo","cum.allo.med","cum.allo.lo","cum.allo.up", "stage","rejection")
# tmp.tiss <- res.tmp.tissue[,c("abs.allo", "abs.allo.lo","abs.allo.up",
#                               "rel.allo","rel.allo.lo","rel.allo.up",
#                               "cum.allo","cum.allo.lo","cum.allo.up", "stage","rejection")]
# tmp.tiss[,c("rel.allo")] <- as.numeric(tmp.tiss[,c("rel.allo")])/100
# tmp.tiss[,c("rel.allo.lo")] <- as.numeric(tmp.tiss[,c("rel.allo.lo")])/100
# tmp.tiss[,c("rel.allo.up")] <- as.numeric(tmp.tiss[,c("rel.allo.up")])/100
# 
# res.all <- rbind(tmp.tiss,res.all[,c("abs.allo", "abs.allo.lo","abs.allo.up",
#                               "rel.allo","rel.allo.lo","rel.allo.up",
#                               "cum.allo","cum.allo.lo","cum.allo.up", "stage","rejection")])
# res.all$patID <- c("R34", "R43", "R34", "R34", "R43", "R43","R42", "R42","R30","R30","R148","R148", "R190","R190")
# 
# res.all <- res.all %>%
#   relocate(patID, .before=abs.allo) %>%
#   relocate(stage, .before=abs.allo) %>%
#   relocate(rejection, .before=abs.allo) %>%
#   rbind(c("R30","tissue1",rep(0,10))) %>%
#   rbind(c("R42","tissue1",1,rep(0,9))) %>%
#   rbind(c("R148","tissue1",rep(0,10))) %>%
#   rbind(c("R190","tissue1",rep(0,10)))
# 
# res.all[,4:12] <- apply(res.all[,4:12],2, as.numeric)
# 
# write.xlsx(res.all, paste0(out.path, "table_tcopies_dwns_PrePostTissue_readjust.xlsx"), overwrite=T)

if(F){ #deactivate for now depends on file >>table_tcopies_dwns_PrePostTissue_readjust.xlsx<< which currently can not be generated as variable res.tmp.tissue is missing
res.all <- read.xlsx(paste0(out.path, "table_tcopies_dwns_PrePostTissue_readjust.xlsx"))
res.all[res.all$rejection==0,]$rejection <- "Controls"
res.all[res.all$rejection==1,]$rejection <- "Rejectors"
res.all[res.all == 0]<-0.00011

my_cols <- c("#A6CEE3", "#1F78B4", "red4")
p <- ggplot(data=res.all, aes(y=rel.allo, x=patID, fill=stage))+
  geom_col(position = position_dodge2(preserve = "single", padding = 0.1))+
  scale_x_discrete( "") +
  scale_y_continuous("Alloclone fraction", limits=c(0,0.05), breaks = seq(0,0.05,0.01)) +
  theme_bw() +
  scale_fill_manual(values=my_cols,name = "", labels = c("Pre-Transplant", "Circulation at Biopsy", "Allograft at Biopsy")) +
  ggtitle("") +
  theme(plot.title = element_text(hjust = 0.5),legend.position="bottom", text = element_text(size=14)) +
  facet_wrap(~rejection, scales="free_x")

filename <- paste0("plot_allofrac_blood+tissue_barchart_3x3_tcopies")
p
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p, width = 20, height = 20, dpi = 640, units="cm", limitsize = F)

my_cols <- c("#A6CEE3", "#1F78B4", "red4")
tmp <- res.all[res.all$patID %in% c("R30", "R190", "R34", "R43"),]
p <- ggplot(data=tmp, aes(y=rel.allo, x=patID, fill=stage))+
  geom_col(position = position_dodge2(preserve = "single", padding = 0.1))+
  scale_x_discrete( "") +
  scale_y_continuous("Alloclone fraction", limits=c(0,0.05), breaks = seq(0,0.05,0.01)) +
  theme_bw() +
  scale_fill_manual(values=my_cols,name = "", labels = c("Pre-Transplant", "Circulation at Biopsy", "Allograft at Biopsy")) +
  ggtitle("") +
  theme(plot.title = element_text(hjust = 0.5),legend.position="bottom", text = element_text(size=14)) +
  facet_wrap(~rejection, scales="free_x")

filename <- paste0("plot_allofrac_blood+tissue_barchart_2x2_tcopies")
p
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p, width = 20, height = 20, dpi = 640, units="cm", limitsize = F)
}
```





```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE}
#---------- CD4:CD8 ratio in the tissue
tissue_34$pheno <- "Unknown"
tissue_34[tissue_34$cdr3nt %in% df.34.CD4$cdr3nt,]$pheno <- "CD4"
tissue_34[tissue_34$cdr3nt %in% df.34.CD8$cdr3nt,]$pheno <- "CD8"

tissue_43$pheno <- "Unknown"
tissue_43[tissue_43$cdr3nt %in% df.43.CD4$cdr3nt,]$pheno <- "CD4"
tissue_43[tissue_43$cdr3nt %in% df.43.CD8$cdr3nt,]$pheno <- "CD8"

cdratio.tissue <- data.frame(cbind("Patient"=c("R34","R43"),rbind(table(tissue_34$pheno), table(tissue_43$pheno))))

cdratio.tissue %>%
  mutate_at('CD4', as.numeric) %>%
  mutate_at('CD8', as.numeric) %>%
  mutate_at('Unknown', as.numeric) %>%
  mutate(CD4.perc=round(CD4/(CD4+CD8+Unknown)*100,2),
         CD8.perc=round(CD8/(CD4+CD8+Unknown)*100,2),
         Unknown.perc=round(Unknown/(CD4+CD8+Unknown)*100,2)) %>%
  relocate(CD4.perc, .after = CD4) %>% 
  relocate(CD8.perc, .after = CD8) %>% 
  flextable() %>%
  autofit %>%
  bold(part = "header") %>%
  set_caption("CD4:CD8 ratio of clones in the allograft", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))

cdratio.tissue.counts <- data.frame(cbind(c("R34", "R43"),rbind(c(sum(tissue_34$count), sum(tissue_34[tissue_34$pheno %in% "CD4",]$count), sum(tissue_34[tissue_34$pheno %in% "CD8",]$count), sum(tissue_34[tissue_34$pheno %in% "Unknown",]$count)), c(sum(tissue_43$count), sum(tissue_43[tissue_43$pheno %in% "CD4",]$count), sum(tissue_43[tissue_43$pheno %in% "CD8",]$count), sum(tissue_43[tissue_43$pheno %in% "Unknown",]$count)))))
colnames(cdratio.tissue.counts) <- c("Patient","Total", "CD4", "CD8", "Unknown")

cdratio.tissue.counts %>%
  mutate_at('CD4', as.numeric) %>%
  mutate_at('CD8', as.numeric) %>%
  mutate_at('Unknown', as.numeric) %>%
  mutate(CD4.perc=round(CD4/(CD4+CD8+Unknown)*100,2),
         CD8.perc=round(CD8/(CD4+CD8+Unknown)*100,2),
         Unknown.perc=round(Unknown/(CD4+CD8+Unknown)*100,2)) %>%
  relocate(CD4.perc, .after = CD4) %>% 
  relocate(CD8.perc, .after = CD8) %>% 
  flextable() %>%
  autofit %>%
  bold(part = "header") %>%
  set_caption("CD4:CD8 ratio of clone counts in the allograft", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))
```
## Diversity estimates

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE}
stages <- c("overall", "BL", "low", "late")

calc_CloneOverlap <- function(file, blood, tissue, stage="overall"){
  if(stage!="overall"){
    blood_tmp <- blood[,c(1,grep(stage, colnames(blood)))]
  }else{blood_tmp <- blood}
  ind <- apply(blood_tmp[,2:ncol(blood_tmp)], 1, function(x) all(is.na(x)))
  blood_tmp <- blood_tmp[ !ind, ]

  overlap_cdr3nt <- tissue[which(tissue$cdr3nt %in% blood_tmp$cdr3nt),]
  new_cdr3nt <-tissue[which(!(tissue$cdr3nt %in% blood_tmp$cdr3nt)),]
  new_perc <- format(round(nrow(new_cdr3nt)/nrow(tissue),3)*100, nsmall=1)
  overlap_perc <- format(round(nrow(overlap_cdr3nt)/nrow(tissue),3)*100, nsmall=1)
  return(data.frame("sample"=file,"stage"=stage, "clones_blood"=nrow(blood_tmp), "clones_tissue"=length(unique(tissue$cdr3nt)), 
                    "new_clones_tissue"=nrow(new_cdr3nt), "new_clones_tissue_perc"=new_perc,
                    "old_clones_tissue"=nrow(overlap_cdr3nt), "old_clones_tissue_perc"=overlap_perc))
}


blood_34 <- merged[["R34"]]
ind <- apply(blood_34[,2:ncol(blood_34)], 1, function(x) all(is.na(x)))
blood_34 <- blood_34[ !ind, ]

# Check if CD4 clones in CD8 clones
blood_CD4 <- blood_34[,grepl("CD4", colnames(blood_34))]
blood_CD4$CD4 <- apply(blood_CD4, 1, function(x) any(!is.na(x)))
blood_CD8 <- blood_34[,grepl("CD8", colnames(blood_34))]
blood_CD8$CD8 <- apply(blood_CD8, 1, function(x) any(!is.na(x)))
#sum(isTRUE(blood_CD8$CD8)&isTRUE(blood_CD4$CD4))

# Descriptive table tissue samples
get_stats <- function(df, blood){
   
  clt <- length(unique(df$cdr3nt))
  clt.count <- sum(df$count)
  clt.allo <- length(unique(df[df$allo==1,]$cdr3nt))
  clt.allo.count <- sum(df[df$allo==1,]$count)
  clt.new <- clt - sum(df$cdr3nt %in% blood$cdr3nt)
  clt.new.count <- sum(df[!(df$cdr3nt %in% blood$cdr3nt),]$count)
  clt.old <- sum(df$cdr3nt %in% blood$cdr3nt)
  clt.old.count <- sum(df[(df$cdr3nt %in% blood$cdr3nt),]$count)


  return(c(clt, clt.count, clt.allo, clt.allo.count, clt.new, clt.new.count, clt.old, clt.old.count))
}


# Overlap Table
res_tissue_34 <- bind_rows(lapply(stages, function(x) calc_CloneOverlap(file="R34", blood=blood_34, tissue = tissue_34, stage = x)))
blood_43 <- merged[["R43"]]


df.tissue <- data.frame(rbind(c("pat"="R34", "sample"="tissue", get_stats(tissue_34, blood_34)), c("pat"="R43", "sample"="tissue",get_stats(tissue_43, blood_43))))
colnames(df.tissue) <- c("Patient","Sample","clt", "clt.count", "clt.allo", "clt.allo.count", 
                         "clt.new", "clt.new.count", "clt.old", "clt.old.count")

df.tissue %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Overview of the allograft immune repertoire of patient R34 and R43 at time of biopsy", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))

# Overlap Table
res_tissue_43 <- bind_rows(lapply(stages, function(x) calc_CloneOverlap(file="R43", blood=blood_43, tissue = tissue_43, stage = x)))
res_tissue <- data.frame(rbind(res_tissue_34, res_tissue_43))
res_tissue %>% 
  `colnames<-`(c("Patient", "Sample", "Clones - blood", "Clones - tissue", "New Clones - tissue", "New Clones - tissue (%)", 
                 "Old Clones - tissue", "Old Clones - tissue (%)")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Overview of the immune repertoire samples of patient R34 and R43 (pre-transplant, alloreative, post-transplant and allograft)", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))


# Diversity estimates of the tissue samples
div.tissue <- data.frame(cbind("Sample"=c(rep("R34", 4),rep("R43",4)),
                               "Time.point"=rep(c("Tissue", "PreTX", "PostTX","donor-reactive"),2),
                               rbind(calc_RepertoireStatistics(data=tissue_34),calc_RepertoireStatistics(data=df.34.BL),
                                     calc_RepertoireStatistics(data=df.34.late), calc_RepertoireStatistics(data=df.34.low),
                                     calc_RepertoireStatistics(data=tissue_43),calc_RepertoireStatistics(data=df.43.BL), 
                                     calc_RepertoireStatistics(data=df.43.late), calc_RepertoireStatistics(data=df.43.low))))

div.tissue %>%
  dplyr::select(Sample, Time.point,Reads, largest.clone.count, largest.clone.freq, SpeciesRichness.Nuc, R20, R20allo, Clonality) %>%
  `colnames<-`(c("Patient", "Time point","Reads", "Largest Count", "Largest Frequency", "Species Richness (NT)", "R20", 
                 "R20 allo", "Clonality")) %>%
  flextable() %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Diversity estimates of the tissue samples R34 and R43", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))


```

## Circos Plots for R34 and R43

```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE, fig.align="center", fig.show="hold", out.width="50%", fig.width=12, fig.height=6, fig.cap="VJ usage comparison between post-transplant and allograft sample of one selected individual (R34). Frequencies of the CDR3 sequences in the post-transplant sample were readjusted according to the provided CD4 and CD8 ratios."}
# Selected plot in report
plot_VJcircos(data.tcr=df.34.BL, file.tcr=NULL, name="blood_34_PreTX_CD4+CD8")
plot_VJcircos(df.34.late, file.tcr=NULL, name="blood_34_PostTX_CD4+CD8")
plot_VJcircos(tissue_34, file.tcr=NULL, name="tissue_34")

```


```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE, fig.align="center", fig.show="hold", out.width="50%", fig.width=12, fig.height=6, fig.cap="VJ usage comparison between post-transplant and allograft sample of one selected individual (R43). Frequencies of the CDR3 sequences in the post-transplant sample were readjusted according to the provided CD4 and CD8 ratios."}
# Selected plot in report
plot_VJcircos(df.43.BL, file.tcr=NULL, name="blood_43_PreTX_CD4+CD8")
plot_VJcircos(df.43.late, file.tcr=NULL, name="blood_43_PostTX_CD4+CD8")
plot_VJcircos(tissue_43, file.tcr=NULL, name="tissue_43")

```



Since the rows of the frequency tables of VJ combinations include a lot of 1 entries, the JSD of VJ usage between the readjusted post-transplant repertoire and the allograft repertoire is computed for patients R34 and R43.

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE}
JensenShannonDivergence <- function(x, y, classes=2, by.var="cdr3nt"){

  df.merged <- dplyr::full_join(x,y, by=by.var)
  p <- df.merged$freq.x
  q <- df.merged$freq.y
  p[is.na(p)] <- 0
  q[is.na(q)] <- 0

  JSD <- -sum(((p+q)/2)* log(((p+q)/2),2)) - (1/2)* (-sum(p * log(p,2), na.rm = T) - sum(q*log(q,2),na.rm = T))

  return(JSD)
}

# R34
# Test if VJ usage is different between readjusted PostTX and allograft
df.34.late$group <- paste0(df.34.late$v, "x",df.34.late$j)
tissue_34$group <- paste0(tissue_34$v, "x",tissue_34$j)

dfVJ <- dplyr::full_join(melt(wtd.table(df.34.late$group, weights = df.34.late$count)), melt(wtd.table(tissue_34$group, weights = tissue_34$count)), by="Var1")
dfVJ[is.na(dfVJ)] <- 0
colnames(dfVJ) <- c("VJsegments", "PostTX", "Tissue")
chitest.34 <- chisq.test(dfVJ[,-1])

# JSD of VJ usage between PostTX and allograft
#df.34.BL$group <- paste0(df.34.BL$v, "x",df.34.BL$j)

df.34.late$group <- paste0(df.34.late$v, "x",df.34.late$j)
blood34.VJusage <- melt(wtd.table(df.34.late$group, weights=df.34.late$count))
blood34.VJusage <- blood34.VJusage %>%
  mutate(freq=value/sum(value))

tissue_34 <- tissue_34 %>%
  mutate(group=paste0(v,"x",j))
tissue34.VJusage <- melt(wtd.table(tissue_34$group, weights=tissue_34$count))  %>%
  mutate(freq=value/sum(value))

jsd.vjusage.34<- JensenShannonDivergence(x=blood34.VJusage, y=tissue34.VJusage, by.var="Var1")

# Save table with which JSD was claculated
R34.VJ <- dplyr::full_join(blood34.VJusage, tissue34.VJusage, by="Var1")
colnames(R34.VJ) <- c("VJpair", "PostTX.count", "PostTX.freq", "Tissue.count", "Tissue.freq")
write.xlsx(R34.VJ, paste0(out.path, "tbl_VJusage_R34.xlsx"), overwrite=T)

# 
# require(elrm)
# tmp <- rbind(cbind(df.34.late[,c("v","j")], "type"=rep("blood", nrow(df.34.late))),cbind(tissue_34[,c("v","j")], "type"=rep("tissue", nrow(tissue_34))))
# tmp <- tmp[order(tmp$v),]
# x <- data.frame(xtabs(~type + interaction(v, j), data = tmp))
# vjpairs <- do.call(rbind, strsplit(as.character(x$interaction.v..j.), ".", fixed=T))
# colnames(vjpairs) <- c("v","j")
# 
# cdat <- data.frame(vjpairs, x)
# 
# m.female <- elrm(formula = type/Freq ~ v, interest = ~v, iter = 22000, 
#     dataset = cdat, burnIn = 2000)

# R43
# Test if VJ usage is different
df.43.late$group <- paste0(df.43.late$v, "x",df.43.late$j)
tissue_43$group <- paste0(tissue_43$v, "x",tissue_43$j)

dfVJ <- dplyr::full_join(melt(wtd.table(df.43.late$group, weights=df.43.late$count)), melt(wtd.table(tissue_43$group, weights=tissue_43$count)), by="Var1")
dfVJ[is.na(dfVJ)] <- 0
colnames(dfVJ) <- c("VJsegments", "PostTX", "Tissue")
chitest.43 <- chisq.test(dfVJ[,-1])

# JSD of VJ usage
df.43.BL$group <- paste0(df.43.BL$v, "x",df.43.BL$j)
df.43.late$group <- paste0(df.43.late$v, "x",df.43.late$j)

blood43.VJusage <- melt(wtd.table(df.43.late$group, weights=df.43.late$count))
blood43.VJusage <- blood43.VJusage %>%
  mutate(freq=value/sum(value))

tissue_43 <- tissue_43 %>%
  mutate(group=paste0(v,"x",j))
tissue43.VJusage <- melt(wtd.table(tissue_43$group, weights=tissue_43$count))  %>%
  mutate(freq=value/sum(value))

jsd.vjusage.43 <- JensenShannonDivergence(x=blood43.VJusage, y=tissue43.VJusage, by.var="Var1")

# Save table with which JSD was claculated
R43.VJ <- dplyr::full_join(blood43.VJusage, tissue43.VJusage, by="Var1")
colnames(R43.VJ) <- c("VJpair", "PostTX.count", "PostTX.freq", "Tissue.count", "Tissue.freq")
write.xlsx(R43.VJ, paste0(out.path, "tbl_VJusage_R43.xlsx"), overwrite = T)

# Summarize output
tbl.VJusage <- data.frame("Variable"=c("VJ-usage", "VJ-usage"),
                          "Patient"=c("R34","R43"),
                          "Sample.1"=c("PostTX","PostTX"),
                          "Sample.2"=c("Tissue","Tissue"), 
                          "Statistic"=round(c(chitest.34$statistic, chitest.43$statistic),1), 
                          "df"=c(chitest.34$parameter,chitest.43$parameter), 
                          "p"=c(chitest.34$p.value,chitest.43$p.value), 
                          "JSD"=c(jsd.vjusage.34, jsd.vjusage.43))


```



Differences in V usage and VJ usage were analyzed by conducting the chi-square test of independence and computing the Jensen Shannon divergence between the post-transplant sample and the allograft sample.

```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE}

# Test if V usage is different
# R34
dfV <- dplyr::full_join(melt(wtd.table(df.34.late$v, weights = df.34.late$count)),melt(wtd.table(tissue_34$v, weights = tissue_34$count)), by="Var1")
dfV[is.na(dfV)] <- 0
chitest.34 <- chisq.test(dfV[,-1])

blood34.Vusage <- melt(wtd.table(df.34.late$v, weights = df.34.late$count))

blood34.Vusage <- blood34.Vusage %>%
  mutate(freq=value/sum(value))

tissue34.Vusage <- melt(wtd.table(tissue_34$v, weights = tissue_34$count))  %>%
  mutate(freq=value/sum(value))
jsd.34 <- JensenShannonDivergence(x=blood34.Vusage, y=tissue34.Vusage, by.var="Var1")



#R43
dfV <- dplyr::full_join(melt(wtd.table(df.43.late$v, weights = df.43.late$count)),melt(wtd.table(tissue_43$v, weights=tissue_43$count)), by="Var1")
dfV[is.na(dfV)] <- 0
chitest.43 <- chisq.test(dfV[,-1])

blood43.Vusage <- melt(wtd.table(df.43.late$v, weights=df.43.late$count))
blood43.Vusage <- blood43.Vusage %>%
  mutate(freq=value/sum(value))

tissue43.Vusage <- melt(wtd.table(tissue_43$v, weights = tissue_43$count))  %>%
  mutate(freq=value/sum(value))
jsd.43 <- JensenShannonDivergence(x=blood43.Vusage, y=tissue43.Vusage, by.var="Var1")


tbl.Vusage <- data.frame("Variable"=c("V-usage", "V-usage"),"Patient"=c("R34","R43"),"Sample.1"=c("PostTX","PostTX"),"Sample.2"=c("Tissue","Tissue"), "Statistic"=round(c(chitest.34$statistic, chitest.43$statistic),1), "df"=c(chitest.34$parameter,chitest.43$parameter), "p"=round(c(chitest.34$p.value,chitest.43$p.value),3), "JSD"=round(c(jsd.34, jsd.43),3))

tbl.tmp <- rbind(tbl.VJusage, tbl.Vusage)
tbl.tmp$p <- round(as.numeric(tbl.tmp$p),3)
tbl.tmp$JSD <- round(as.numeric(tbl.tmp$JSD),3)

tbl.tmp %>%
  flextable() %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Comparison in V usage and VJ usage between the readjusted PostTX and allograft repertoires of patient R34 and R43", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))
  

```

## Jensen-Shannon Divergence between periphery and allograft

Table XX states the Jensen-Shannon divergence (JSD) between the pre- and post transplant sample and the allograft sample respectively in the individuals R34 and R43. 

```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE}

get_HFcomponent <- function(sample){

    # Separate in bulk and high frequency component
  sample.hf <- sample[sample$freq >= quantile(sample$freq, 0.75),]
  return(sample.hf)
}


list.R34 <- list(df.34.BL, df.34.late, tissue_34)
list.R43 <- list(df.43.BL, df.43.late, tissue_43)

JSD.comp <- cbind(wrapper_JSD(list.R34[c(1,3)], x_filename=F), wrapper_JSD(list.R34[c(2,3)], x_filename=F), wrapper_JSD(list.R34[c(1,2)], x_filename=F),
              wrapper_JSD(list.R43[c(1,3)], x_filename=F), wrapper_JSD(list.R43[c(2,3)], x_filename=F), wrapper_JSD(list.R43[c(1,2)], x_filename=F))


JSD.tissue <- data.frame("Patient"=c("R34","R34","R34","R43","R43", "R43"),"Sample.1"=rep(c("PreTX","PostTX","PreTX"),2), "Sample.2"=rep(c("Tissue","Tissue", "PostTX"),2), "JSD"=t(JSD.comp))
colnames(JSD.tissue)[4:7] <- c("JSD.all", "JSD.1000", "JSD.500", "JSD.100")
JSD.tissue[,4:7] <- apply(JSD.tissue[,4:7],2, function(x) round(x,3))

# Jensen-Shannon Divergence for the high-frequency component
JSD.tissue %>%
  flextable() %>%
  autofit %>%
  colformat_num(digits=2) %>%
  bold(part = "header") %>%
  set_caption("Jensen-Shannon Divergence (JSD)", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))

write.xlsx(JSD.tissue, paste0(out.path, "table_tcopies_JSD_tissue.xlsx"), overwrite=T)
  
JSD.comp <- c(JensenShannonDivergence(get_HFcomponent(df.34.BL), get_HFcomponent(tissue_34)), JensenShannonDivergence(get_HFcomponent(df.34.late), get_HFcomponent(tissue_34)), JensenShannonDivergence(get_HFcomponent(df.34.BL), get_HFcomponent(df.34.late)),
              JensenShannonDivergence(get_HFcomponent(df.43.BL), get_HFcomponent(tissue_43)), JensenShannonDivergence(get_HFcomponent(df.43.late), get_HFcomponent(tissue_43)), JensenShannonDivergence(get_HFcomponent(df.43.BL), get_HFcomponent(df.43.late)))

JSD.tissue.hf <- data.frame("Patient"=c("R34","R34","R34","R43","R43", "R43"),"Sample.1"=rep(c("PreTX","PostTX","PreTX"),2), "Sample.2"=rep(c("Tissue","Tissue", "PostTX"),2), "JSD"=round(JSD.comp,3))

JSD.tissue.hf %>%
  flextable() %>%
  autofit %>%
  colformat_num(digits=2) %>%
  bold(part = "header") %>%
  set_caption("Jensen-Shannon Divergence (JSD) considering only the high frequency component of the pre-transplant (PreTX) and post-transplant (PostTX) samples", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))

```

For Table XX, the pairwise Jensen-Shannon Divergence differs from Table XX since only the high frequency component (HF) of the pre-transplant (PreTX) and the post-transplant (PostTX) repertoires are considered. The HF component comprises all clonotypes which exceed a frequency of the 3rd quartile after CD4.CD8 readjustment of the respective sample.


```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE, fig.align="center", fig.width=12, fig.height=6, fig.cap="Jensen-Shannon divergence (JSD) between periphery and allograft repertoire samples of patient R34"}
tmp.JSD <- JSD.tissue[JSD.tissue$Patient %in% "R34",]
tmp <- melt(tmp.JSD, id.vars = colnames(JSD.tissue)[1:3])
tmp$group <- paste0(tmp$Sample.1, " and ", tmp$Sample.2)
tmp$group <- factor(tmp$group, levels=c("PreTX and Tissue", "PostTX and Tissue", "PreTX and PostTX"))

p.vj <- ggplot(data=tmp, aes(y=value, x=variable, label=Patient)) +
  geom_point(aes(x=variable, label=Patient),stat="identity", position = position_dodge()) +
  geom_line(aes(x=variable, group=group, col=group), alpha=1, size=0.8) +
  scale_x_discrete( "",label=c("All", "Top 1000", "Top 500", "Top 100")) +
  scale_y_continuous("Jensen-Shannon Divergence") +
  scale_color_manual(values=c("limegreen", "forestgreen", "blue1"),"") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5),legend.position="bottom", text = element_text(size=18))


p.vj
ggsave(filename = paste(out.path,"plot_JSD_R34_tcopies.png", sep = ""), plot=p.vj, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)

```

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=12, fig.height=6, fig.cap="Jensen-Shannon divergence (JSD) between periphery and allograft repertoire samples of patient R43"}
tmp.JSD <- JSD.tissue[JSD.tissue$Patient %in% "R43",]
tmp <- melt(tmp.JSD, id.vars = colnames(JSD.tissue)[1:3])
tmp$group <- paste0(tmp$Sample.1, " and ", tmp$Sample.2)

tmp$group <- factor(tmp$group, levels=c("PreTX and Tissue", "PostTX and Tissue", "PreTX and PostTX"))

p.vj <- ggplot(data=tmp, aes(y=value, x=variable, label=Patient))+
  geom_point(aes(x=variable, label=Patient),stat="identity", position = position_dodge()) +
  geom_line(aes(x=variable, group=group, col=group), alpha=1, size=0.8) +
  scale_x_discrete( "",label=c("All", "Top 1000", "Top 500", "Top 100")) +
  scale_y_continuous("Jensen-Shannon Divergence") +
  scale_color_manual(values=c("limegreen", "forestgreen", "blue1"),"") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5),legend.position="bottom", text = element_text(size=18))

p.vj
ggsave(filename = paste(out.path,"plot_JSD_R43_tcopies.png", sep = ""), plot=p.vj, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)

```


```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE}
JensenShannonDivergence <- function(df1, df2, var_freq=F){

  if(!var_freq){
      df.merged <- dplyr::full_join(df1[,c("cdr3nt", "freq")],df2[,c("cdr3nt", "freq")], by="cdr3nt")
    p <- df.merged$freq.x
    q <- df.merged$freq.y
  }else{
    p <- df1
    q <- df2
  }
  p[is.na(p)] <- 0
  q[is.na(q)] <- 0
  
  JSD <- -sum(((p+q)/2)* log(((p+q)/2),2)) - (1/2)* (-sum(p * log(p,2), na.rm = T) - sum(q*log(q,2),na.rm = T))
  
  return(JSD)
}


ovl.R34.BL <- inner_join(df.34.BL[,c("cdr3nt","freq")],tissue_34[,c("cdr3nt","freq")], by="cdr3nt")
ovl.R34.BL.JSD <- JensenShannonDivergence(df1=ovl.R34.BL$freq.x, df2=ovl.R34.BL$freq.y, var_freq = T)

ovl.R34.late <- inner_join(df.34.late[,c("cdr3nt","freq")],tissue_34[,c("cdr3nt","freq")], by="cdr3nt")
ovl.R34.late.JSD <- JensenShannonDivergence(df1=ovl.R34.late$freq.x, df2=ovl.R34.late$freq.y, var_freq = T)

ovl.R43.BL <- inner_join(df.43.BL[,c("cdr3nt","freq")],tissue_43[,c("cdr3nt","freq")], by="cdr3nt")
ovl.R43.BL.JSD <- JensenShannonDivergence(df1=ovl.R43.BL$freq.x, df2=ovl.R43.BL$freq.y, var_freq = T)

ovl.R43.late <- inner_join(df.43.late[,c("cdr3nt","freq")],tissue_43[,c("cdr3nt","freq")], by="cdr3nt")
ovl.R43.late.JSD <- JensenShannonDivergence(df1=ovl.R43.late$freq.x, df2=ovl.R43.late$freq.y, var_freq = T)

JSD.overlap <- data.frame("Patient"=rep(c("R34","R43"),each=2),"Clts"=c(nrow(df.34.BL), nrow(df.34.late), nrow(df.43.BL), nrow(df.43.late)),
                          "Sample.1"=rep(c("PreTX","PostTX"),2), "Sample.2"=rep("Tissue",4), "Clts"=rep(c(nrow(tissue_34), nrow(tissue_43)),each=2),
                          "Ovl.clts"=c(nrow(ovl.R34.BL),nrow(ovl.R34.late),nrow(ovl.R43.BL),nrow(ovl.R43.late)),
                          "JSD.ovl"=c(ovl.R34.BL.JSD,ovl.R34.late.JSD,ovl.R43.BL.JSD ,ovl.R43.late.JSD))
JSD.overlap %>%
  flextable() %>%
  autofit %>%
  colformat_num(digits=2) %>%
  bold(part = "header") %>%
  set_caption("Jensen-Shannon Divergence (JSD) of the overlap between the periphery and the allorgraft sample for patient R34 and R43", autonum = run_autonum(seq_id = "tab", bkm = newBKM()))

```


## Top clonotypes in periphery versus allograft

```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE}


get_ranking <- function(df1, df2, pat, name1, name2){
  # df2=df.43.late; df1=tissue_43; pat="R43"; name2 = "periph"; name1 = "tissue"
  df1 <- df1[order(df1$freq, decreasing = T),]
  df1$rank <- rank(-df1$freq, ties.method = "max")
  df1$clts <- length(unique(df1$cdr3nt))
  df1$reads <- sum(df1$count)
  
  # Separate in bulk and high frequency component  
  tbl.counts <- table(df1$count)  
  threshold.counts <- as.numeric(names(tbl.counts[tbl.counts == 1][2]))  
  df1$expanded <- 0 #bulk 
  df1[df1$count >= threshold.counts,]$expanded <- 1  

  
  df2<- df2[order(df2$freq, decreasing = T),]
  df2$rank <- rank(-df2$freq, ties.method = "max")
  df2$clts <- length(unique(df2$cdr3nt))
  df2$reads <- sum(df2$count)
  
  # Separate in bulk and high frequency component  
  tbl.counts <- table(df2$count)  
  threshold.counts <- as.numeric(names(tbl.counts[tbl.counts == 1][2]))  
  df2$expanded <- 0 #bulk 
  df2[df1$count >= threshold.counts,]$expanded <- 1  
  
  tbl <- left_join(df1[df1$expanded == 1,c("cdr3nt","rank","expanded","freq","count","clts", "reads")], df2[,c("cdr3nt","rank","expanded","freq","count","clts", "reads")], by="cdr3nt")
  tbl <- cbind(rep(pat,nrow(tbl)),tbl)
  colnames(tbl) <- c("Patient","cdr3nt", paste0(name1,"-",c("rank","expanded","freq","count","clts", "reads")), paste0(name2,"-",c("rank","expanded","freq","count","clts", "reads")))
  
  return(tbl)
}

top20ranking <- list("R34-periph"=get_ranking(df.34.late, tissue_34, pat="R34", name1 = "periph", name2 = "tissue"), 
                     "R34-tissue"=get_ranking(tissue_34, df.34.late, pat="R34", name1 = "tissue", name2 = "periph"),
                     "R43-periph"=get_ranking(df.43.late, tissue_43, pat="R43", name1 = "periph", name2 = "tissue"),
                     "R43-tissue"=get_ranking(tissue_43, df.43.late, pat="R43", name1 = "tissue", name2 = "periph"))

write.xlsx(top20ranking, file=paste0(out.path,"table_tcopies_EXPranking_periph_tissue.xlsx"), overwrite=T)


get_ranking <- function(df1, df2, pat, name1, name2){
  # df2=df.43.late; df1=tissue_43; pat="R43"; name2 = "periph"; name1 = "tissue"
  df1 <- df1[order(df1$freq, decreasing = T),]
  df1$rank <- rank(-df1$freq, ties.method = "max")
  df1$clts <- length(unique(df1$cdr3nt))
  df1$reads <- sum(df1$count)
  
  df2<- df2[order(df2$freq, decreasing = T),]
  df2$rank <- rank(-df2$freq, ties.method = "max")
  df2$clts <- length(unique(df2$cdr3nt))
  df2$reads <- sum(df2$count)
  
  
  tbl <- left_join(df1[1:20,c("cdr3nt","rank","freq","count","clts", "reads")], df2[,c("cdr3nt","rank","freq","count","clts", "reads")], by="cdr3nt")
  tbl <- cbind(rep(pat,nrow(tbl)),tbl)
  colnames(tbl) <- c("Patient","cdr3nt", paste0(name1,"-",c("rank","freq","count","clts", "reads")), paste0(name2,"-",c("rank","freq","count","clts", "reads")))
  
  return(tbl)
}

top20ranking <- list("R34-periph"=get_ranking(df.34.late, tissue_34, pat="R34", name1 = "periph", name2 = "tissue"), 
                     "R34-tissue"=get_ranking(tissue_34, df.34.late, pat="R34", name1 = "tissue", name2 = "periph"),
                     "R43-periph"=get_ranking(df.43.late, tissue_43, pat="R43", name1 = "periph", name2 = "tissue"),
                     "R43-tissue"=get_ranking(tissue_43, df.43.late, pat="R43", name1 = "tissue", name2 = "periph"))

write.xlsx(top20ranking, file=paste0(out.path,"table_tcopies_top20ranking_periph_tissue.xlsx"), overwrite=T)

```

```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE}
if(CONVENTIONAL_ONLY){
  knitr::knit_exit()
}
```


# The Network Approach for the Immune Repertoire
TCR similarity networks were constructed from the TCRβ sequences of CD4+ and CD8+ T cells at the amino acid (AA) level in order to determine CDR3 sequence diversity. A TCR repertoire similarity network is an undirected unweighted network G=(V, E) with a set of nodes V (CDR3 AA sequences) and a set of edges (similarity edges). The nodes indicate distinct CDR3 AA sequences linked by an edge defined when the pairwise Levenshtein distance equals 1 i.e. only one allowed AA deletion, insertion or substitution between the two AA sequences (Madi, Poran et al. 2017, Tickotsky, Sagiv et al. 2017). Further, the first and the last 3 amino acid bases were moitted for network construction.


## Network analysis of the top 10.000 most frequent clonotypes

```{r echo=FALSE, message=FALSE, warning=FALSE, size='tiny'}
out.path.nw <- "C:/Users/Mariella/Documents/PhD/Projects/TCOPIES/Output/"
folder_CD4_10000 <- paste0(out.path.nw,"network_tcopies_CD4_10ktop_LV1_final/")
folder_CD8_10000 <- paste0(out.path.nw,"network_tcopies_CD8_10ktop_LV1_final/")
tbl_CD4 <- read.xlsx(paste0(folder_CD4_10000 ,"table_tcopies_networks_CD4_full.xlsx"))
tbl_CD8 <- read.xlsx(paste0(folder_CD8_10000 ,"table_tcopies_networks_CD8_full.xlsx"))
tbl_CD4[tbl_CD4$rejection==1,]$rejection <- tbl_CD8[tbl_CD8$rejection==1,]$rejection <- "Rejectors"
tbl_CD4[tbl_CD4$rejection==0,]$rejection <- tbl_CD8[tbl_CD8$rejection==0,]$rejection <- "Controls"

tbl_CD4 <- tbl_CD4 %>%
  mutate(LC.allo = round(LC.allo*100,2),
         LC.newBX = round(LC.newBX*100,2)) %>%
  mutate(iso.perc = (total.iso/reduced.clones)*100) %>%
  relocate(iso.perc,.before=con.clones)
tbl_CD8 <- tbl_CD8 %>%
  mutate(LC.allo = round(LC.allo*100,2),
         LC.newBX = round(LC.newBX*100,2)) %>%
  mutate(iso.perc = (total.iso/reduced.clones)*100) %>%
  relocate(iso.perc,.before=con.clones)

p1 <- grouped_boxplot("edge.density", yname = "Edge density", title="CD4", tbl_CD4, plot = F)
p2 <- grouped_boxplot("edge.density", yname = "", title="CD8", tbl_CD8, plot = F)
p3 <- grouped_boxplot("modularity", yname = "Modularity", title="CD4", tbl_CD4, plot = F)
p4 <- grouped_boxplot("modularity", yname = "", title="CD8", tbl_CD8, plot = F)
p5 <- grouped_boxplot("clustering.coefficient", yname = "Clustering coefficient", title="CD4", tbl_CD4, plot = F)
p6 <- grouped_boxplot("clustering.coefficient", yname = "", title="CD8", tbl_CD8, plot = F)
p7 <- grouped_boxplot("char.pathlength.unconT", yname = "Characteristic path length", title="CD4", tbl_CD4, plot = F)
p8 <- grouped_boxplot("char.pathlength.unconT", yname = "", title="CD8", tbl_CD8, plot = F)
p9 <- grouped_boxplot("assortativity", yname = "Assortativity", title="CD4", tbl_CD4, plot = F)
p10 <- grouped_boxplot("assortativity", yname = "", title="CD8", tbl_CD8, plot = F)

# ggarrange(p1,p2,nrow=1, common.legend = T)
# ggarrange(p3,p4,nrow=1, common.legend = T)
# ggarrange(p5,p6,nrow=1, common.legend = T)
# ggarrange(p7,p8,nrow=1, common.legend = T)
# ggarrange(p9,p10,nrow=1, common.legend = T)

```

```{r echo=FALSE, cache=T,message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap=""}
my_cols <-brewer.pal(8, "Paired")[c(1,2,7)]
p1 <- ggplot(tbl_CD4, aes(x=LC.allo, y=con.clones, col=stage)) +
    geom_point(data=tbl_CD4, aes(shape=rejection), size=2) +
  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_shape_manual("", values = c(16,17)) +
  scale_x_continuous("Donor-reactive clones in largest cluster (%)",limits=c(0,100)) +
  scale_y_continuous("Connected clones in the full network (%)",limits=c(0,50)) +
  theme_bw() +
  ggtitle("CD4") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

p2 <- ggplot(tbl_CD8, aes(x=LC.allo, y=con.clones, col=stage)) +
    geom_point(data=tbl_CD8, aes(shape=rejection), size=2) +
  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_shape_manual("", values = c(16,17)) +
  scale_x_continuous("Donor-reactive clones in largest cluster (%)",limits=c(0,100)) +
  scale_y_continuous("", limits=c(0,50)) +
  theme_bw() +
  ggtitle("CD8") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

filename <- paste0("dotplot_conlonesvslargestclust_tcopies")
p12 <- ggarrange(p1,p2,nrow=1, common.legend = T)
p12
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p12, width = 20, height = 10, dpi = 640, units="cm", limitsize = F)


# Hypothesis testing
aov.LCallo1 <- summary(aov(tbl_CD4$LC.allo~tbl_CD4$stage))
aov.LCallo2 <- summary(aov(tbl_CD8$LC.allo~tbl_CD8$stage))
aov.full <- data.frame("Pheno"=c("CD4", "CD8"), "Groups"=c("all", "all"), "F"=c(aov.LCallo1[[1]]$`F value`[1],aov.LCallo2[[1]]$`F value`[1]),
           "df"=c(aov.LCallo1[[1]]$Df[1],aov.LCallo2[[1]]$Df[1]), "p"=round(c(aov.LCallo1[[1]]$`Pr(>F)`[1],aov.LCallo2[[1]]$`Pr(>F)`[1]),3))
aov.full %>% 
  `colnames<-`(c("Phenotype","Groups", "F", "df", "p")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Analysis of variance (ANOVA) for multiple comparison to assess overall differences in percentage of donor-reactive clonotypes in the largest network cluster", autonum = tbl_autonum) 

       
ttest.1 <- t.test(tbl_CD4[tbl_CD4$stage %in% "BL1",]$LC.allo, tbl_CD4[tbl_CD4$stage %in% "late1",]$LC.allo, paired=T)
ttest.2 <- t.test(tbl_CD4[tbl_CD4$stage %in% "BL1",]$LC.allo, tbl_CD4[tbl_CD4$stage %in% "low1",]$LC.allo, paired=T)
ttest.3 <- t.test(tbl_CD4[tbl_CD4$stage %in% "late1",]$LC.allo, tbl_CD4[tbl_CD4$stage %in% "low1",]$LC.allo, paired=T)

ttest.4 <- t.test(tbl_CD8[tbl_CD8$stage %in% "BL1",]$LC.allo, tbl_CD8[tbl_CD8$stage %in% "late1",]$LC.allo, paired=T)
ttest.5 <- t.test(tbl_CD8[tbl_CD8$stage %in% "BL1",]$LC.allo, tbl_CD8[tbl_CD8$stage %in% "low1",]$LC.allo, paired=T)
ttest.6 <- t.test(tbl_CD8[tbl_CD8$stage %in% "late1",]$LC.allo, tbl_CD8[tbl_CD8$stage %in% "low1",]$LC.allo, paired=T)

ttest.full <- data.frame("Pheno"=c(rep("CD4", 3),rep("CD8",3)), "Groups"=rep(c("Pre-Post", "Pre-allo", "Post-allo"),2),
                          "t"=c(ttest.1$statistic, ttest.2$statistic, ttest.3$statistic, ttest.4$statistic, ttest.5$statistic, ttest.6$statistic),
                          "df"=c(ttest.1$parameter, ttest.2$parameter, ttest.3$parameter, ttest.4$parameter, ttest.5$parameter, ttest.6$parameter), 
                          "p"=c(ttest.1$p.value, ttest.2$p.value, ttest.3$p.value, ttest.4$p.value, ttest.5$p.value, ttest.6$p.value), 
                          "Mean diff."=c(ttest.1$estimate[1], ttest.2$estimate[1], ttest.3$estimate[1], 
                                         ttest.4$estimate[1], ttest.5$estimate[1], ttest.6$estimate[1]),
                          "ci.lo"=c(ttest.1$conf.int[1],ttest.2$conf.int[1],ttest.3$conf.int[1],
                                    ttest.4$conf.int[1],ttest.5$conf.int[1],ttest.6$conf.int[1]),
                          "ci.up"=c(ttest.1$conf.int[2],ttest.2$conf.int[2],ttest.3$conf.int[2],
                                    ttest.4$conf.int[2],ttest.4$conf.int[2],ttest.6$conf.int[2]))
ttest.full[,c(3:8)] <- apply(ttest.full[,c(3:8)],2, function(x) round(x,3))

ttest.full %>% 
  unite("CI", ci.lo:ci.up, sep=",") %>%
  mutate(CI = paste0("(",CI,")")) %>%
  `colnames<-`(c("Phenotype","Groups", "t", "df", "p", "Mean Difference", "95% CI")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Results of multiple paired Student's t testing to assess differences in percentage of donor-reactive clonotypes in the largest network cluster", autonum = tbl_autonum) 
```

To better illustrate the difference in the pre-transplant and the post-transplant samples, we will omit the donor-reactive samples from the figure above.

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap=""}
my_cols <-brewer.pal(8, "Paired")[c(1,2,7)]
p1 <- ggplot(tbl_CD4[!tbl_CD4$stage %in% "low1",], aes(x=LC.allo, y=con.clones, col=stage)) +
    geom_point(data=tbl_CD4[!tbl_CD4$stage %in% "low1",], aes(shape=rejection), size=2) +
  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant")) +
    scale_shape_manual("", values = c(16,17)) +
  scale_x_continuous("Donor-reactive clones in largest cluster (%)") +
  scale_y_continuous("Connected clones in the full network (%)", limits = c(0,55)) +
  theme_bw() +
  ggtitle("CD4") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

p2 <- ggplot(tbl_CD8[!tbl_CD8$stage %in% "low1",], aes(x=LC.allo, y=con.clones, col=stage)) +
    geom_point(data=tbl_CD8[!tbl_CD8$stage %in% "low1",], aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +
  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant")) +
  scale_x_continuous("Donor-reactive clones in largest cluster (%)") +
  scale_y_continuous("", limits = c(0,55)) +
  theme_bw() +
  ggtitle("CD8") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

filename <- paste0("dotplot_conneclonesvslargestclust2_tcopies")
p12<- ggarrange(p1,p2,nrow=1, common.legend = T)
p12
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p12, width = 20, height = 10, dpi = 640, units="cm", limitsize = F)


```



```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap="Scatterplot of G(Vdegree) vs G(Csize)"}
my_cols <-brewer.pal(8, "Paired")[c(1,2,7)]
p3<-ggplot(tbl_CD4, aes(x=Gini.cl, y=Gini.degree, col=stage)) +
    geom_point(data=tbl_CD4, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Gini(Cluster size)", limits=c(0,0.3)) +
  scale_y_continuous("Gini(Vertex degree)", limits=c(0.65,1)) +
  theme_bw() +
  ggtitle("CD4") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

p4<- ggplot(tbl_CD8, aes(x=Gini.cl, y=Gini.degree, col=stage)) +
    geom_point(data=tbl_CD8, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Gini(Cluster size)", limits=c(0,0.3)) +
  scale_y_continuous("", limits=c(0.65,1)) +
  theme_bw() +
  ggtitle("CD8") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

filename <- paste0("dotplot_gini_tcopies")
p34 <- ggarrange(p3,p4,nrow=1, common.legend = T)
p34
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p34, width = 20, height = 10, dpi = 640, units="cm", limitsize = F)


aov.1 <- summary(aov(tbl_CD4$Gini.degree~tbl_CD4$stage))
aov.2 <- summary(aov(tbl_CD8$Gini.degree~tbl_CD8$stage))
aov.full <- data.frame("Pheno"=c("CD4", "CD8"), "Groups"=c("all", "all"), "F"=c(aov.1[[1]]$`F value`[1],aov.2[[1]]$`F value`[1]),
           "df"=c(aov.1[[1]]$Df[1],aov.2[[1]]$Df[1]), "p"=round(c(aov.1[[1]]$`Pr(>F)`[1],aov.2[[1]]$`Pr(>F)`[1]),3))
aov.full %>% 
  `colnames<-`(c("Phenotype","Groups", "F", "df", "p")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Analysis of variance (ANOVA) for multiple comparison to assess overall differences in Gini index values of degree distribution of the subject-specific networks across the time point groups", autonum = tbl_autonum) 

```
```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap="Scatterplot of G(BC) vs G(LocalEff)"}
my_cols <-brewer.pal(8, "Paired")[c(1,2,7)]
p5<-ggplot(tbl_CD4, aes(x=Gini.localeff, y=Gini.bc, col=stage)) +
    geom_point(data=tbl_CD4, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Gini(Local efficiency)", limits=c(0.65,1)) +
  scale_y_continuous("Gini(Betweenness centrality)", limits=c(0.65,1)) +
  theme_bw() +
  ggtitle("CD4") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

p6<- ggplot(tbl_CD8, aes(x=Gini.localeff, y=Gini.bc, col=stage)) +
    geom_point(data=tbl_CD8, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Gini(Local efficiency)", limits=c(0.65,1)) +
  scale_y_continuous("", limits=c(0.65,1)) +
  theme_bw() +
  ggtitle("CD8") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

filename <- paste0("dotplot_gini2_tcopies")
p56 <- ggarrange(p5,p6,nrow=1, common.legend = T)
p56
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p56, width = 20, height = 10, dpi = 640, units="cm", limitsize = F)

# Hypothesis testing
aov.1 <- summary(aov(tbl_CD4$Gini.localeff~tbl_CD4$stage))
aov.2 <- summary(aov(tbl_CD8$Gini.localeff~tbl_CD8$stage))
aov.full <- data.frame("Pheno"=c("CD4", "CD8"), "Groups"=c("all", "all"), "F"=c(aov.1[[1]]$`F value`[1],aov.2[[1]]$`F value`[1]),
           "df"=c(aov.1[[1]]$Df[1],aov.2[[1]]$Df[1]), "p"=round(c(aov.1[[1]]$`Pr(>F)`[1],aov.2[[1]]$`Pr(>F)`[1]),3))
aov.full %>% 
  `colnames<-`(c("Phenotype","Groups", "F", "df", "p")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Analysis of variance (ANOVA) for multiple comparison to assess overall differences in local efficiency of the subject-specific networks across the time point groups", autonum = tbl_autonum) 

ttest.1 <- t.test(tbl_CD4[tbl_CD4$stage %in% "BL1",]$Gini.localeff, tbl_CD4[tbl_CD4$stage %in% "late1",]$Gini.localeff, paired=T)
ttest.2 <- t.test(tbl_CD4[tbl_CD4$stage %in% "BL1",]$Gini.localeff, tbl_CD4[tbl_CD4$stage %in% "low1",]$Gini.localeff, paired=T)
ttest.3 <- t.test(tbl_CD4[tbl_CD4$stage %in% "late1",]$Gini.localeff, tbl_CD4[tbl_CD4$stage %in% "low1",]$Gini.localeff, paired=T)

ttest.4 <- t.test(tbl_CD8[tbl_CD8$stage %in% "BL1",]$Gini.localeff, tbl_CD8[tbl_CD8$stage %in% "late1",]$Gini.localeff, paired=T)
ttest.5 <- t.test(tbl_CD8[tbl_CD8$stage %in% "BL1",]$Gini.localeff, tbl_CD8[tbl_CD8$stage %in% "low1",]$Gini.localeff, paired=T)
ttest.6 <- t.test(tbl_CD8[tbl_CD8$stage %in% "late1",]$Gini.localeff, tbl_CD8[tbl_CD8$stage %in% "low1",]$Gini.localeff, paired=T)

ttest.full <- data.frame("Pheno"=c(rep("CD4", 3),rep("CD8",3)), "Groups"=rep(c("Pre-Post", "Pre-allo", "Post-allo"),2),
                          "t"=c(ttest.1$statistic, ttest.2$statistic, ttest.3$statistic, ttest.4$statistic, ttest.5$statistic, ttest.6$statistic),
                          "df"=c(ttest.1$parameter, ttest.2$parameter, ttest.3$parameter, ttest.4$parameter, ttest.5$parameter, ttest.6$parameter), 
                          "p"=c(ttest.1$p.value, ttest.2$p.value, ttest.3$p.value, ttest.4$p.value, ttest.5$p.value, ttest.6$p.value), 
                          "Mean diff."=c(ttest.1$estimate[1], ttest.2$estimate[1], ttest.3$estimate[1], 
                                         ttest.4$estimate[1], ttest.5$estimate[1], ttest.6$estimate[1]),
                          "ci.lo"=c(ttest.1$conf.int[1],ttest.2$conf.int[1],ttest.3$conf.int[1],
                                    ttest.4$conf.int[1],ttest.5$conf.int[1],ttest.6$conf.int[1]),
                          "ci.up"=c(ttest.1$conf.int[2],ttest.2$conf.int[2],ttest.3$conf.int[2],
                                    ttest.4$conf.int[2],ttest.4$conf.int[2],ttest.6$conf.int[2]))
ttest.full[,c(3:8)] <- apply(ttest.full[,c(3:8)],2, function(x) round(x,3))

ttest.full %>% 
  unite("CI", ci.lo:ci.up, sep=",") %>%
  mutate(CI = paste0("(",CI,")")) %>%
  `colnames<-`(c("Phenotype","Groups", "t", "df", "p", "Mean Difference", "95% CI")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Results of multiple paired Student's t testing to assess differences in the Gini index measure for the local efficiency distribution between time point groups", autonum = tbl_autonum) 

```

```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap="Scatterplot of G(BC) vs G(LocalEff)"}
my_cols <-brewer.pal(8, "Paired")[c(1,2,7)]
p7<-ggplot(tbl_CD4, aes(x=clustering.coefficient, y=char.pathlength.unconT, col=stage)) +
    geom_point(data=tbl_CD4, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Clustering coefficient", limits=c(0,1)) +
  scale_y_continuous("Characteristic path length", limits=c(1,15), breaks = seq(1,15,2)) +
  theme_bw() +
  ggtitle("CD4") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

p8<- ggplot(tbl_CD8, aes(x=clustering.coefficient, y=char.pathlength.unconT, col=stage)) +
    geom_point(data=tbl_CD8, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Clustering coefficient", limits=c(0,1)) +
  scale_y_continuous("", limits=c(1,15), breaks = seq(1,15,2)) +
  theme_bw() +
  ggtitle("CD8") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

filename <- paste0("dotplot_cc_cpl_tcopies")
p78 <- ggarrange(p7,p8,nrow=1, common.legend = T)
p78
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p78, width = 20, height = 10, dpi = 640, units="cm", limitsize = F)



# Hypothesis testing
aov.1 <- summary(aov(tbl_CD4$char.pathlength.unconT~tbl_CD4$stage))
aov.2 <- summary(aov(tbl_CD8$char.pathlength.unconT~tbl_CD8$stage))
aov.full <- data.frame("Pheno"=c("CD4", "CD8"), "Groups"=c("all", "all"), "F"=c(aov.1[[1]]$`F value`[1],aov.2[[1]]$`F value`[1]),
           "df"=c(aov.1[[1]]$Df[1],aov.2[[1]]$Df[1]), "p"=round(c(aov.1[[1]]$`Pr(>F)`[1],aov.2[[1]]$`Pr(>F)`[1]),3))
aov.full %>% 
  `colnames<-`(c("Phenotype","Groups", "F", "df", "p")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Analysis of variance (ANOVA) for multiple comparison to assess overall differences in characteristic path length of the subject-specific networks across the time point groups", autonum = tbl_autonum) 



ttest.1 <- t.test(tbl_CD4[tbl_CD4$stage %in% "BL1",]$char.pathlength.unconT, tbl_CD4[tbl_CD4$stage %in% "late1",]$char.pathlength.unconT, paired=T)
ttest.2 <- t.test(tbl_CD4[tbl_CD4$stage %in% "BL1",]$char.pathlength.unconT, tbl_CD4[tbl_CD4$stage %in% "low1",]$char.pathlength.unconT, paired=T)
ttest.3 <- t.test(tbl_CD4[tbl_CD4$stage %in% "late1",]$char.pathlength.unconT, tbl_CD4[tbl_CD4$stage %in% "low1",]$char.pathlength.unconT, paired=T)

ttest.4 <- t.test(tbl_CD8[tbl_CD8$stage %in% "BL1",]$char.pathlength.unconT, tbl_CD8[tbl_CD8$stage %in% "late1",]$char.pathlength.unconT, paired=T)
ttest.5 <- t.test(tbl_CD8[tbl_CD8$stage %in% "BL1",]$char.pathlength.unconT, tbl_CD8[tbl_CD8$stage %in% "low1",]$char.pathlength.unconT, paired=T)
ttest.6 <- t.test(tbl_CD8[tbl_CD8$stage %in% "late1",]$char.pathlength.unconT, tbl_CD8[tbl_CD8$stage %in% "low1",]$char.pathlength.unconT, paired=T)

ttest.full <- data.frame("Pheno"=c(rep("CD4", 3),rep("CD8",3)), "Groups"=rep(c("Pre-Post", "Pre-allo", "Post-allo"),2),
                          "t"=c(ttest.1$statistic, ttest.2$statistic, ttest.3$statistic, ttest.4$statistic, ttest.5$statistic, ttest.6$statistic),
                          "df"=c(ttest.1$parameter, ttest.2$parameter, ttest.3$parameter, ttest.4$parameter, ttest.5$parameter, ttest.6$parameter), 
                          "p"=c(ttest.1$p.value, ttest.2$p.value, ttest.3$p.value, ttest.4$p.value, ttest.5$p.value, ttest.6$p.value), 
                          "Mean diff."=c(ttest.1$estimate[1], ttest.2$estimate[1], ttest.3$estimate[1], 
                                         ttest.4$estimate[1], ttest.5$estimate[1], ttest.6$estimate[1]),
                          "ci.lo"=c(ttest.1$conf.int[1],ttest.2$conf.int[1],ttest.3$conf.int[1],
                                    ttest.4$conf.int[1],ttest.5$conf.int[1],ttest.6$conf.int[1]),
                          "ci.up"=c(ttest.1$conf.int[2],ttest.2$conf.int[2],ttest.3$conf.int[2],
                                    ttest.4$conf.int[2],ttest.4$conf.int[2],ttest.6$conf.int[2]))
ttest.full[,c(3:8)] <- apply(ttest.full[,c(3:8)],2, function(x) round(x,3))

ttest.full %>% 
  unite("CI", ci.lo:ci.up, sep=",") %>%
  mutate(CI = paste0("(",CI,")")) %>%
  `colnames<-`(c("Phenotype","Groups", "t", "df", "p", "Mean Difference", "95% CI")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Results of multiple paired Student's t testing to assess differences in characteristic path length of the subject-specific networks between time point groups", autonum = tbl_autonum) 

```

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap="Scatterplot of G(BC) vs G(LocalEff)"}
my_cols <-brewer.pal(8, "Paired")[c(1,2,7)]
p7<-ggplot(tbl_CD4, aes(x=modularity, y=assortativity, col=stage)) +
    geom_point(data=tbl_CD4, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Modularity", limits=c(0,1)) +
  scale_y_continuous("Assortativity", limits=c(0,1)) +
  theme_bw() +
  ggtitle("CD4") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

p8<- ggplot(tbl_CD8, aes(x=modularity, y=assortativity, col=stage)) +
    geom_point(data=tbl_CD8, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Modularity", limits=c(0,1)) +
  scale_y_continuous("", limits=c(0,1)) +
  theme_bw() +
  ggtitle("CD8") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

filename <- paste0("dotplot_mod_ass_tcopies")
p78 <- ggarrange(p7,p8,nrow=1, common.legend = T)
p78
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p78, width = 20, height = 10, dpi = 640, units="cm", limitsize = F)



# Hypothesis testing
aov.1 <- summary(aov(tbl_CD4$modularity~tbl_CD4$stage))
aov.2 <- summary(aov(tbl_CD8$modularity~tbl_CD8$stage))
aov.full <- data.frame("Pheno"=c("CD4", "CD8"), "Groups"=c("all", "all"), "F"=c(aov.1[[1]]$`F value`[1],aov.2[[1]]$`F value`[1]),
           "df"=c(aov.1[[1]]$Df[1],aov.2[[1]]$Df[1]), "p"=round(c(aov.1[[1]]$`Pr(>F)`[1],aov.2[[1]]$`Pr(>F)`[1]),3))
aov.full %>% 
  `colnames<-`(c("Phenotype","Groups", "F", "df", "p")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Analysis of variance (ANOVA) for multiple comparison to assess overall differences in modularity of the subject-specific networks across the time point groups", autonum = tbl_autonum) 


ttest.1 <- t.test(tbl_CD4[tbl_CD4$stage %in% "BL1",]$modularity, tbl_CD4[tbl_CD4$stage %in% "late1",]$modularity, paired=T)
ttest.2 <- t.test(tbl_CD4[tbl_CD4$stage %in% "BL1",]$modularity, tbl_CD4[tbl_CD4$stage %in% "low1",]$modularity, paired=T)
ttest.3 <- t.test(tbl_CD4[tbl_CD4$stage %in% "late1",]$modularity, tbl_CD4[tbl_CD4$stage %in% "low1",]$modularity, paired=T)

ttest.4 <- t.test(tbl_CD8[tbl_CD8$stage %in% "BL1",]$modularity, tbl_CD8[tbl_CD8$stage %in% "late1",]$modularity, paired=T)
ttest.5 <- t.test(tbl_CD8[tbl_CD8$stage %in% "BL1",]$modularity, tbl_CD8[tbl_CD8$stage %in% "low1",]$modularity, paired=T)
ttest.6 <- t.test(tbl_CD8[tbl_CD8$stage %in% "late1",]$modularity, tbl_CD8[tbl_CD8$stage %in% "low1",]$modularity, paired=T)

ttest.full <- data.frame("Pheno"=c(rep("CD4", 3),rep("CD8",3)), "Groups"=rep(c("Pre-Post", "Pre-allo", "Post-allo"),2),
                          "t"=c(ttest.1$statistic, ttest.2$statistic, ttest.3$statistic, ttest.4$statistic, ttest.5$statistic, ttest.6$statistic),
                          "df"=c(ttest.1$parameter, ttest.2$parameter, ttest.3$parameter, ttest.4$parameter, ttest.5$parameter, ttest.6$parameter), 
                          "p"=c(ttest.1$p.value, ttest.2$p.value, ttest.3$p.value, ttest.4$p.value, ttest.5$p.value, ttest.6$p.value), 
                          "Mean diff."=c(ttest.1$estimate[1], ttest.2$estimate[1], ttest.3$estimate[1], 
                                         ttest.4$estimate[1], ttest.5$estimate[1], ttest.6$estimate[1]),
                          "ci.lo"=c(ttest.1$conf.int[1],ttest.2$conf.int[1],ttest.3$conf.int[1],
                                    ttest.4$conf.int[1],ttest.5$conf.int[1],ttest.6$conf.int[1]),
                          "ci.up"=c(ttest.1$conf.int[2],ttest.2$conf.int[2],ttest.3$conf.int[2],
                                    ttest.4$conf.int[2],ttest.4$conf.int[2],ttest.6$conf.int[2]))
ttest.full[,c(3:8)] <- apply(ttest.full[,c(3:8)],2, function(x) round(x,3))

ttest.full %>% 
  unite("CI", ci.lo:ci.up, sep=",") %>%
  mutate(CI = paste0("(",CI,")")) %>%
  `colnames<-`(c("Phenotype","Groups", "t", "df", "p", "Mean Difference", "95% CI")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Results of multiple paired Student's t testing to assess differences in modularity of the subject-specific immune networks between time point groups", autonum = tbl_autonum) 

```

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap=""}
my_cols <-brewer.pal(8, "Paired")[c(1,2,7)]
p1<-ggplot(tbl_CD4, aes(x=modularity, y=char.pathlength.unconT, col=stage)) +
    geom_point(data=tbl_CD4, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Modularity", limits=c(0,1)) +
  scale_y_continuous("Characteristic path length", limits=c(1,15), breaks = seq(1,15,2)) +
  theme_bw() +
  ggtitle("CD4") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

p2<- ggplot(tbl_CD8, aes(x=modularity, y=char.pathlength.unconT, col=stage)) +
    geom_point(data=tbl_CD8, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Modularity", limits=c(0,1)) +
  scale_y_continuous("", limits=c(1,15), breaks = seq(1,15,2)) +
  theme_bw() +
  ggtitle("CD8") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

filename <- paste0("dotplot_mod_cpl_tcopies")
p12 <- ggarrange(p1,p2,nrow=1, common.legend = T)
p12
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p12, width = 20, height = 10, dpi = 640, units="cm", limitsize = F)
```
```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap=""}
my_cols <-brewer.pal(8, "Paired")[c(1,2,7)]
p1<-ggplot(tbl_CD4, aes(x=modularity, y=clustering.coefficient, col=stage)) +
    geom_point(data=tbl_CD4, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Modularity", limits=c(0,1)) +
  scale_y_continuous("Clustering coefficient", limits=c(0,1), breaks = seq(0,1,0.25)) +
  theme_bw() +
  ggtitle("CD4") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

p2<- ggplot(tbl_CD8, aes(x=modularity, y=clustering.coefficient, col=stage)) +
    geom_point(data=tbl_CD8, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Modularity", limits=c(0,1)) +
  scale_y_continuous("", limits=c(0,1), breaks = seq(0,1,0.25)) +
  theme_bw() +
  ggtitle("CD8") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

filename <- paste0("dotplot_mod_cc_tcopies")
p12 <- ggarrange(p1,p2,nrow=1, common.legend = T)
p12
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p12, width = 20, height = 10, dpi = 640, units="cm", limitsize = F)

aov.1 <- summary(aov(tbl_CD4$clustering.coefficient~tbl_CD4$stage))
aov.2 <- summary(aov(tbl_CD8$clustering.coefficient~tbl_CD8$stage))
aov.full <- data.frame("Pheno"=c("CD4", "CD8"), "Groups"=c("all", "all"), "F"=c(aov.1[[1]]$`F value`[1],aov.2[[1]]$`F value`[1]),
           "df"=c(aov.1[[1]]$Df[1],aov.2[[1]]$Df[1]), "p"=round(c(aov.1[[1]]$`Pr(>F)`[1],aov.2[[1]]$`Pr(>F)`[1]),3))
aov.full %>% 
  `colnames<-`(c("Phenotype","Groups", "F", "df", "p")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Analysis of variance (ANOVA) for multiple comparison to assess overall differences in the clustering coefficient of the subject-specific networks across the time point groups", autonum = tbl_autonum) 
```



```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap=""}
my_cols <-brewer.pal(8, "Paired")[c(1,2,7)]
p1<-ggplot(tbl_CD4, aes(x=modularity, y=iso.perc, col=stage)) +
    geom_point(data=tbl_CD4, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Modularity", limits=c(0,1)) +
  scale_y_continuous("Percentage of isolates", limits=c(0,100), breaks = seq(0,100,25)) +
  theme_bw() +
  ggtitle("CD4") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

p2<- ggplot(tbl_CD8, aes(x=modularity, y=iso.perc, col=stage)) +
    geom_point(data=tbl_CD8, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Modularity", limits=c(0,1)) +
  scale_y_continuous("", limits=c(0,100), breaks = seq(0,100,25)) +
  theme_bw() +
  ggtitle("CD8") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

filename <- paste0("dotplot_mod_iso_tcopies")
p12 <- ggarrange(p1,p2,nrow=1, common.legend = T)
p12
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p12, width = 20, height = 10, dpi = 640, units="cm", limitsize = F)

aov.1 <- summary(aov(tbl_CD4$iso.perc~tbl_CD4$stage))
aov.2 <- summary(aov(tbl_CD8$iso.perc~tbl_CD8$stage))
aov.full <- data.frame("Pheno"=c("CD4", "CD8"), "Groups"=c("all", "all"), "F"=c(aov.1[[1]]$`F value`[1],aov.2[[1]]$`F value`[1]),
           "df"=c(aov.1[[1]]$Df[1],aov.2[[1]]$Df[1]), "p"=round(c(aov.1[[1]]$`Pr(>F)`[1],aov.2[[1]]$`Pr(>F)`[1]),3))
aov.full %>% 
  `colnames<-`(c("Phenotype","Groups", "F", "df", "p")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Analysis of variance (ANOVA) for multiple comparison to assess overall differences in percentage of isolated clonotypes in the subject-specific networks across the time point groups", autonum = tbl_autonum) 
```


```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap=""}
my_cols <-brewer.pal(8, "Paired")[c(1,2,7)]
p1<-ggplot(tbl_CD4, aes(x=modularity, y=con.clones, col=stage)) +
    geom_point(data=tbl_CD4, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Modularity", limits=c(0,1)) +
  scale_y_continuous("Connected clones (%)", limits=c(0,100), breaks = seq(0,100,25)) +
  theme_bw() +
  ggtitle("CD4") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

p2<- ggplot(tbl_CD8, aes(x=modularity, y=con.clones, col=stage)) +
    geom_point(data=tbl_CD8, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Modularity", limits=c(0,1)) +
  scale_y_continuous("", limits=c(0,100), breaks = seq(0,100,25)) +
  theme_bw() +
  ggtitle("CD8") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

filename <- paste0("dotplot_mod_conc_tcopies")
p12 <- ggarrange(p1,p2,nrow=1, common.legend = T)
p12
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p12, width = 20, height = 10, dpi = 640, units="cm", limitsize = F)

aov.1 <- summary(aov(tbl_CD4$con.clones~tbl_CD4$stage))
aov.2 <- summary(aov(tbl_CD8$con.clones~tbl_CD8$stage))
aov.full <- data.frame("Pheno"=c("CD4", "CD8"), "Groups"=c("all", "all"), "F"=c(aov.1[[1]]$`F value`[1],aov.2[[1]]$`F value`[1]),
           "df"=c(aov.1[[1]]$Df[1],aov.2[[1]]$Df[1]), "p"=round(c(aov.1[[1]]$`Pr(>F)`[1],aov.2[[1]]$`Pr(>F)`[1]),3))
aov.full %>% 
  `colnames<-`(c("Phenotype","Groups", "F", "df", "p")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Analysis of variance (ANOVA) for multiple comparison to assess overall differences in connected clonotypes in the subject-specific networks across the time point groups", autonum = tbl_autonum) 

aov.1 <- summary(aov(tbl_CD4$edge.density~tbl_CD4$stage))
aov.2 <- summary(aov(tbl_CD8$edge.density~tbl_CD8$stage))
aov.full <- data.frame("Pheno"=c("CD4", "CD8"), "Groups"=c("all", "all"), "F"=c(aov.1[[1]]$`F value`[1],aov.2[[1]]$`F value`[1]),
           "df"=c(aov.1[[1]]$Df[1],aov.2[[1]]$Df[1]), "p"=round(c(aov.1[[1]]$`Pr(>F)`[1],aov.2[[1]]$`Pr(>F)`[1]),3))
aov.full %>% 
  `colnames<-`(c("Phenotype","Groups", "F", "df", "p")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Analysis of variance (ANOVA) for multiple comparison to assess overall differences in edge density in the subject-specific networks across the time point groups", autonum = tbl_autonum) 
```



```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap=""}
my_cols <-brewer.pal(8, "Paired")[c(1,2,7)]
p1<-ggplot(tbl_CD4, aes(x=modularity, y=edge.density, col=stage)) +
    geom_point(data=tbl_CD4, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Modularity", limits=c(0,1)) +
  scale_y_continuous("Edge density", limits=c(0,0.05), breaks = seq(0,0.05,0.01)) +
  theme_bw() +
  ggtitle("CD4") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

p2 <- ggplot(tbl_CD8, aes(x=modularity, y=edge.density, col=stage)) +
    geom_point(data=tbl_CD8, aes(shape=rejection), size=2) +
      scale_shape_manual("", values = c(16,17)) +  scale_color_manual("", values=my_cols,labels=c("Pre-Transplant", "Post-Transplant", "donor-reactive")) +
  scale_x_continuous("Modularity", limits=c(0,1)) +
  scale_y_continuous("", limits=c(0,0.05), breaks = seq(0,0.05,0.01)) +
  theme_bw() +
  ggtitle("CD8") +
  theme(plot.title = element_text(hjust = 0.5), text = element_text(size=10), legend.position = "bottom")

filename <- paste0("dotplot_mod_edged_tcopies")
p12 <- ggarrange(p1,p2,nrow=1, common.legend = T)
p12
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p12, width = 20, height = 10, dpi = 640, units="cm", limitsize = F)

```


<!-- ```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap=""} -->
<!-- my_cols <-brewer.pal(8, "Paired")[c(1,2,7)] -->
<!-- p2a <- grouped_dotplot("LC.allo", yname = "Connected clones in largest subgraph", title="CD4", tbl_CD4, plot = F) -->
<!-- p3a <- grouped_dotplot("edge.density", yname = "Edge density", title="CD4", tbl_CD4, plot = F) -->
<!-- p4a <- grouped_dotplot("leven1", yname = "Percentage of clones with LV1", title="CD4", tbl_CD4, plot = F) -->
<!-- p5a <- grouped_dotplot("Gini.degree", yname = "Gini index of vertex degree", title="CD4", tbl_CD4, plot = F) -->

<!-- p2b <- grouped_dotplot("LC.allo", yname = "Connected clones in largest subgraph", title="CD8", tbl_CD8, plot = F) -->
<!-- p3b <- grouped_dotplot("edge.density", yname = "Edge density", title="CD8", tbl_CD8, plot = F) -->
<!-- p4b <- grouped_dotplot("leven1", yname = "Percentage of clones with LV1", title="CD8", tbl_CD8, plot = F) -->
<!-- p5b <- grouped_dotplot("Gini.degree", yname = "Gini index of vertex degree", title="CD8", tbl_CD8, plot = F) -->

<!-- ggarrange(p2a,p2b,common.legend = T) -->
<!-- ggarrange(p3a,p3b,common.legend = T) -->
<!-- ggarrange(p4a,p4b,common.legend = T) -->
<!-- ggarrange(p5a,p5b,common.legend = T) -->

<!-- ``` -->

<!-- ```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap=""} -->
<!-- my_cols <- c("#A6CEE3", "#1F78B4", "red4") -->
<!-- pa.1 <- ggplot()+ -->
<!--   geom_bar(data=tbl_CD4, aes(y=cluster.no, x=group, fill=stage),stat="identity", position = position_dodge()) + -->
<!--   scale_x_discrete( "",guide = guide_axis(angle = 45)) + -->
<!--   scale_y_continuous("Number of clusters") + -->
<!--   theme_bw() + -->
<!--   scale_fill_manual(values=my_cols,name = "", labels = c("Pre-Transplant", "Circulation at Biopsy", "Allograft at Biopsy")) + -->
<!--   ggtitle("CD4") + -->
<!--   facet_wrap(~rejection,scales="free_x") + -->
<!--   theme(plot.title = element_text(hjust = 0.5),legend.position="bottom", text = element_text(size=12)) -->

<!-- pa.2 <- ggplot()+ -->
<!--   geom_bar(data=tbl_CD8, aes(y=cluster.no, x=group, fill=stage),stat="identity", position = position_dodge()) + -->
<!--   scale_x_discrete( "",guide = guide_axis(angle = 45)) + -->
<!--   scale_y_continuous("Number of clusters") + -->
<!--   theme_bw() + -->
<!--   scale_fill_manual(values=my_cols,name = "", labels = c("Pre-Transplant", "Circulation at Biopsy", "Allograft at Biopsy")) + -->
<!--   ggtitle("CD8") + -->
<!--   facet_wrap(~rejection,scales="free_x") + -->
<!--   theme(plot.title = element_text(hjust = 0.5),legend.position="none", text = element_text(size=12)) -->
<!-- ggarrange(pa.1,pa.2, common.legend = T) -->

<!-- ``` -->

<!-- ```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap="Edge density of the connected top 10.000 most frequent clonotypes in the pre- and post-transplant networks grouped by rejection"} -->

<!-- filename <- paste0("plot_edgedensity_top10knet_tcopies") -->
<!-- ggarrange(p1, p2, ncol=2, nrow=1, common.legend = TRUE, legend="bottom") -->
<!-- ggsave(filename = paste(out.path,filename,".png", sep = ""), width = 20, height = 20, dpi = 640, units="cm", limitsize = F) -->

<!-- ``` -->

<!-- ```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap="Percentage of the connected top 10.000 most frequent clonotypes in the pre- and post-transplant networks grouped by rejection"} -->
<!-- filename <- paste0("plot_percLV1_top10knet_tcopies") -->
<!-- ggarrange(p3, p4, ncol=2, nrow=1, common.legend = TRUE, legend="bottom") -->
<!-- ggsave(filename = paste(out.path,filename,".png", sep = ""), width = 20, height = 20, dpi = 640, units="cm", limitsize = F) -->

<!-- ``` -->

<!-- ```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap="Boxplot of the clustering coefficient estimate of the top 10.000 most frequent clonotypes in the pre- and post-transplant networks grouped by rejection"} -->
<!-- filename <- paste0("plot_cc_top10knet_tcopies") -->
<!-- ggarrange(p5, p6, ncol=2, nrow=1, common.legend = TRUE, legend="bottom") -->
<!-- ggsave(filename = paste(out.path,filename,".png", sep = ""), width = 20, height = 20, dpi = 640, units="cm", limitsize = F) -->
<!-- ``` -->

<!-- ```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap="Boxplot of the characteristic path length estimate of the top 10.000 clonotypes in the pre- and post-transplant networks grouped by rejection"} -->
<!-- filename <- paste0("plot_cpl_top10knet_tcopies") -->
<!-- ggarrange(p7, p8, ncol=2, nrow=1, common.legend = TRUE, legend="bottom") -->
<!-- ggsave(filename = paste(out.path,filename,".png", sep = ""), width = 20, height = 20, dpi = 640, units="cm", limitsize = F) -->

<!-- ``` -->



```{r echo=FALSE,  message=FALSE, warning=FALSE}

tbl_netmetrics <- CreateTableOne(colnames(tbl_CD4 )[15:ncol(tbl_CD4)], strata=c("stage"), data=tbl_CD4, test=F, smd=F)
tbl_netmetrics_mat <- data.frame(print(tbl_netmetrics,includeNA = TRUE, printToggle = FALSE))
rownames(tbl_netmetrics_mat) <- sapply(rownames(tbl_netmetrics_mat), function(x) strsplit(x," ")[[1]][1])
colnames(tbl_netmetrics_mat)[1:3] <- c("PreTX","PostTX","donor.reactive")
tbl_netmetrics_mat$Variable <- c("N", "Connected clones (%)", "Connected alloclones (%)","Clones in the network (%)", "Alloclones in the network (%)", "Median degree", "Mean degree", "Median degree (allo)", "Mean degree (allo)", "Mean CDR3 length", "Ratio of isolates and connected", "Ratio of isolates and connected (allo)", "Ratio of singletons and doubletons", "Ratio of nodes with degree>1 and doubletons", "Clustering coefficient", "Number of clusters", "Clones in LC", "Average number of clones in LC", "Median number of clones in LC", "Number of pairs", "Number of clusters larger than 2 clones", "Ratio between clusters >2 and =2", "Alloclones in LC (%)", "Newly detected clones in LC", "Number of clones in LC", "Characteristic path length 1", "Characteristic path length 2", "Global efficiency", "Edge density", "Edge density 2", "Modularity", "Assortativity", "Diameter", "Radius", 
"Gini(Degree)", "Gini(Cluster size)", "Gini(Betweenness)", "Gini(Local efficiency)")

tbl_netmetrics_mat %>%
  relocate(Variable, .before=PreTX) %>%
  dplyr::select(Variable, PreTX, PostTX, donor.reactive) %>%
  `colnames<-`(c("Variable","PreTX","PostTX","donor-reactive")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% 
  set_table_properties(layout = "autofit") %>%
  set_caption("Results of the network analysis for the CD4 samples", autonum = tbl_autonum)
  

```

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE}

tbl_netmetrics <- CreateTableOne(colnames(tbl_CD8 )[15:ncol(tbl_CD8)], strata=c("stage"), data=tbl_CD8, test=F, smd=F)
tbl_netmetrics_mat <- data.frame(print(tbl_netmetrics,includeNA = TRUE, printToggle = FALSE))
rownames(tbl_netmetrics_mat) <- sapply(rownames(tbl_netmetrics_mat), function(x) strsplit(x," ")[[1]][1])
colnames(tbl_netmetrics_mat)[1:3] <-c("PreTX","PostTX","donor.reactive")
tbl_netmetrics_mat$Variable <- c("N", "Connected clones (%)", "Connected alloclones (%)","Clones in the network (%)", "Alloclones in the network (%)", "Median degree", "Mean degree", "Median degree (allo)", "Mean degree (allo)", "Mean CDR3 length", "Ratio of isolates and connected", "Ratio of isolates and connected (allo)", "Ratio of singletons and doubletons", "Ratio of nodes with degree>1 and doubletons", "Clustering coefficient", "Number of clusters", "Clones in LC", "Average number of clones in LC", "Median number of clones in LC", "Number of pairs", "Number of clusters larger than 2 clones", "Ratio between clusters >2 and =2", "Alloclones in LC (%)", "Newly detected clones in LC", "Number of clones in LC", "Characteristic path length 1", "Characteristic path length 2", "Global efficiency", "Edge density", "Edge density 2","Modularity", "Assortativity", "Diameter", "Radius", 
"Gini(Degree)", "Gini(Cluster size)", "Gini(Betweenness)", "Gini(Local efficiency)")


tbl_netmetrics_mat %>%
  relocate(Variable, .before=PreTX) %>%
  dplyr::select(Variable, PreTX, PostTX, donor.reactive) %>%
  `colnames<-`(c("Variable","PreTX","PostTX","donor-reactive")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% 
  set_table_properties(layout = "autofit") %>%
  set_caption("Results of the network analysis for the CD8 samples", autonum = tbl_autonum)
  

```



<!-- # Networks of the donor-reactive clonotypes and their neighbors -->
<!-- Only vertices with an donor-reactive neighbor are contained in the network. Isolated vertices are removed from further analysis -->

<!-- Summary of the subject-specific network attributes of the CD4 TCR repertoire stratified by timepoint of sample acquisition -->
<!-- ```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, size='tiny'} -->
<!-- net.CD4 <- read.xlsx(paste0(folder_CD4_allo ,"table_tcopies_networks_CD4_full.xlsx")) -->
<!-- net.CD4$conn.clones.perc <- (net.CD4$clones.network/net.CD4$sample.clones)*100 -->
<!-- net.CD4$LC.allo <- round(net.CD4$LC.allo*100,2) -->
<!-- net.CD4$LC.newBX <- round(net.CD4$LC.newBX*100,2) -->

<!-- netmat.CD4 <- net.CD4[,c(1:15)] -->
<!-- colnames(netmat.CD4) <- c("Group","Phenotype","Stage","Rejection","Total clones", "Clones after pruning", "Clones in network", "Links in network", "Total allo", "Allo after pruning", "Allo in network", "Total isolates", "Allo isolates", "Clones in network (%)","Allo in network (%)") -->
<!-- netmat.CD4 <- netmat.CD4[order(netmat.CD4$Group),] -->
<!-- rownames(netmat.CD4) <- NULL -->


<!-- tbl.netmetrics <- CreateTableOne(colnames(net.CD4)[16:ncol(net.CD4)], strata=c("rejection", "stage"), data=net.CD4, test=F, smd=F) -->
<!-- tbl.netmetrics_mat <- print(tbl.netmetrics,includeNA = TRUE, printToggle = FALSE) -->
<!-- rownames(tbl.netmetrics_mat) <- sapply(rownames(tbl.netmetrics_mat), function(x) strsplit(x," ")[[1]][1]) -->
<!-- rownames(tbl.netmetrics_mat) <- c("Patients (n)","Median degree", "Mean degree", "Median degree (allo)", "Mean degree (allo)", "Mean CDR3 seq length", "Ratio of isolated and connected clones", "Ratio of isolated and connected allo clones", "Ratio of clones with degree=1 and degree=2", "Ratio of clones with degree>1 and degree=1", "Clustering coefficient", "Number of clusters", "Number of clones in largest cluster", "Mean number of clones in a cluster", "Median number of clones in a cluster", "Number of clusters consisting of 2 clones", "Number of clusters consisting of >2 clones", "Ratio of clusters of size > 2 and size=2", "Allo in largest cluster (%)", "New BX in largest cluster (%)", "Mean cell count of clones in largest cluster", "Characteristic path length (only existing paths)", "Characteristic path length (missing paths included)", "Global efficiency", "Edge density","Clones with LV of 1 (%)","Alloclones with LV of 1 (%)", "Connected clones (%)") -->
<!-- colnames(tbl.netmetrics_mat) <- rep(c("0","1"),2) -->

<!-- # netmat.CD4 %>% -->
<!-- #   flextable()  %>% -->
<!-- #     bold(part = "header") %>% -->
<!-- #   set_caption("test1", autonum = tbl_autonum) %>% -->
<!-- #   autofit() %>% -->
<!-- #   fit_to_width(7) -->
<!-- # -->
<!-- # tbl.netmetrics_mat %>% -->
<!-- #   data.frame() %>% -->
<!-- #   flextable()  %>% -->
<!-- #     bold(part = "header") %>% -->
<!-- #   set_caption("test2", autonum = tbl_autonum) %>% -->
<!-- #   autofit() %>% -->
<!-- #   fit_to_width(7) -->
<!-- ``` -->



## Networks of the tissue samples

First and last 3 amino acids were removed from the CDR3 aa sequences sequences. Each vertex in the network corresponds to an unique CDR3 amino acid sequence in an individual's repertoire. Vertices are connected by edges if they only differ by one amino acid in their CDR3 sequence (Levenshtein distance of one). Vertices without neighbours were removed from the visual graph representation.

```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE}
get_eff <- function(i,g){return((1/(length(V(g)) - 1))*sum(1/distances(g, V(g)[i])[-i]))}

# ----- R34
tissue_34$cdr3aa.short <- str_sub(tissue_34$cdr3aa, 4, -4)
load(file = paste0("TCopies_AA_merged.RData"))
blood_34_aa <- merged[["R34"]]
detected_aa <- str_sub(blood_34_aa$cdr3aa, 4, -4)

adjmat_34 <- stringdistmatrix(unique(tissue_34$cdr3aa.short),unique(tissue_34$cdr3aa.short), method="lv")
adjmat_34[adjmat_34!=1] <- 0
colnames(adjmat_34) <- rownames(adjmat_34) <- unique(tissue_34$cdr3aa.short)
net_34 <- graph_from_adjacency_matrix(adjmatrix = adjmat_34, diag=F, mode="undirected")

df.cl <- data.frame(table(clusters(net_34)$membership))
colnames(df.cl) <- c("cluster", "size")
df.cl$cluster <- as.factor(df.cl$cluster)
  
avg_degree_34 <- data.frame("cdr3aa.short"= rownames(adjmat_34),degree=apply(adjmat_34, 1, sum), 
                            cluster=as.factor(clusters(net_34)$membership))
avg_degree_34$new <- 0
avg_degree_34[!avg_degree_34$cdr3 %in% detected_aa,]$new <- 1
avg_degree_34 <- left_join(avg_degree_34,df.cl,by="cluster")

net_34 <- graph_from_adjacency_matrix(adjmatrix = adjmat_34, diag=F, mode="undirected")

tmp <- left_join(avg_degree_34, tissue_34[,c("cdr3aa.short", "allo", "cdr3aa","count","freq")], by="cdr3aa.short")
tmp.bc <- data.frame("cdr3aa.short"=names(betweenness(net_34)),"bc"=betweenness(net_34),"bc.norm"=betweenness(net_34, normalized = T))
tmp <- left_join(tmp, tmp.bc, by="cdr3aa.short")
tmp <- tmp[order(tmp$bc, decreasing = T),]


tmp_34 <- tmp %>%
  relocate(cdr3aa, .before=degree) %>%
  relocate(freq, .before=degree) %>%
  relocate(count, .before=degree) %>%
  relocate(allo, .before=degree) %>%
  relocate(new, .before=degree) %>%
  mutate(degree.norm=degree/size,
         freq=round(freq,3),
         bc.norm=round(bc.norm,3),
         degree.norm=round(degree.norm,3)) %>% 
  relocate(degree, .before=degree.norm) 

write.xlsx(tmp_34, file = paste0(out.path,"table_localfeatures_R34_tissue.xlsx"), overwrite=T)


cl.34 <- clusters(net_34)
me <- which(cl.34$csize == max(cl.34$csize))
res.34 <- unlist(split(names(cl.34$membership), cl.34$membership)[me])
largest.cluster_34 <- tmp_34[tmp_34$cdr3aa.short %in% res,]
```


```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE}
#------ R43
tissue_43$cdr3aa.short <- str_sub(tissue_43$cdr3aa, 4, -4)
load(file = paste0("TCopies_AA_merged.RData"))
blood_43_aa <- merged[["R43"]]
detected_aa <- str_sub(blood_43_aa$cdr3aa, 4, -4)

adjmat_43 <- stringdistmatrix(unique(tissue_43$cdr3aa.short),unique(tissue_43$cdr3aa.short), method="lv")
adjmat_43[adjmat_43!=1] <- 0
colnames(adjmat_43) <- rownames(adjmat_43) <- unique(tissue_43$cdr3aa.short)
net_43 <- graph_from_adjacency_matrix(adjmatrix = adjmat_43, diag=F, mode="undirected")

df.cl <- data.frame(table(clusters(net_43)$membership))
colnames(df.cl) <- c("cluster", "size")
df.cl$cluster <- as.factor(df.cl$cluster)
  
avg_degree_43 <- data.frame("cdr3aa.short"= rownames(adjmat_43),degree=apply(adjmat_43, 1, sum), 
                            cluster=as.factor(clusters(net_43)$membership))
avg_degree_43$new <- 0
avg_degree_43[!avg_degree_43$cdr3 %in% detected_aa,]$new <- 1
avg_degree_43 <- left_join(avg_degree_43,df.cl,by="cluster")

net_43 <- graph_from_adjacency_matrix(adjmatrix = adjmat_43, diag=F, mode="undirected")

tmp <- left_join(avg_degree_43, tissue_43[,c("cdr3aa.short", "allo", "cdr3aa","count","freq")], by="cdr3aa.short")
tmp.bc <- data.frame("cdr3aa.short"=names(betweenness(net_43)),"bc"=betweenness(net_43),"bc.norm"=betweenness(net_43, normalized = T))
tmp <- left_join(tmp, tmp.bc, by="cdr3aa.short")
tmp <- tmp[order(tmp$bc, decreasing = T),]

tmp_43 <- tmp %>%
  relocate(cdr3aa, .before=degree) %>%
  relocate(freq, .before=degree) %>%
  relocate(count, .before=degree) %>%
  relocate(allo, .before=degree) %>%
  relocate(new, .before=degree) %>%
  mutate(degree.norm=degree/size,
         freq=round(freq,3),
         bc.norm=round(bc.norm,3),
         degree.norm=round(degree.norm,3)) %>% 
  relocate(degree, .before=degree.norm) 

write.xlsx(tmp_43, file = paste0(out.path,"table_localfeatures_R43_tissue.xlsx"), overwrite=T)

  
cl.43 <- clusters(net_43)
me <- which(cl.43$csize == max(cl.43$csize))
res.43 <- unlist(split(names(cl.43$membership), cl.43$membership)[me])
largest.cluster_43<- tmp_43[tmp_43$cdr3aa.short %in% res,]
```

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.show="hold", out.width="50%", fig.height=6}
comp_allo_34 <- print(CreateTableOne(c("bc","degree"),  argsNonNormal = c("bc","degree"),strata="allo", data=tmp_34[tmp_34$size>1,]), printToggle=F)

comp_allo_34 %>%
  data.frame() %>% 
  `rownames<-`(c("N", "Betweenness centrality",  "Degree")) %>%
  flextable()  %>% 
  bold(part = "header") %>% 
  set_caption("Comparison of non- and donor-reactive clonotypes within the immune repertoire network of R34", autonum = tbl_autonum) %>% 
  autofit() %>% 
  fit_to_width(7) 

comp_allo_43 <- print(CreateTableOne(c("bc","degree"), argsNonNormal = c("bc","degree")  , strata="allo", data=tmp_43[tmp_43$size>2,]), printToggle=F)

comp_allo_43 %>%
  data.frame() %>% 
  `rownames<-`(c("N", "Betweenness centrality",  "Degree")) %>%
  flextable()  %>% 
  bold(part = "header") %>% 
  set_caption("Comparison of non- and donor-reactive clonotypes within the immune repertoire network of R34", autonum = tbl_autonum) %>% 
  autofit() %>% 
  fit_to_width(7) 

```

```{r echo=FALSE, cache=T,message=FALSE, warning=FALSE, fig.align="center", fig.show="hold", out.width="50%", fig.height=6}
ratio.con34 <- sum(igraph::degree(net_34)>1)/sum(igraph::degree(net_34)==1)
ratio.con43 <- sum(igraph::degree(net_43)>1)/sum(igraph::degree(net_43)==1)
plan(multisession, workers=10)
local.eff.34 <- future_sapply(1:length(V(net_34)), function(i) get_eff(i,g=net_34))
local.eff.43 <- future_sapply(1:length(V(net_43)), function(i) get_eff(i,g=net_43))
plan(sequential)

res <- data.frame("R34"=c(nrow(tissue_34), sum(tissue_34$count), sum(igraph::degree(net_34)>0)/nrow(tissue_34)*100,
                          sum(tissue_34$allo==1),
                          sum(tissue_34[tissue_34$cdr3aa.short %in% names(which(igraph::degree(net_34)>0)),]$allo),
                          transitivity(net_34,isolates = "NaN"),
                          mean_distance(net_34, directed = F, unconnected = TRUE), ratio.con34, edge_density(net_34),
                          modularity(net_34, cl.34$membership),assortativity.degree(net_34), 
                          Gini(igraph::degree(net_34)), Gini(cl.34$csize), Gini(betweenness(net_34)), Gini(local.eff.43)),
                  "R43"=c(nrow(tissue_43), sum(tissue_43$count),sum(igraph::degree(net_43)>0)/nrow(tissue_43)*100, 
                          sum(tissue_43$allo==1),
                          sum(tissue_43[tissue_43$cdr3aa.short %in% names(which(igraph::degree(net_43)>0)),]$allo),
                          transitivity(net_43,isolates = "NaN"),
                          mean_distance(net_43, directed = F, unconnected = TRUE), ratio.con43, edge_density(net_43),
                          modularity(net_43, cl.43$membership), assortativity.degree(net_43),
                          Gini(igraph::degree(net_43)),Gini(cl.43$csize), Gini(betweenness(net_43)), Gini(local.eff.43)))
res$Variable <- c("Clonotypes", "Clones", "Connected clonotypes (%)",
                  "Alloclonotypes", "Connected alloclonotypes (%)",
                  "Clustering coefficient", "Characteristic path length", 
                  "Ratio: degree>1/degree==1", "Edge density","Modularity", "Assortativity", 
                  "Gini(Degree)", "Gini(Cluster size)", "Gini(Betweenness)", "Gini(Local eff.)")

res$R34 <- round(res$R34, 3)
res$R43 <- round(res$R43, 3)

res %>%
  relocate(Variable, .before=R34) %>%
  flextable() %>%
  autofit() %>%
  set_caption("Network results of the tissue samples", autonum = tbl_autonum) %>%
  bold(part = "header") %>%
  colformat_num(digits=2)  %>%
  fit_to_width(7) %>%set_table_properties(layout = "autofit")
```

```{r echo=FALSE, results='hide',cache=T,message=FALSE, warning=FALSE, fig.align="center", fig.show="hold", out.width="50%", fig.height=6}
# Plot the networks
net_34 <- delete.vertices(net_34, names(which(igraph::degree(net_34) == 0)))
E(net_34)$color <- 'black'
V(net_34)$color <- "grey"
V(net_34)[tmp_34$cdr3aa.short[1]]$color<-"red"
vertex.color <- data.frame(clone=names(V(net_34)),group=tissue_34[which(names(V(net_34)) %in% tissue_34$cdr3aa.short),]$allo)

l <- layout_with_fr(net_34,dim=3)
plot(net_34, vertex.label=NA, remove.multiple = T, edge.arrow.size=0.1, vertex.size=3, vertex.color = as.factor(vertex.color$group),palette=c("grey","red"),
       layout=l, main="R34")

filename <- paste0("plot_network_R34_tissue")
png(file = paste(out.path, filename,".png", sep = ""), width = 8*600, height = 6*600, units = "px", res = 600)
plot(net_34, vertex.label=NA, remove.multiple = T, edge.arrow.size=0.1, vertex.size=3, vertex.color = as.factor(vertex.color$group),palette=c("grey","red"),
       layout=l, main="R34")
dev.off()

cl.34 <- clusters(net_34)
me <- which(cl.34$csize == max(cl.34$csize))
res.34 <- unlist(split(names(cl.34$membership), cl.34$membership)[me])
largest.cluster_34 <- tmp_34[tmp_34$cdr3aa.short %in% res,]

g34 <- delete_vertices(net_34, V(net_34)[!(cl.34$membership %in% which.max(cl.34$csize))])
l <- layout_with_fr(g34,dim=3)
plot(g34, vertex.label=NA, remove.multiple = T, edge.arrow.size=0.1,
     vertex.size=3,palette=c("grey","red"),layout=l,main="R34")

filename <- paste0("plot_network_R34_tissue_2")
png(file = paste(out.path, filename,".png", sep = ""), width = 8*600, height = 6*600, units = "px", res = 600)
plot(g34, vertex.label=NA, remove.multiple = T, edge.arrow.size=0.1,
     vertex.size=3,palette=c("grey","red"),layout=l,main="R34")
dev.off()

# Plot largest cluster only
cl.34 <- clusters(net_34)
me <- which(cl.34$csize == max(cl.34$csize))
res.34 <- unlist(split(names(cl.34$membership), cl.34$membership)[me])
largest.cluster_34 <- tmp_34[tmp_34$cdr3aa.short %in% res,]

V(net_34)$size <- 2
V(net_34)[tmp_34$cdr3aa.short[1]]$size <- 5
g34 <- delete_vertices(net_34, V(net_34)[!(cl.34$membership %in% which.max(cl.34$csize))])
l <- layout_with_kk(g34)
plot(g34, vertex.label=NA, remove.multiple = T, edge.arrow.size=0.1,
     palette=c("grey","red"),layout=l,main="R34")

filename <- paste0("plot_network_R34_tissue_2")
png(file = paste(out.path, filename,".png", sep = ""), width = 8*600, height = 6*600, units = "px", res = 600)
plot(g34, vertex.label=NA, remove.multiple = T, edge.arrow.size=0.1,
     palette=c("grey","red"),layout=l,main="R34")
dev.off()

#R43 --------------------
net_43 <- delete.vertices(net_43, names(which(igraph::degree(net_43) == 0)))
E(net_43)$color <- 'black'
V(net_43)$color <- "grey"
V(net_43)[tmp_43$cdr3aa.short[1]]$color<-"red"
V(net_43)[tmp_43$cdr3aa.short[1]]$size <- 5

vertex.color <- data.frame(clone=names(V(net_43)),group=tissue_43[which(names(V(net_43)) %in% tissue_43$cdr3aa.short),]$allo)

l <- layout_with_fr(net_43,dim=3)
plot(net_43, vertex.label=NA, remove.multiple = T, edge.arrow.size=0.1, vertex.size=3, vertex.color = as.factor(vertex.color$group),palette=c("grey","red"),
       layout=l, main="R43")

filename <- paste0("plot_network_R43_tissue")
png(file = paste(out.path, filename,".png", sep = ""), width = 8*600, height = 6*600, units = "px", res = 600)
plot(net_43, vertex.label=NA, remove.multiple = T, edge.arrow.size=0.1, vertex.size=3, vertex.color = as.factor(vertex.color$group),palette=c("grey","red"),
       layout=l, main="R43")
dev.off()

# Plot largest cluster only
cl.43 <- clusters(net_43)
me <- which(cl.43$csize == max(cl.43$csize))
res.43 <- unlist(split(names(cl.43$membership), cl.43$membership)[me])
largest.cluster_43 <- tmp_43[tmp_43$cdr3aa.short %in% res,]

V(net_43)$size <- 2
V(net_43)[tmp_43$cdr3aa.short[1]]$size <- 5
g43 <- delete_vertices(net_43, V(net_43)[!(cl.43$membership %in% which.max(cl.43$csize))])
l <- layout_with_kk(g43)
plot(g43, vertex.label=NA, remove.multiple = T, edge.arrow.size=0.1,
     palette=c("grey","red"),layout=l,main="R43")

filename <- paste0("plot_network_R43_tissue_2")
png(file = paste(out.path, filename,".png", sep = ""), width = 8*600, height = 6*600, units = "px", res = 600)
plot(g43, vertex.label=NA, remove.multiple = T, edge.arrow.size=0.1,
     palette=c("grey","red"),layout=l,main="R43")
dev.off()


```


## Levenshtein distance of tissue clonotypes and donor-reactive clonotypes
For the clonotypes within the tissue samples, the Levenhsteint distance of each to a clonotype included in either the CD4 or the CD8 MLR samples was assessed. The percentage of clonotypes from the tissue samples which are connected to at least one clonotype within the MLR samples (CD4 and CD8) is computed. For the MLR samples, only clonotypes were considered which were identified as alloreactive according to the 5-fold expansion criterion.

```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE}
# df.34.low$cdr3aa.short <- str_sub(df.34.low$cdr3aa, 4, -4)
# tissue_34$cdr3aa.short <- str_sub(tissue_34$cdr3aa, 4, -4)
# res.simallo <- list()
# 
# tmp<- stringdistmatrix(tissue_34$cdr3aa.short,df.34.low[df.34.low$allo==1,]$cdr3aa.short, method="lv")
# tmp[tmp!=1] <- 0
# tmp.sim.34 <- data.frame("cdr3aa.short"=tissue_34$cdr3aa.short, "cdr3aa"=tissue_34$cdr3aa, "allo"=tissue_34$allo,"NR.eq"=apply(tmp, 1, sum))
# 
# clts.similarallo <- sum(tmp.sim.34[tmp.sim.34$allo==0,]$NR.eq >0)/nrow(tmp.sim.34[tmp.sim.34$allo==0,])*100
# vec.simallo.34 <- c(nrow(tissue_34),nrow(df.34.low[df.34.low$allo==1,]), nrow(tmp.sim.34[tmp.sim.34$allo==0,]), sum(tmp.sim.34[tmp.sim.34$allo==0,]$NR.eq >0), clts.similarallo)
# 
# #-----------------------------
# df.43.low$cdr3aa.short <- str_sub(df.43.low$cdr3aa, 4, -4)
# tissue_43$cdr3aa.short <- str_sub(tissue_43$cdr3aa, 4, -4)
# 
# try_LV <- function(x){
#   mat <- stringdist(x,df.43.low[df.43.low$allo==1,]$cdr3aa.short, method="lv")
#   mat[mat!=1] <- 0
#   return(sum(mat))
# }
# 
# plan(multisession, workers=10)
# tmp <- future_sapply(tissue_43$cdr3aa.short, function(x) try_LV(x))
# plan(sequential)
# 
# tmp.sim.43 <- data.frame("cdr3aa.short"=tissue_43$cdr3aa.short, "cdr3aa"=tissue_43$cdr3aa, "allo"=tissue_43$allo,"NR.eq"=tmp)
# 
# clts.similarallo <- sum(tmp.sim.43[tmp.sim.43$allo==0,]$NR.eq >0)/nrow(tmp.sim.43[tmp.sim.43$allo==0,])*100
# vec.simallo.43 <- c(nrow(tissue_43),nrow(df.43.low[df.43.low$allo==1,]), nrow(tmp.sim.43[tmp.sim.43$allo==0,]), sum(tmp.sim.43[tmp.sim.43$allo==0,]$NR.eq >0), clts.similarallo)
# 
# write.xlsx(list("R34"=tmp.sim.34, "R43"=tmp.sim.43), paste0(out.path, "table_tcopies_LV1_MLRvsTissue.xlsx"), overwrite=T)
# 
# df.simallo <- data.frame("Variable"=c("Clts.tissue", "Clts.MLRallo", "Clts.tissue.nonallo", "Clts.nonallo.LV1", "Perc.nonallo.LV1"), "R34"=vec.simallo.34, "R43"=vec.simallo.43)
# write.xlsx(df.simallo, paste0(out.path, "table_tcopies_tiss_simallo.xlsx"), overwrite=T)
df.simallo <- read.xlsx(paste0(out.path, "table_tcopies_tiss_simallo.xlsx"))
df.simallo <- df.simallo %>%
  mutate(R34=round(R34,2),
         R43=round(R43,2))

df.simallo %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("How similar are non-allo in the tissue to alloreactive clonotypes?", autonum = tbl_autonum)

```

```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE}

compute_vec <- function(i, v, threshold=1){
  res.LV1 = sum(stringdist(i, v, method = "lv") == threshold)

  return(res.LV1)
}

compute_lv1 <- function(v1, v2, workers = 5) {
  n1 = nrow(v1) 
  n2 = nrow(v2) 
  v1.short <- str_sub(v1$cdr3aa, 4, -4)
  v2.short <- str_sub(v2, 4, -4)
  
  plan(multisession, workers = 5)  
  res <- future_lapply(v1.short, function(i) compute_vec(i, v=v2.short))
  plan(sequential)
  
  res.full <- list("cdr3aa"=v1$cdr3aa, "cdr3aa.short"=v1.short, "allo"=v1$allo,"LV1"=res)
  return(res.full)  
}

#----------------------------------

# res.34.lv1.bl <- data.frame(do.call(cbind,compute_lv1(v1=df.34.BL[,c("cdr3aa","allo")],v2=df.34.low[df.34.low$allo==1,]$cdr3aa)))
# res.43.lv1.bl <- data.frame(do.call(cbind,compute_lv1(v1=df.43.BL[,c("cdr3aa","allo")],v2=df.43.low[df.43.low$allo==1,]$cdr3aa)))
# res.34.lv1.late <- data.frame(do.call(cbind,compute_lv1(v1=df.34.late[,c("cdr3aa","allo")],v2=df.34.low[df.34.low$allo==1,]$cdr3aa)))
# res.43.lv1.late <- data.frame(do.call(cbind,compute_lv1(v1=df.43.late[,c("cdr3aa","allo")],v2=df.43.low[df.43.low$allo==1,]$cdr3aa)))
# res.34.lv1.tissue <- data.frame(do.call(cbind,compute_lv1(v1=tissue_34[,c("cdr3aa","allo")],v2=df.34.low[df.34.low$allo==1,]$cdr3aa)))
# res.43.lv1.tissue <- data.frame(do.call(cbind,compute_lv1(v1=tissue_43[,c("cdr3aa","allo")],v2=df.43.low[df.43.low$allo==1,]$cdr3aa)))
# 
# 
# res.all <- list("R34.BL"=res.34.lv1.bl, "R34.late"=res.34.lv1.late,  "R34.tissue"=res.34.lv1.tissue,
#                 "R43.BL"=res.43.lv1.bl, "R43.late"=res.43.lv1.late,  "R43.tissue"=res.43.lv1.tissue)
# write.xlsx(res.all, paste0(out.path, "table_tcopies_LV1_MLRvsPeriph.xlsx"), overwrite=T)

# res.all <- list("R34.BL"=read.xlsx(paste0(out.path, "table_tcopies_LV1_MLRvsPeriph.xlsx"), sheet=1), 
#                 "R34.late"=read.xlsx(paste0(out.path, "table_tcopies_LV1_MLRvsPeriph.xlsx"), sheet=2),
#                 "R34.tissue"=read.xlsx(paste0(out.path, "table_tcopies_LV1_MLRvsPeriph.xlsx"), sheet=3),
#                  "R43.BL"=read.xlsx(paste0(out.path, "table_tcopies_LV1_MLRvsPeriph.xlsx"), sheet=4), 
#                 "R43.late"=read.xlsx(paste0(out.path, "table_tcopies_LV1_MLRvsPeriph.xlsx"), sheet=5),  
#                 "R43.tissue"=read.xlsx(paste0(out.path, "table_tcopies_LV1_MLRvsPeriph.xlsx"), sheet=6))
# 
# res.lv1 <- sapply(res.all, function(x) c(nrow(x),sum(x$allo==0),sum(x[x$allo==0,]$LV1 >0),
#                                          sum(x[x$allo==0,]$LV1>0)/nrow(x[x$allo==0,])*100))
# res <- data.frame(rep(c("PreTX","PostTX","Tissue"),2),
#                   c(rep(sum(df.34.low$allo==1),3),rep(sum(df.43.low$allo==1),3)),
#                     res.lv1[1,], res.lv1[2,], res.lv1[3,], res.lv1[4,])
# colnames(res) <- c("Time", "MLR clts allo", "Clts","Clts non-allo", "Clts non-allo+LV1", "Clts non-allo+LV1 (%)")
# write.xlsx(res, paste0(out.path, "table_tcopies_LV1_MLRvsPeriph_results.xlsx"), overwrite=T)

res <- read.xlsx(paste0(out.path, "table_tcopies_LV1_MLRvsPeriph_results.xlsx"))

res %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("How similar are non-allo in the periphery and allograft compared to alloreactive clonotypes in terms of CDR3 AA sequence?", autonum = tbl_autonum)

```

Switch the MLR repertoire for R34 and R43

```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE}

compute_vec <- function(i, v, threshold=1){
  res.LV1 = sum(stringdist(i, v, method = "lv") == threshold)

  return(res.LV1)
}

compute_lv1 <- function(v1, v2, workers = 5) {
  n1 = nrow(v1) 
  n2 = nrow(v2) 
  v1.short <- str_sub(v1$cdr3aa, 4, -4)
  v2.short <- str_sub(v2, 4, -4)
  
  plan(multisession, workers = 5)  
  res <- future_lapply(v1.short, function(i) compute_vec(i, v=v2.short))
  plan(sequential)
  
  res.full <- list("cdr3aa"=v1$cdr3aa, "cdr3aa.short"=v1.short, "allo"=v1$allo,"LV1"=res)
  return(res.full)  
}

#----------------------------------
# 
# res.34.lv1.bl <- data.frame(do.call(cbind,compute_lv1(v1=df.34.BL[,c("cdr3aa","allo")],v2=df.43.low[df.43.low$allo==1,]$cdr3aa)))
# res.43.lv1.bl <- data.frame(do.call(cbind,compute_lv1(v1=df.43.BL[,c("cdr3aa","allo")],v2=df.34.low[df.34.low$allo==1,]$cdr3aa)))
# res.34.lv1.late <- data.frame(do.call(cbind,compute_lv1(v1=df.34.late[,c("cdr3aa","allo")],v2=df.43.low[df.43.low$allo==1,]$cdr3aa)))
# res.43.lv1.late <- data.frame(do.call(cbind,compute_lv1(v1=df.43.late[,c("cdr3aa","allo")],v2=df.34.low[df.34.low$allo==1,]$cdr3aa)))
# res.34.lv1.tissue <- data.frame(do.call(cbind,compute_lv1(v1=tissue_34[,c("cdr3aa","allo")],v2=df.43.low[df.43.low$allo==1,]$cdr3aa)))
# res.43.lv1.tissue <- data.frame(do.call(cbind,compute_lv1(v1=tissue_43[,c("cdr3aa","allo")],v2=df.34.low[df.34.low$allo==1,]$cdr3aa)))
# 
# 
# res.all <- list("R34.BL"=res.34.lv1.bl, "R34.late"=res.34.lv1.late,  "R34.tissue"=res.34.lv1.tissue,
#                 "R43.BL"=res.43.lv1.bl, "R43.late"=res.43.lv1.late,  "R43.tissue"=res.43.lv1.tissue)
# write.xlsx(res.all, paste0(out.path, "table_tcopies_LV1_MLRvsPeriph_switched.xlsx"), overwrite=T)
# 
# res.all <- list("R34.BL"=read.xlsx(paste0(out.path, "table_tcopies_LV1_MLRvsPeriph_switched.xlsx"), sheet=1),
#                 "R34.late"=read.xlsx(paste0(out.path, "table_tcopies_LV1_MLRvsPeriph_switched.xlsx"), sheet=2),
#                 "R34.tissue"=read.xlsx(paste0(out.path, "table_tcopies_LV1_MLRvsPeriph_switched.xlsx"), sheet=3),
#                  "R43.BL"=read.xlsx(paste0(out.path, "table_tcopies_LV1_MLRvsPeriph_switched.xlsx"), sheet=4),
#                 "R43.late"=read.xlsx(paste0(out.path, "table_tcopies_LV1_MLRvsPeriph_switched.xlsx"), sheet=5),
#                 "R43.tissue"=read.xlsx(paste0(out.path, "table_tcopies_LV1_MLRvsPeriph_switched.xlsx"), sheet=6))
# 
# res.lv1 <- sapply(res.all, function(x) c(nrow(x),sum(x$allo==0),sum(x[x$allo==0,]$LV1 >0),
#                                          sum(x[x$allo==0,]$LV1>0)/nrow(x[x$allo==0,])*100))
# res <- data.frame(rep(c("PreTX","PostTX","Tissue"),2),
#                   c(rep(sum(df.43.low$allo==1),3),rep(sum(df.34.low$allo==1),3)),
#                     res.lv1[1,], res.lv1[2,], res.lv1[3,], res.lv1[4,])
# colnames(res) <- c("Time", "MLR clts allo", "Clts","Clts non-allo", "Clts non-allo+LV1", "Clts non-allo+LV1 (%)")
# write.xlsx(res, paste0(out.path, "table_tcopies_LV1_MLRvsPeriph_switched_results.xlsx"), overwrite=T)

res <- read.xlsx(paste0(out.path, "table_tcopies_LV1_MLRvsPeriph_switched_results.xlsx"))

res %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("How similar are non-allo in the periphery and allograft compared to alloreactive clonotypes in terms of CDR3 AA sequence? - switched alloclone repertoires for R34 and R43
              ", autonum = tbl_autonum)

```




# Analysis of the Weinberger data

## Descriptive Analysis
```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=4, fig.cap="Absolute number of unique clonotypes grouped by phenotype groups TRBA and TRBB for the blood and biopsy repertoire samples"}
get_info <- function(filename){
  # filename <- files.wein[1]
  # filename <- files.wein[91]

  if(str_sub(filename, start = 5, end=5)=="B"){
    patID <- paste0("P",str_sub(filename, start = 8, end=9))
    stage <- str_sub(filename, start = 5, end=5)
    primer <- str_sub(filename, start = 10, end=13)
  }else{
    patID <- paste0("P",str_sub(filename, start = 7, end=8))
    stage <- str_sub(filename, start = 5, end=6)
    primer <- str_sub(filename, start = 9, end=12)
  }

  rejection <- ifelse(patID %in% wein.TCMGRset, 1,  0)
  
  info <- c(filename=filename,patID=patID, stage=stage, primer=primer, rejection=rejection)
}


descriptive_analysis <- function(filename){
  tmp <- read.table(paste0(wein.path,filename), header=T, stringsAsFactors = F)
  file.info <- get_info(filename)
  
  unique.clones <- length(unique(tmp$cdr3nt))
  counts <- sum(tmp$count)
  larg.clone <- max(tmp$count)
  frac.singletons <- sum(tmp$count==1)/counts

  return(c(file.info[2:5], unique.clones, counts, larg.clone, frac.singletons))
}

files.wein <- grep(list.files(wein.path), pattern=".txt",value=T)
files.wein <- files.wein[!(grepl("Rep", files.wein))]

df_da <- data.frame(t(sapply(files.wein, function(filename) descriptive_analysis(filename))))
df_da$V8 <- round(as.numeric(df_da$V8),4)
colnames(df_da) <- c("patID", "stage", "primer", "rejection", "clonotypes", "total count", "largest count", "fraction of singletons")
rownames(df_da) <- NULL

write.xlsx(df_uc, paste0(out.path, "table_weinberger_descrstats.xlsx"), overwrite=T)

```


## Unique clone counts
```{r echo=FALSE, cache=T,  message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=4, fig.cap="Absolute number of unique clonotypes grouped by phenotype groups TRBA and TRBB for the blood and biopsy repertoire samples"}


get_uc <- function(filename){
   
  #print(filename)
  tmp <- read.table(paste0(wein.path,filename), header=T, stringsAsFactors = F)
  unique.clones <- length(unique(tmp$cdr3nt))
  unique.allo <- length(unique(tmp[tmp$allo==1,]$cdr3nt))
  file.info <- get_info(filename)
  
  return(c(file.info[2:5], unique.clones, unique.allo))
}

files.wein <- grep(list.files(wein.path), pattern=".txt",value=T)
files.wein <- files.wein[!(grepl("Rep", files.wein))]

df_uc <- data.frame(t(sapply(files.wein, function(filename) get_uc(filename))))
df_uc$V5 <- as.numeric(df_uc$V5)
df_uc[df_uc$stage == "B",]$stage <- "Blood"
df_uc[df_uc$stage == "Tx",]$stage <- "Biopsy"

df_uc$tmp <- paste0(df_uc$primer," " ,df_uc$stage)
df_uc$tmp <- factor(df_uc$tmp,levels = c("TRBA Blood", "TRBB Blood", "TRBA Biopsy", "TRBB Biopsy"))
df_uc$stage <- factor(df_uc$stage, levels=c("Blood","Biopsy"))
write.xlsx(df_uc, paste0(out.path, "table_weinberger_unique_clones.xlsx"), overwrite=T)

df_uc <- read.xlsx(paste0(out.path, "table_weinberger_unique_clones.xlsx"))
df_uc$stage <- factor(df_uc$stage, levels=c("Blood","Biopsy"))


my_cols <-brewer.pal(8, "Paired")[c(1,2)]
p1 <- ggplot(df_uc, aes(x = stage, y = V5)) +
  stat_boxplot(geom ='errorbar', width = 0.2) +
  geom_boxplot(aes(x=stage, y=V5, fill=stage),width=0.5) +
  geom_dotplot(binaxis='y', stackdir='center', position=position_dodge(1), dotsize=0.3) +
  scale_fill_manual(values=my_cols) +
  scale_y_continuous(expand=c(0,0),"Unique clones",limits=c(0,50000), breaks = seq(0,50000, 10000)) +
  scale_x_discrete(expand=c(0,0),"",guide = guide_axis(angle = 45)) +
  theme_bw() + 
  theme(legend.position = "none", text = element_text(size=11)) +
  facet_wrap(~primer)

filename <- paste0("plot_uniqueclones_weinberger")
p1
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p1, width = 20, height = 20, dpi = 640, units="cm", limitsize = F)

```

## Diversity analysis
```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE}



calc_diversity <- function(filename, path=wein.path){

  sample <- read.table(paste0(path,filename), header=T, stringsAsFactors = F)
  fileinfo <- get_info(filename)

  
  reads <- sum(sample$count)
  largest.clone.freq <- max(sample$freq)
  largest.clone.count <- max(sample$count)
  sr.nuc <- length(unique(sample$cdr3nt))
  sr.aa <- length(unique(sample$cdr3aa))
  
  # RX diversity
  R10 <- min(which(cumsum(sample$freq)>=0.10))/nrow(sample)
  R20 <- min(which(cumsum(sample$freq)>=0.20))/nrow(sample)
  R50 <- min(which(cumsum(sample$freq)>=0.50))/nrow(sample)

  sample.allo <- sample[sample$allo==1,]
  sample.allo$freq.allo <- sample.allo$count/sum(sample.allo$count)
  R10allo <- min(which(cumsum(sample.allo$freq.allo)>=0.10))/nrow(sample.allo)
  R20allo <- min(which(cumsum(sample.allo$freq.allo)>=0.20))/nrow(sample.allo)
  R50allo <- min(which(cumsum(sample.allo$freq.allo)>=0.50))/nrow(sample.allo)
  
  Hobs <- -sum(sample$freq*log(sample$freq,2)) #Shannon
  Hmax <- log(nrow(sample),2)
  pielou <- Hobs/Hmax
  clonality <- 1- pielou
  si <- sum(sample$freq^2)
  invsi <- 1/si
 # ginic <- Gini(sample$count)
  bpi <- max(sample$freq)
  invbpi <- 1/bpi

  # The clonotype abundance frequency distribution  (x=frequency, y=number of clones with frequency x)
  tbl.counts <- table(sample$count)
  threshold.counts <- as.numeric(names(tbl.counts[tbl.counts == 1][2]))
  sample.bulk <- sample[sample$count < threshold.counts,]
  clone_nr <- c(table(sample.bulk$freq))
  freq <- as.numeric(as.character(names(clone_nr)))
  names(clone_nr) <- NULL
  if(all(is.na(log(freq))) | all(is.na(log(clone_nr)))){
    slope<-NA
  }else{
    slope <- as.numeric(abs(lm(log(clone_nr)~log(freq))$coefficients[2]))
  }

  # Non-parametric estimators of species richness
  df <- cbind(ceiling(as.numeric(names(table(sample$count)))),table(sample$count))
  chao1 <- chao1984(df, conf = 0.95)
  ACE_t10 <- ChaoLee1992(df, t = 10, method = "ACE",conf = 0.95)
  ACEbc_t10 <- ChaoLee1992(df, t = 10, method = "ACE-1",conf = 0.95)

  repstats <- c(c(fileinfo), round(c(Reads=reads, largest.clone.freq=largest.clone.freq, largest.clone.count=largest.clone.count,
                                     SpeciesRichness.Nuc=sr.nuc, SpeciesRichness.AA=sr.aa,
                                     R10=R10, R10allo=R10allo, R20=R20,R20allo=R20allo, R50=R50, R50allo=R50allo,
                                     ShannonE=Hobs, maxH=Hmax,PielouDiversity = pielou, Clonality=clonality,
                                     SimpsonIndex=si, invSimpsonIndex=invsi,
                                     BergerParkerIndex=bpi, invBergerParkerIndex=invbpi, pl.slope=slope,
                                     Chao1.est=chao1$Nhat, Chao1.se=chao1$SE,
                                     ACEbc.t10.est=ACEbc_t10$Nhat, ACEbc.t10.se=ACEbc_t10$SE),5))
}


files.wein <- grep(list.files(wein.path), pattern=".txt",value=T)
files.wein <- files.wein[!(grepl("Rep", files.wein))]

wein.stats <- data.frame(t(sapply(files.wein, calc_diversity, USE.NAMES = F)))
wein.stats[wein.stats$rejection==0,]$rejection <- "Controls"
wein.stats[wein.stats$rejection==1,]$rejection <- "Rejectors"
wein.stats[,6:29] <- apply(wein.stats[,6:29],2,as.numeric)
wein.stats[duplicated(wein.stats[,c("patID","stage","primer")]),]$patID <- paste0(wein.stats[duplicated(wein.stats[,c("patID","stage","primer")]),]$patID ,"-2")
write.xlsx(wein.stats, paste0(out.path, "table_weinberger_diversity.xlsx"), overwrite=T)

```


```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE}
wein.stats <- read.xlsx(paste0(out.path, "table_weinberger_diversity.xlsx"))
wein.stats$stage <- ifelse(wein.stats$stage %in% "B", "Blood", "Biopsy")
wein.stats$stage <- factor(wein.stats$stage, levels = c("Blood", "Biopsy"))
wein.stats[,6:29] <- apply(wein.stats[,6:29],2,as.numeric)


#------- TRBA
tmp.TRBA <- wein.stats[wein.stats$primer%in% "TRBA",]
p1 <-  ggplot(tmp.TRBA, aes(x=stage, y=R20, fill=rejection)) +
  geom_boxplot(position=position_dodge(0.8), notch=F, width = .4, alpha=0.8) +
  geom_dotplot(binaxis = "y", stackdir = "center",  position=position_dodge(0.8), dotsize=0.45, alpha=1) +
  stat_boxplot(geom ='errorbar', width = 0.2, position=position_dodge(0.8), alpha=0.8) +
  scale_y_continuous(expand=c(0,0),"R20", limits=c(0,0.075)) +
  scale_x_discrete(expand=c(0,0),"", labels = c("Blood", "Biopsy")) +
  scale_fill_brewer(palette="Dark2",name = "", labels = c("Controls", "Rejectors")) +
  ggtitle("TRBA") +
  theme_bw() +
  theme(legend.position="none", plot.title = element_text(hjust = 0.5), text = element_text(size=14))


p2 <- ggplot(tmp.TRBA, aes(x=stage, y=Clonality, fill=rejection)) +
  geom_boxplot(position=position_dodge(0.8), notch=F, width = .4, alpha=0.8) +
  geom_dotplot(binaxis = "y", stackdir = "center",  position=position_dodge(0.8), dotsize=0.45, alpha=1) +
  stat_boxplot(geom ='errorbar', width = 0.2, position=position_dodge(0.8), alpha=0.8) +
  scale_y_continuous(expand=c(0,0),"Clonality", limits = c(0,0.55)) +
  scale_x_discrete(expand=c(0,0),"", labels = c("Blood", "Biopsy")) +
  scale_fill_brewer(palette="Dark2",name = "", labels = c("Controls", "Rejectors")) +
  ggtitle("TRBA") +
  theme_bw() +
  theme(legend.position="none", plot.title = element_text(hjust = 0.5), text = element_text(size=14))



# ---- TRBB
tmp.TRBB <- wein.stats[wein.stats$primer%in%"TRBB",]
p3 <-  ggplot(tmp.TRBB, aes(x=stage, y=R20, fill=rejection)) +
  geom_boxplot(position=position_dodge(0.8), notch=F, width = .4, alpha=0.8) +
  geom_dotplot(binaxis = "y", stackdir = "center",  position=position_dodge(0.8), dotsize=0.45, alpha=1) +
  stat_boxplot(geom ='errorbar', width = 0.2, position=position_dodge(0.8), alpha=0.8) +
  scale_y_continuous(expand=c(0,0),"", limits=c(0,0.0075)) +
  scale_x_discrete(expand=c(0,0),"", labels = c("Blood", "Biopsy")) +
  scale_fill_brewer(palette="Dark2",name = "", labels = c("Controls", "Rejectors")) +
  ggtitle("TRBB") +
  theme_bw() +
  theme(legend.position="none", plot.title = element_text(hjust = 0.5), text = element_text(size=14))


p4 <- ggplot(tmp.TRBB, aes(x=stage, y=Clonality, fill=rejection)) +
  geom_boxplot(position=position_dodge(0.8), notch=F, width = .4, alpha=0.8) +
  geom_dotplot(binaxis = "y", stackdir = "center",  position=position_dodge(0.8), dotsize=0.45, alpha=1) +
  stat_boxplot(geom ='errorbar', width = 0.2, position=position_dodge(0.8), alpha=0.8) +
  scale_y_continuous(expand=c(0,0),"", limits = c(0,0.55)) +
  scale_x_discrete(expand=c(0,0),"", labels = c("Blood", "Biopsy")) +
  scale_fill_brewer(palette="Dark2",name = "", labels = c("Controls", "Rejectors")) +
  ggtitle("TRBB") +
  theme_bw() +
  theme(legend.position="none", plot.title = element_text(hjust = 0.5), text = element_text(size=14))
```

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap="Boxplot of R20 estimates in the blood and biopsy repertoire samples grouped by rejection"}

filename <- paste0("plot_R20_weinberger")
p13 <- ggarrange(p1, p3, ncol=2, nrow=1, common.legend = TRUE, legend="bottom")
p13
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p13, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)

```


```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE, fig.align="center", fig.width=6, fig.height=3, fig.cap="Boxplot of clonality in the blood and biopsy repertoire samples grouped by rejection"}

filename <- paste0("plot_clonality_weinberger")
p24 <- ggarrange(p2, p4, ncol=2, nrow=1, common.legend = TRUE, legend="bottom")
p24
ggsave(filename = paste(out.path,filename,".png", sep = ""), plot=p24, width = 30, height = 20, dpi = 640, units="cm", limitsize = F)

```



The unpaired t test was performed to investigate whether there exist a statistically significant difference in $\Delta R20=R20_{BX}-R20_{BL}$, $\Delta R20^{allo}=R20^{allo}_{BX}-R20^{allo}_{BL}$ and $\Delta C=C_{BX}-C_{BL}$ (C...Clonality) 
between the controls and the rejectors.

```{r echo=FALSE, cache=T, message=FALSE, warning=FALSE}

tmpA.new <- merge(tmp.TRBA[tmp.TRBA$stage %in% "Blood",c("patID", "stage", "rejection", "R20",  "Clonality")],
                  tmp.TRBA[tmp.TRBA$stage %in% "Biopsy",c("patID", "stage", "rejection", "R20", "Clonality")], by="patID")
tmpA.new$deltaR20 <- tmpA.new$R20.y-tmpA.new$R20.x
tmpA.new$deltaClonality <- tmpA.new$Clonality.y-tmpA.new$Clonality.x

tmpB.new <- merge(tmp.TRBB[tmp.TRBB$stage %in% "Blood",c("patID", "stage", "rejection", "R20",  "Clonality")], 
                  tmp.TRBB[tmp.TRBB$stage %in% "Biopsy",c("patID", "stage", "rejection", "R20",  "Clonality")], by="patID")
tmpB.new$deltaR20 <- tmpB.new$R20.y-tmpB.new$R20.x
tmpB.new$deltaClonality <- tmpB.new$Clonality.y-tmpB.new$Clonality.x

perform_ttests<- function(x, y){
  test <- t.test(x,y)
  
  res <-c("t"=test$statistic, test$parameter, "p"=test$p.value, "dd.value"=as.numeric(test$estimate[1]- test$estimate[2]),  "ci.lo"=test$conf.int[1], "ci.up"=test$conf.int[2])
  return(res)
}

res.ttests <- matrix(NA, nrow=4, ncol=6)
res.ttests[1,] <- perform_ttests(x=tmpA.new[tmpA.new$rejection.x %in% "Controls",]$deltaR20, 
                                 y=tmpA.new[tmpA.new$rejection.x %in% "Rejectors",]$deltaR20)
res.ttests[2,] <- perform_ttests(x=tmpA.new[tmpA.new$rejection.x %in% "Controls",]$deltaClonality, 
                                 y=tmpA.new[tmpA.new$rejection.x %in% "Rejectors",]$deltaClonality)

res.ttests[3,] <- perform_ttests(x=tmpB.new[tmpB.new$rejection.x %in% "Controls",]$deltaR20, 
                                 y=tmpB.new[tmpB.new$rejection.x %in% "Rejectors",]$deltaR20)
res.ttests[4,] <- perform_ttests(x=tmpB.new[tmpB.new$rejection.x %in% "Controls",]$deltaClonality, 
                                 y=tmpB.new[tmpB.new$rejection.x %in% "Rejectors",]$deltaClonality)
colnames(res.ttests) <- c("t", "df", "p", "dd.value", "ci.lo", "ci.up")
res.ttests <- round(res.ttests, 3)

res <- data.frame("Primer mix"=c(rep("TRBA",2), rep("TRBB",2)), "Variable"=rep(c("deltaR20", "deltaClonality"),2),res.ttests)

res %>%
  unite("95% CI", ci.lo:ci.up, sep="-") %>%
  `colnames<-`(c("Primer mix","Variable", "t", "df", "p", "Mean Difference", "95% CI")) %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Results of the two-sample t tests for the diversity estimates between rejectors and controls for TRBA and TRBB", autonum = tbl_autonum)
  

#- Range of diversity estimates
div.df <- data.frame("Pheno"=c(rep("TRBA",4),rep("TRBB",4)), "Stage"=rep(c("Blood","Biopsy"),4), 
                     "Variable"=rep(c("R20","R20","Clonality","Clonality"),2),
                     rbind(summary(tmp.TRBA[tmp.TRBA$stage %in% "Blood",]$R20),summary(tmp.TRBA[tmp.TRBA$stage %in% "Biopsy",]$R20), 
      summary(tmp.TRBA[tmp.TRBA$stage %in% "Blood",]$Clonality), summary(tmp.TRBA[tmp.TRBA$stage %in% "Biopsy",]$Clonality),
      summary(tmp.TRBB[tmp.TRBB$stage %in% "Blood",]$R20),summary(tmp.TRBB[tmp.TRBB$stage %in% "Biopsy",]$R20), 
      summary(tmp.TRBB[tmp.TRBB$stage %in% "Blood",]$Clonality), summary(tmp.TRBB[tmp.TRBB$stage %in% "Biopsy",]$Clonality)))
div.df[,4:9] <- apply(div.df[,4:9],2,function(x) round(x,3))

div.df %>%
  flextable()  %>%
  bold(part = "header") %>%
  autofit()  %>% fit_to_width(7)  %>% set_table_properties(layout = "autofit") %>%
  set_caption("Distribution of the diversity estimates across the immune repertoires", autonum = tbl_autonum)


```




